// @momentum-design/components/dist/components/buttongroup/buttongroup.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { Component } from '../../models';
import { DEFAULTS } from './buttongroup.constants';
import styles from './buttongroup.styles';
/**
 * `mdc-buttongroup` is a wrapper component that groups multiple buttons together.
 *
 * ## Supported button types
 * - Icon buttons
 * - Pill buttons
 * - Combination of icon and pill buttons
 *
 * ## Features
 * - Horizontal or vertical orientation
 * - Unified size and variant applied to all child buttons
 * - Optional compact mode for reduced height
 * - Supported sizes: 24, 28 (default), 32, 40
 *
 * ## Usage
 * - Only `mdc-button` components are allowed as direct children
 * - The `size` and `variant` set on buttongroup override individual button settings
 *
 * @tagname mdc-buttongroup
 *
 * @slot default - This is a default/unnamed slot, which contains the buttons
 *
 * @cssproperty --mdc-buttongroup-border-radius - The border radius of the buttongroup
 * @cssproperty --mdc-buttongroup-border-color - The border color of the buttongroup
 * @cssproperty --mdc-buttongroup-divider-color - The color of the divider between buttons within the buttongroup
 *
 * @csspart container - The container of the buttongroup.
 */
class ButtonGroup extends Component {
    constructor() {
        super(...arguments);
        /**
         * Layout direction of the buttons within the group.
         * - `horizontal`: Buttons are arranged side by side (default)
         * - `vertical`: Buttons are stacked vertically
         * @default 'horizontal'
         */
        this.orientation = DEFAULTS.ORIENTATION;
        /**
         * Visual style variant applied to all buttons in the group.
         * - `primary`: Solid background color
         * - `secondary`: Transparent background with solid border
         * - `tertiary`: No background or border, text-only appearance
         * @default 'primary'
         */
        this.variant = DEFAULTS.VARIANT;
        /**
         * Size applied to all buttons in the group.
         * Available sizes: 24, 28, 32, 40
         * @default 28
         */
        this.size = DEFAULTS.SIZE;
        /**
         * Enables compact mode with reduced height.
         * When true, all buttons have a fixed height of 24px regardless of their size setting.
         * Useful for space-constrained layouts.
         * @default false
         */
        this.compact = false;
    }
    /**
     * Handles the slotchange event, setting the size and variant of the buttons
     * @internal
     */
    handleSlotChange() {
        this.buttons.forEach(button => {
            button.setAttribute('size', this.size.toString());
            button.setAttribute('variant', this.variant);
        });
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('size') || changedProperties.has('variant')) {
            this.handleSlotChange();
        }
    }
    render() {
        return html `<div part="container"><slot @slotchange=${this.handleSlotChange}></slot></div>`;
    }
}
ButtonGroup.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], ButtonGroup.prototype, "orientation", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], ButtonGroup.prototype, "variant", void 0);
__decorate([
    property({ type: Number, reflect: true }),
    __metadata("design:type", Number)
], ButtonGroup.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], ButtonGroup.prototype, "compact", void 0);
__decorate([
    queryAssignedElements({ selector: 'mdc-button' }),
    __metadata("design:type", Array)
], ButtonGroup.prototype, "buttons", void 0);
export default ButtonGroup;
