// @momentum-design/components/dist/components/buttongroup/buttongroup.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { Component } from '../../models';
import { DEFAULTS } from './buttongroup.constants';
import styles from './buttongroup.styles';
/**
 * buttongroup component, is a styled wrapper for multiple buttons.
 * It can support icon buttons, combination of icon and pill buttons, and text buttons.
 * They are available in horizontal and vertical orientation.
 *
 * @tagname mdc-buttongroup
 *
 * @slot default - This is a default/unnamed slot, which contains the buttons
 *
 * @cssproperty --mdc-buttongroup-border-radius - The border radius of the buttongroup
 * @cssproperty --mdc-buttongroup-border-color - The border color of the buttongroup
 * @cssproperty --mdc-buttongroup-divider-color - The color of the divider between buttons within the buttongroup
 */
class ButtonGroup extends Component {
    constructor() {
        super(...arguments);
        /**
         * Orientation of the buttongroup.
         * @default 'horizontal'
         */
        this.orientation = DEFAULTS.ORIENTATION;
        /**
         * Variant of the buttons within the buttongroup.
         * @default 'primary'
         */
        this.variant = DEFAULTS.VARIANT;
        /**
         * Size of the buttons within the buttongroup.
         * @default '28'
         */
        this.size = DEFAULTS.SIZE;
        /**
         * When this is true, the buttons within the buttongroup will be compact.
         * i.e. Irrespective of the size of the buttons, they will have a height of 24px.
         * @default false
         */
        this.compact = false;
    }
    /**
     * Handles the slotchange event, setting the size and variant of the buttons
     * @internal
     */
    handleSlotChange() {
        this.buttons.forEach(button => {
            button.setAttribute('size', this.size.toString());
            button.setAttribute('variant', this.variant);
        });
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('size') || changedProperties.has('variant')) {
            this.handleSlotChange();
        }
    }
    render() {
        return html `<div part="container"><slot @slotchange=${this.handleSlotChange}></slot></div>`;
    }
}
ButtonGroup.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], ButtonGroup.prototype, "orientation", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], ButtonGroup.prototype, "variant", void 0);
__decorate([
    property({ type: Number, reflect: true }),
    __metadata("design:type", Number)
], ButtonGroup.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], ButtonGroup.prototype, "compact", void 0);
__decorate([
    queryAssignedElements({ selector: 'mdc-button' }),
    __metadata("design:type", Array)
], ButtonGroup.prototype, "buttons", void 0);
export default ButtonGroup;
