// @momentum-design/components/dist/react/password/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/password';
import { TAG_NAME } from '../../components/password/password.constants';
/**
 * `mdc-password` is a component that allows users to input their password.
 *  It extends the `mdc-input` component and provides additional features specific to password fields.
 *  It contains:
 * - `label` field - describe the password field.
 * - `password` field - contains the value
 * - `help-text` or `validation-message` - displayed below the password field.
 * - `help-text-type` - type of the help text, can be 'default', 'error', 'warning', 'success', 'priority'.
 * - `show-hide-button-aria-label` - aria label for the trailing show/hide button.
 * - all the attributes of the native password field.
 *
 * @tagname mdc-password
 *
 * @event input - (React: onInput) This event is dispatched when the value of the password field changes (every press).
 * @event change - (React: onChange) This event is dispatched when the value of the password field changes (on blur).
 * @event focus - (React: onFocus) This event is dispatched when the password receives focus.
 * @event blur - (React: onBlur) This event is dispatched when the password loses focus.
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-button
 * @dependency mdc-toggletip
 *
 * @cssproperty --mdc-input-disabled-border-color - Border color for the password container when disabled
 * @cssproperty --mdc-input-disabled-text-color - Text color for the password field when disabled
 * @cssproperty --mdc-input-disabled-background-color - Background color for the password field when disabled
 * @cssproperty --mdc-input-border-color - Border color for the password container
 * @cssproperty --mdc-input-text-color - Text color for the password field
 * @cssproperty --mdc-input-background-color - Background color for the password field
 * @cssproperty --mdc-input-selection-background-color - Background color for the selected text
 * @cssproperty --mdc-input-selection-text-color - Text color for the selected text
 * @cssproperty --mdc-input-support-text-color - Text color for the help text
 * @cssproperty --mdc-input-hover-background-color - Background color for the password field when hovered
 * @cssproperty --mdc-input-focused-background-color - Background color for the password field when focused
 * @cssproperty --mdc-input-focused-border-color - Border color for the password container when focused
 * @cssproperty --mdc-input-error-border-color - Border color for the password container when error
 * @cssproperty --mdc-input-warning-border-color - Border color for the password container when warning
 * @cssproperty --mdc-input-success-border-color - Border color for the password container when success
 * @cssproperty --mdc-input-primary-border-color - Border color for the password container when primary
 *
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onInput: 'input',
        onChange: 'change',
        onFocus: 'focus',
        onBlur: 'blur',
        onClear: 'clear',
    },
    displayName: 'Password',
});
export default reactWrapper;
