// @momentum-design/components/dist/react/datepicker/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/datepicker';
import { TAG_NAME } from '../../components/datepicker/datepicker.constants';
/**
 * mdc-datepicker is a component that allows users to select a date
 * or enter a date manually. It supports two variants:
 *
 * - `input`: Manual date entry with spinbutton inputs for month, day, and year
 * - `default`: A select-like trigger that opens a calendar for date range selection
 *
 * The component uses `mdc-calendar` inside a popover for date selection
 * and supports localization via a BCP 47 locale string.
 *
 * @tagname mdc-datepicker
 *
 * @dependency mdc-button
 * @dependency mdc-calendar
 * @dependency mdc-icon
 * @dependency mdc-popover
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @event input - (React: onInput) This event is dispatched when the date value changes.
 * @event change - (React: onChange) This event is dispatched when the date value is committed.
 * @event focus - (React: onFocus) This event is dispatched when the datepicker receives focus.
 * @event blur - (React: onBlur) This event is dispatched when the datepicker loses focus.
 *
 * @slot label - Slot for the label element.
 * @slot toggletip - Slot for the toggletip info icon button.
 * @slot help-icon - Slot for the helper/validation icon.
 * @slot help-text - Slot for the helper/validation text.
 *
 * @cssproperty --mdc-datepicker-background-color - Background color of the datepicker input.
 * @cssproperty --mdc-datepicker-border-color - Border color of the datepicker input.
 * @cssproperty --mdc-datepicker-text-color - Text color of the datepicker input.
 * @cssproperty --mdc-datepicker-height - The height of the datepicker trigger control.
 * @cssproperty --mdc-datepicker-width - Width of the datepicker component.
 *
 * @csspart label - The label element.
 * @csspart label-text - The container for the label and required indicator elements.
 * @csspart required-indicator - The required indicator element.
 * @csspart info-icon-btn - The info icon button element.
 * @csspart label-toggletip - The toggletip element.
 * @csspart help-text - The helper/validation text element.
 * @csspart helper-icon - The helper/validation icon element.
 * @csspart help-text-container - The container for helper/validation elements.
 * @csspart container - The outer container for the input and popover.
 * @csspart base-container - The input container with border and background.
 * @csspart spinbutton-group - The container for spinbutton elements.
 * @csspart spinbutton - A spinbutton input element.
 * @csspart separator - The separator between spinbuttons.
 * @csspart icon-container - The calendar/dropdown button container.
 * @csspart native-input - The hidden native input for form participation.
 * @csspart select-text - The selected value display text (default variant).
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onInput: 'input',
        onChange: 'change',
        onFocus: 'focus',
        onBlur: 'blur',
    },
    displayName: 'DatePicker',
});
export default reactWrapper;
