// @momentum-design/components/dist/react/calendar/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/calendar';
import { TAG_NAME } from '../../components/calendar/calendar.constants';
/**
 * mdc-calendar is a component that displays a monthly calendar grid
 * allowing users to select dates, weeks, or date ranges.
 *
 * The component supports three selection modes:
 * - `single`: Select a single date
 * - `week`: Select an entire week
 * - `range`: Select a date range (start and end)
 *
 * @tagname mdc-calendar
 *
 * @dependency mdc-button
 *
 * @event date-selected - Fires when a date or date range is selected.
 * @event month-changed - Fires when the displayed month changes.
 *
 * @cssproperty --mdc-calendar-background-color - Background color of the calendar.
 * @cssproperty --mdc-calendar-text-color - Text color of the calendar.
 * @cssproperty --mdc-calendar-day-selected-bg - Background color of selected days.
 * @cssproperty --mdc-calendar-day-selected-text-color - Text color of selected days.
 * @cssproperty --mdc-calendar-day-today-border-color - Border color for today's date.
 * @cssproperty --mdc-calendar-day-hover-bg - Background color on hover.
 * @cssproperty --mdc-calendar-range-bg - Background color for days within a range.
 * @cssproperty --mdc-calendar-day-outside-month-text-color - Text color for days outside the current month.
 * @cssproperty --mdc-calendar-day-disabled-text-color - Text color for disabled days.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onMonthChanged: 'month-changed',
        onDateSelected: 'date-selected',
    },
    displayName: 'Calendar',
});
export default reactWrapper;
