// @momentum-design/components/dist/components/toggletip/toggletip.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, queryAssignedNodes, state } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import Popover from '../popover/popover.component';
import { POPOVER_PLACEMENT } from '../popover/popover.constants';
import { DEFAULTS } from './toggletip.constants';
import styles from './toggletip.styles';
/**
 * A toggletip is triggered by clicking a trigger element.
 *
 * It can contain interactive content and can be closed by
 * clicking outside the toggletip or pressing the escape key.
 *
 * It can have optional close button to close the toggletip.
 *
 * Toggletip component uses `mdc-screenreaderannouncer` internally to
 * announce the toggletip text content with screen readers when the toggletip is shown.
 *
 * `screenreader-announcer-identity` attribute can be used to provide ID of an element
 * in DOM to which announcement elements are added. If not set, a visually hidden
 * div element is created in DOM to which announcement elements are added.
 *
 * Please refer to the `mdc-screenreaderannouncer` component for more details.
 *
 * @dependency mdc-screenreaderannouncer
 * @dependency mdc-button
 *
 * @tagname mdc-toggletip
 *
 * @event shown - (React: onShown) This event is dispatched when the toggletip is shown
 * @event hidden - (React: onHidden) This event is dispatched when the toggletip is hidden
 * @event created - (React: onCreated) This event is dispatched when the toggletip is created (added to the DOM)
 * @event destroyed - (React: onDestroyed) This event is dispatched when the toggletip
 * is destroyed (removed from the DOM)
 *
 * @cssproperty --mdc-toggletip-max-width - The maximum width of the toggletip.
 * @cssproperty --mdc-toggletip-text-color - The text color of the toggletip.
 * @cssproperty --mdc-toggletip-text-color-contrast - The text color of the toggletip
 *  when the color is contrast.
 *
 * @slot - Default slot for the toggletip content
 */
class ToggleTip extends Popover {
    constructor() {
        super(...arguments);
        /** @internal */
        this.currentAnnouncement = '';
        /**
         * The placement of the popover.
         * - **top**
         * - **top-start**
         * - **top-end**
         * - **bottom**
         * - **bottom-start**
         * - **bottom-end**
         * - **left**
         * - **left-start**
         * - **left-end**
         * - **right**
         * - **right-start**
         * - **right-end**
         * @default bottom
         */
        this.placement = DEFAULTS.PLACEMENT;
    }
    connectedCallback() {
        var _a;
        super.connectedCallback();
        this.closeButton = (_a = this.closeButton) !== null && _a !== void 0 ? _a : DEFAULTS.CLOSE_BUTTON;
        this.closeButtonAriaLabel = DEFAULTS.CLOSE_BUTTON_ARIA_LABEL;
        this.trigger = DEFAULTS.CLICK;
        this.showArrow = DEFAULTS.SHOW_ARROW;
        this.interactive = true;
        this.backdrop = true;
        this.hideOnBlur = true;
        this.hideOnEscape = true;
        this.hideOnOutsideClick = true;
        this.disableAriaExpanded = false;
        this.focusBackToTrigger = true;
        this.focusTrap = true;
    }
    /**
     * @returns The text content of all the nodes in the default slot, joined by a space.
     *          If there are no nodes, an empty string is returned.
     */
    getToggleTipText() {
        var _a, _b;
        return (((_b = (_a = this.defaultSlotNodes) === null || _a === void 0 ? void 0 : _a.map((node) => node.textContent).join(' ')) === null || _b === void 0 ? void 0 : _b.trim()) || '');
    }
    /**
     * Updates the placement attribute if it is not a valid placement.
     * Default placement for toggle tip is top.
     */
    onPlacementUpdated() {
        if (!Object.values(POPOVER_PLACEMENT).includes(this.placement)) {
            this.placement = DEFAULTS.PLACEMENT;
        }
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('placement')) {
            this.onPlacementUpdated();
        }
        if (changedProperties.has('visible')) {
            this.currentAnnouncement = this.visible ? this.getToggleTipText() : '';
        }
    }
    render() {
        return html `
      ${super.render()}
      <mdc-screenreaderannouncer
        identity="${ifDefined(this.screenreaderAnnouncerIdentity)}"
        announcement="${this.currentAnnouncement}"
        delay="300"
      >
      </mdc-screenreaderannouncer>
    `;
    }
}
ToggleTip.styles = [...Popover.styles, ...styles];
__decorate([
    queryAssignedNodes(),
    __metadata("design:type", Array)
], ToggleTip.prototype, "defaultSlotNodes", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], ToggleTip.prototype, "currentAnnouncement", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'screenreader-announcer-identity' }),
    __metadata("design:type", String)
], ToggleTip.prototype, "screenreaderAnnouncerIdentity", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], ToggleTip.prototype, "placement", void 0);
export default ToggleTip;
