// @momentum-design/components/dist/react/toggle/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/toggle';
import { TAG_NAME } from '../../components/toggle/toggle.constants';
/**
 * Toggle Component is an interactive control used to switch between two mutually exclusive options,
 * such as On/Off, Active/Inactive. These are commonly used in settings panels, forms, and preference selections
 * where users need to enable or disable a feature.
 * It contains an optional label and an optional helper text.
 *
 * To create a group of toggles, use the FormFieldGroup component.
 *
 * Note: It internally renders a checkbox styled as a toggle switch.
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-statictoggle
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @tagname mdc-toggle
 *
 * @event change - (React: onChange) Event that gets dispatched when the toggle state changes.
 * @event focus - (React: onFocus) Event that gets dispatched when the toggle receives focus.
 *
 * @cssproperty --mdc-toggle-width - Width of the toggle
 * @cssproperty --mdc-toggle-height - Height of the toggle
 * @cssproperty --mdc-toggle-width-compact - Width of the toggle when it's size is compact
 * @cssproperty --mdc-toggle-height-compact - Height of the toggle when it's size is compact
 * @cssproperty --mdc-toggle-label-lineheight - Line height of the toggle label
 * @cssproperty --mdc-toggle-label-fontsize - Font size of the toggle label
 * @cssproperty --mdc-toggle-label-fontweight - Font weight of the toggle label
 * @cssproperty --mdc-toggle-label-color-disabled - Color of the toggle label and help text in disabled state
 * @cssproperty --mdc-toggle-help-text-color - Color of the help text label
 * @cssproperty --mdc-toggle-active-hover-color - Background color of the active toggle in hover state
 * @cssproperty --mdc-toggle-active-pressed-color - Background color of the active toggle in pressed state
 * @cssproperty --mdc-toggle-inactive-hover-color - Background color of the inactive toggle in hover state
 * @cssproperty --mdc-toggle-inactive-pressed-color - Background color of the inactive toggle in pressed state
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change',
        onFocus: 'focus',
    },
    displayName: 'Toggle',
});
export default reactWrapper;
