// @momentum-design/components/dist/components/toggle/toggle.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { KEYS } from '../../utils/keys';
import { DataAriaLabelMixin } from '../../utils/mixins/DataAriaLabelMixin';
import { FormInternalsMixin } from '../../utils/mixins/FormInternalsMixin';
import { ROLE } from '../../utils/roles';
import FormfieldWrapper from '../formfieldwrapper';
import { DEFAULTS as FORMFIELD_DEFAULTS } from '../formfieldwrapper/formfieldwrapper.constants';
import { DEFAULTS, TOGGLE_SIZE } from './toggle.constants';
import styles from './toggle.styles';
/**
 * Toggle Component is an interactive control used to switch between two mutually exclusive options,
 * such as On/Off, Active/Inactive. These are commonly used in settings panels, forms, and preference selections
 * where users need to enable or disable a feature.
 * It contains an optional label and an optional helper text.
 *
 * To create a group of toggles, use the FormFieldGroup component.
 *
 * Note: It internally renders a checkbox styled as a toggle switch.
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-statictoggle
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @tagname mdc-toggle
 *
 * @event change - (React: onChange) Event that gets dispatched when the toggle state changes.
 * @event focus - (React: onFocus) Event that gets dispatched when the toggle receives focus.
 *
 * @cssproperty --mdc-toggle-width - Width of the toggle
 * @cssproperty --mdc-toggle-height - Height of the toggle
 * @cssproperty --mdc-toggle-width-compact - Width of the toggle when it's size is compact
 * @cssproperty --mdc-toggle-height-compact - Height of the toggle when it's size is compact
 * @cssproperty --mdc-toggle-label-lineheight - Line height of the toggle label
 * @cssproperty --mdc-toggle-label-fontsize - Font size of the toggle label
 * @cssproperty --mdc-toggle-label-fontweight - Font weight of the toggle label
 * @cssproperty --mdc-toggle-label-color-disabled - Color of the toggle label and help text in disabled state
 * @cssproperty --mdc-toggle-help-text-color - Color of the help text label
 * @cssproperty --mdc-toggle-active-hover-color - Background color of the active toggle in hover state
 * @cssproperty --mdc-toggle-active-pressed-color - Background color of the active toggle in pressed state
 * @cssproperty --mdc-toggle-inactive-hover-color - Background color of the inactive toggle in hover state
 * @cssproperty --mdc-toggle-inactive-pressed-color - Background color of the inactive toggle in pressed state
 */
class Toggle extends FormInternalsMixin(DataAriaLabelMixin(FormfieldWrapper)) {
    constructor() {
        super(...arguments);
        /**
         * Determines whether the toggle is active or inactive.
         * @default false
         */
        this.checked = false;
        /**
         * Determines toggle size in rem (height is specified here).
         * - **Default**: 1.5
         * - **Compact**: 1
         * @default default
         */
        this.size = DEFAULTS.SIZE;
        /**
         * Automatically focus on the element when the page loads.
         * [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autofocus)
         * @default false
         */
        this.autofocus = false;
    }
    connectedCallback() {
        super.connectedCallback();
        // Toggle does not contain helpTextType property.
        this.helpTextType = undefined;
    }
    /** @internal
     * Resets the checkbox to its initial state.
     * The checked property is set to false.
     */
    formResetCallback() {
        this.checked = false;
    }
    /** @internal */
    formStateRestoreCallback(state) {
        if (state) {
            this.checked = true;
        }
    }
    /**
     * Manages the required state of the checkbox.
     * If the checkbox is not checked and the required property is set, then the checkbox is invalid.
     * If the validationMessage is set, it will be used as the custom validity message.
     * If the validationMessage is not set, it will clear the custom validity message.
     */
    manageRequired() {
        if (!this.checked && this.required) {
            if (this.validationMessage) {
                this.inputElement.setCustomValidity(this.validationMessage);
            }
            else {
                this.inputElement.setCustomValidity('');
            }
            this.setValidity();
        }
        else {
            this.internals.setValidity({});
        }
    }
    /**
     * Updates the form value to reflect the current state of the toggle.
     * If toggle is switched on, the value is set to either the user-provided value or 'isActive' if no value is provided.
     * If toggle is switched off, the value is set to null.
     */
    setFormValue() {
        let actualValue = null;
        if (this.checked) {
            actualValue = !this.value ? 'isActive' : this.value;
        }
        else {
            actualValue = null;
        }
        this.manageRequired();
        this.internals.setFormValue(actualValue);
    }
    /**
     * Toggles the state of the toggle element.
     * If the element is not disabled, then the checked property is toggled.
     */
    toggleState() {
        if (!this.disabled) {
            this.checked = !this.checked;
        }
    }
    /**
     * Handles the keydown event on the toggle element.
     * When the user presses Enter, the form is submitted.
     * @param event - The keyboard event.
     */
    handleKeyDown(event) {
        var _a;
        if (event.key === KEYS.ENTER) {
            (_a = this.form) === null || _a === void 0 ? void 0 : _a.requestSubmit();
        }
    }
    /**
     * Toggles the state of the toggle element.
     * and dispatch the new change event.
     */
    handleChange(event) {
        this.toggleState();
        const EventConstructor = event.constructor;
        this.dispatchEvent(new EventConstructor(event.type, event));
    }
    /**
     * Sets the size attribute for the toggle component.
     * If the provided size is not included in the TOGGLE_SIZE,
     * it defaults to the value specified in DEFAULTS.SIZE.
     *
     * @param size - The size to set.
     */
    setToggleSize(size) {
        this.setAttribute('size', Object.values(TOGGLE_SIZE).includes(size) ? size : DEFAULTS.SIZE);
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('checked')) {
            this.setFormValue();
        }
        if (changedProperties.has('size')) {
            this.setToggleSize(this.size);
        }
    }
    render() {
        var _a;
        return html `
      <mdc-statictoggle
        ?checked="${this.checked}"
        ?disabled="${this.disabled}"
        size="${this.size}"
        class="mdc-focus-ring"
        part="container"
      >
        <input
          id="${this.inputId}"
          type="checkbox"
          part="toggle-input"
          role="${ROLE.CHECKBOX}"
          ?autofocus="${this.autofocus}"
          ?required="${this.required}"
          name="${ifDefined(this.name)}"
          value="${ifDefined(this.value)}"
          .checked="${this.checked}"
          aria-checked="${this.checked}"
          .disabled="${this.disabled}"
          aria-describedby="${ifDefined(this.helpText ? FORMFIELD_DEFAULTS.HELPER_TEXT_ID : '')}"
          aria-label="${(_a = this.dataAriaLabel) !== null && _a !== void 0 ? _a : ''}"
          tabindex="${this.disabled ? -1 : 0}"
          @change="${this.handleChange}"
          @keydown="${this.handleKeyDown}"
        />
      </mdc-statictoggle>
      ${this.renderLabel()} ${this.renderHelperText()}
    `;
    }
}
Toggle.styles = [...FormfieldWrapper.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], Toggle.prototype, "checked", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Toggle.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], Toggle.prototype, "autofocus", void 0);
export default Toggle;
