// @momentum-design/components/dist/react/toggle/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/toggle';
import { TAG_NAME } from '../../components/toggle/toggle.constants';
/**
 * The Toggle component is an interactive control used to switch between two mutually exclusive states,
 * such as On/Off or Active/Inactive. It is commonly used in settings panels, forms, and preference selections
 * where users need to enable or disable a feature.
 *
 * To create a group of toggles, use the `mdc-formfieldgroup` component.
 *
 * **Note:** This component internally renders a native checkbox input element with the `switch` role, styled as a toggle switch.
 *
 * ## When to use
 * Use toggles for binary choices where the change takes effect immediately, such as enabling/disabling settings or features.
 *
 * ## Accessibility
 * - Provide clear labels that describe what the toggle controls
 * - Use `data-aria-label` when a visual label is not present
 * - Keyboard navigation: Space to toggle, Tab to navigate, Enter to submit form
 *
 * ## Styling
 * Use the `static-toggle` part to apply custom styles to the toggle switch visual element.
 * This part exposes the underlying [StaticToggle](?path=/docs/components-decorator-statictoggle--docs) component for advanced styling.
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-statictoggle
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @tagname mdc-toggle
 *
 * @event change - (React: onChange) Event that gets dispatched when the toggle state changes.
 * @event focus - (React: onFocus) Event that gets dispatched when the toggle receives focus.
 *
 * @csspart label - The label element.
 * @csspart label-text - The container for the label and required indicator elements.
 * @csspart required-indicator - The required indicator element that is displayed next to the label when the `required` property is set to true.
 * @csspart info-icon-btn - The info icon button element that is displayed next to the label when the `toggletip-text` property is set.
 * @csspart label-toggletip - The toggletip element that is displayed when the info icon button is clicked.
 * @csspart help-text - The helper/validation text element.
 * @csspart helper-icon - The helper/validation icon element that is displayed next to the helper/validation text.
 * @csspart help-text-container - The container for the helper/validation icon and text elements.
 * @csspart text-container - The container for the label and helper text elements.
 * @csspart static-toggle - The statictoggle that provides the visual toggle switch appearance.
 * @csspart toggle-input - The native input element with switch role that provides the interactive functionality.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change',
        onFocus: 'focus',
    },
    displayName: 'Toggle',
});
export default reactWrapper;
