// @momentum-design/components/dist/components/skeleton/skeleton.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { Component } from '../../models';
import { DEFAULTS } from './skeleton.constants';
import styles from './skeleton.styles';
/**
 * `mdc-skeleton` is a component that shows a grey placeholder area.
 * It provides visual feedback to users that content is being loaded.
 *
 * **Skeleton Variants:**
 * - **rectangular**: Default variant with 0.25rem border radius
 * - **rounded**: Has 0.5rem border radius
 * - **circular**: Has 50% border radius for circular shapes
 * - **button**: Optimized for button placeholders with 1.25rem border radius
 *
 * **Sizing Behavior:**
 * 1. If wrapping content, takes dimensions of wrapped content
 * 2. Otherwise grows to fill parent container
 *
 * @tagname mdc-skeleton
 *
 * @slot - Content to wrap (optional). When provided, skeleton takes dimensions of this content.
 *
 * @cssproperty --mdc-skeleton-background-color - background color of the skeleton
 * @cssproperty --mdc-skeleton-height - height of the skeleton
 * @cssproperty --mdc-skeleton-width - width of the skeleton
 */
class Skeleton extends Component {
    constructor() {
        super(...arguments);
        /**
         * The variant of skeleton to display
         * - **rectangular**: Default rectangular shape with 0.25rem border radius
         * - **rounded**: Rounded rectangle with 0.5rem border radius
         * - **circular**: Circular shape with 50% border radius
         * - **button**: Button placeholder with 1.25rem border radius
         * @default rectangular
         */
        this.variant = DEFAULTS.VARIANT;
    }
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('aria-hidden', 'true');
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.checkSlotContent();
    }
    checkSlotContent() {
        var _a;
        const slot = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('slot');
        if (slot) {
            const hasContent = slot.assignedNodes().length > 0;
            if (hasContent) {
                this.setAttribute('has-content', '');
            }
            else {
                this.removeAttribute('has-content');
            }
        }
    }
    render() {
        return html `<slot @slotchange=${this.checkSlotContent}></slot>`;
    }
}
/**
 * Styles associated with this component.
 */
Skeleton.styles = [...Component.styles, styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Skeleton.prototype, "variant", void 0);
export default Skeleton;
