// @momentum-design/components/dist/components/coachmark/coachmark.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import Popover from '../popover/popover.component';
import { DEFAULTS } from './coachmark.constants';
/**
 * A Coachmark is a contextual guidance component used to highlight a specific UI element and explain its purpose or functionality.
 * It is typically shown as part of onboarding, feature discovery, or progressive education within the product.
 *
 * Coachmark component based on top of the popover component,
 * with the default value of certain properties changed.
 * Coachmark component will always have an arrow attached to it.
 * The color of the coachmark is contrast.
 *
 * ## When to use
 * - Introducing a new or complex feature to users.
 * - Drawing attention to a UI element that might otherwise be missed.
 * - Providing step-by-step guidance in an onboarding or walkthrough flow.
 *
 * ## Accessibility
 * - The default role of coachmark is `"dialog"`, which can be changed via the `role` attribute.
 * - aria-label or aria-labelledby can be provided to give the coachmark an accessible name.
 *
 * @dependency mdc-popover
 *
 * @tagname mdc-coachmark
 *
 * @event shown - (React: onShown) This event is dispatched when the coachmark is shown
 * @event hidden - (React: onHidden) This event is dispatched when the coachmark is hidden
 * @event created - (React: onCreated) This event is dispatched when the coachmark is created (added to the DOM)
 * @event destroyed - (React: onDestroyed) This event is dispatched when the coachmark is
 * destroyed (removed from the DOM)
 *
 * @cssproperty --mdc-popover-arrow-border-radius - radius of the arrow border
 * @cssproperty --mdc-popover-arrow-border - border of the arrow
 * @cssproperty --mdc-popover-primary-background-color - primary background color of the popover
 * @cssproperty --mdc-popover-border-color - border color of the popover
 * @cssproperty --mdc-popover-inverted-background-color - inverted background color of the popover
 * @cssproperty --mdc-popover-inverted-border-color - inverted border color of the popover
 * @cssproperty --mdc-popover-inverted-text-color - inverted text color of the popover
 * @cssproperty --mdc-popover-elevation-3 - elevation of the popover
 *
 * @slot - Default slot for modal container
 */
class Coachmark extends Popover {
    constructor() {
        super(...arguments);
        /**
         * Determines the events that cause the Coachmark to show.
         * Multiple event names should be separated by spaces.
         * For example to allow both click and hover, use 'click mouseenter' as the trigger.
         * - **click**
         * - **mouseenter**
         * - **focusin**
         * - **manual**
         * @default manual
         */
        this.trigger = DEFAULTS.TRIGGER;
        /**
         * The arrow visibility of the Coachmark.
         * @default true
         */
        this.showArrow = DEFAULTS.ARROW;
        /**
         * The close button visibility of the Coachmark.
         * @default true
         */
        this.closeButton = DEFAULTS.CLOSE_BUTTON;
        /**
         * Disable aria-expanded attribute on trigger element.
         * @default true
         */
        this.disableAriaExpanded = DEFAULTS.DISABLE_ARIA_EXPANDED;
    }
    connectedCallback() {
        super.connectedCallback();
        this.color = DEFAULTS.COLOR;
        this.showArrow = DEFAULTS.ARROW;
    }
}
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Coachmark.prototype, "trigger", void 0);
__decorate([
    property({ type: Boolean, attribute: 'show-arrow' }),
    __metadata("design:type", Boolean)
], Coachmark.prototype, "showArrow", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'close-button' }),
    __metadata("design:type", Boolean)
], Coachmark.prototype, "closeButton", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'disable-aria-expanded' }),
    __metadata("design:type", Boolean)
], Coachmark.prototype, "disableAriaExpanded", void 0);
export default Coachmark;
