// @momentum-design/components/dist/components/illustration/illustration.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, state } from 'lit/decorators.js';
import { Component } from '../../models';
import providerUtils from '../../utils/provider';
import IllustrationProvider from '../illustrationprovider/illustrationprovider.component';
import { ROLE } from '../../utils/roles';
import styles from './illustration.styles';
import { svgFetch } from './illustration.utils';
import { DEFAULTS } from './illustration.constants';
/**
 * Illustration component that dynamically displays SVG illustrations based on a valid name.
 *
 * This component must be mounted within an `IllustrationProvider` component.
 *
 * The `IllustrationProvider` defines the source URL from which illustrations are consumed.
 * The `Illustration` component accepts a `name` attribute, which corresponds to
 * the file name of the illustration to be loaded from the specified URL.
 *
 * Once fetched, the illustration will be rendered. If the fetching process is unsuccessful,
 * no illustration will be displayed.
 *
 * Default sizing of the illustration is controlled by choosing a different illustration name, can be overridden with the `--mdc-illustration-size` CSS property.
 * Coloring of the illustration is currently baked into the svg, meaning that the illustration name determines
 * the coloring.
 *
 * Regarding accessibility, there are two types of illustrations: decorative and informative.
 *
 * ### Decorative Illustrations
 * - Decorative illustrations do not convey any essential information to the content of a page.
 * - They should be hidden from screen readers (SR) to prevent confusion for users.
 * - For decorative illustrations, an `aria-label` is not required, and the `role` will be set to null.
 *
 * ### Informative Illustrations
 * - Informative illustrations convey important information that is not adequately represented
 *   by surrounding text or components.
 * - They provide valuable context and must be announced by assistive technologies.
 * - For informative illustrations, an `aria-label` is required, and the `role` will be set to "img" automatically.
 * - If an `aria-label` is provided, the role will be set to 'img'; if it is absent,
 *   the role will be unset.
 *
 * @tagname mdc-illustration
 *
 * @cssproperty --mdc-illustration-size - Allows customization of the illustration size.
 *
 * @csspart illustration - The svg inside the illustration element.
 */
class Illustration extends Component {
    constructor() {
        super();
        /**
         * Name of the illustration (= filename)
         */
        this.name = DEFAULTS.NAME;
        /**
         * Aria-label attribute to be set for accessibility
         */
        this.ariaLabel = null;
        /**
         * Aria-labelledby attribute to be set for accessibility
         */
        this.ariaLabelledby = null;
        this.illustrationProviderContext = providerUtils.consume({
            host: this,
            context: IllustrationProvider.Context,
        });
        this.abortController = new AbortController(); // Initialize AbortController
    }
    /**
     * Parse the illustration string to an html element, set the attributes and
     * return the illustration element
     *
     * @param illustrationData - The illustration string to be parsed
     * @returns illustrationElement
     */
    prepareIllustrationElement(illustrationData) {
        const illustrationElement = new DOMParser().parseFromString(illustrationData, 'text/html').body.children[0];
        if (this.name) {
            illustrationElement.setAttribute('data-name', this.name);
        }
        illustrationElement.setAttribute('part', 'illustration');
        // set aria-hidden=true for SVG to avoid screen readers
        illustrationElement.setAttribute('aria-hidden', 'true');
        return illustrationElement;
    }
    /**
     * Fetches the illustration (currently only svg) and sets state and attributes once fetched successfully
     *
     * This method uses abortController.signal to cancel the fetch request when the component is disconnected or updated.
     * If the request is aborted after the fetch() call has been fulfilled but before the response body has been read,
     * then attempting to read the response body will reject with an AbortError exception.
     */
    async getIllustrationData() {
        if (this.illustrationProviderContext.value) {
            const { fileExtension, url, cacheName, illustrationSet, cacheStrategy } = this.illustrationProviderContext.value;
            if (illustrationSet === 'custom-illustrations' && url && fileExtension && this.name) {
                // function to abort the fetch request and create a new signal
                // (directly passing the abortcontroller to the fetch request per reference
                // will not work due to JS call-by-sharing behavior)
                const renewSignal = () => {
                    var _a;
                    (_a = this.abortController) === null || _a === void 0 ? void 0 : _a.abort();
                    this.abortController = new AbortController();
                    return this.abortController.signal;
                };
                // fetch illustration data (including caching logic)
                return svgFetch({
                    url,
                    name: this.name,
                    fileExtension,
                    cacheName,
                    cacheStrategy,
                    renewSignal,
                })
                    .then(illustrationData => {
                    // parse the fetched illustration string to an html element and set the attributes
                    const illustrationElement = this.prepareIllustrationElement(illustrationData);
                    this.handleIllustrationLoadedSuccess(illustrationElement);
                })
                    .catch(error => {
                    this.handleIllustrationLoadedFailure(error);
                });
            }
            if (illustrationSet === 'momentum-illustrations' && this.name) {
                // dynamic import of the lit template from the momentum illustrations package
                return import(`@momentum-design/illustrations/dist/ts/${this.name}.ts`)
                    .then(module => {
                    this.handleIllustrationLoadedSuccess(module.default());
                })
                    .catch(error => {
                    this.handleIllustrationLoadedFailure(error);
                });
            }
        }
        const noIllustrationProviderError = new Error('IllustrationProvider not found or not properly set up.');
        this.handleIllustrationLoadedFailure(noIllustrationProviderError);
        return Promise.reject(noIllustrationProviderError);
    }
    /**
     * Sets the illustrationData state to the fetched illustration.
     * Dispatches a 'load' event on the component once the illustration has been successfully loaded.
     * @param illustrationHtml - The illustration html element which has been fetched from the illustration provider.
     */
    handleIllustrationLoadedSuccess(illustrationHtml) {
        // update illustrationData state once fetched:
        this.illustrationData = illustrationHtml;
        // when illustration is fetched successfully, trigger illustration load event.
        const loadEvent = new Event('load', {
            bubbles: true,
            cancelable: true,
        });
        this.dispatchEvent(loadEvent);
    }
    /**
     * Dispatches an 'error' event on the component when the illustration fetching has failed.
     * This event bubbles and is cancelable.
     * The error detail is set to the error object.
     */
    handleIllustrationLoadedFailure(error) {
        const errorEvent = new CustomEvent('error', {
            bubbles: true,
            cancelable: true,
            detail: { error },
        });
        this.dispatchEvent(errorEvent);
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('name')) {
            // fetch illustration data if name changes:
            this.getIllustrationData().catch(err => {
                if (err.name !== 'AbortError' && this.onerror) {
                    this.onerror(err);
                }
            });
        }
        if (changedProperties.has('ariaLabel') || changedProperties.has('ariaLabelledby')) {
            this.role = this.ariaLabel || this.ariaLabelledby ? ROLE.IMG : null;
        }
    }
    disconnectedCallback() {
        var _a;
        super.disconnectedCallback();
        // abort the fetch request when the component is disconnected
        (_a = this.abortController) === null || _a === void 0 ? void 0 : _a.abort();
        this.abortController = undefined; // reset the abort controller
    }
    render() {
        return html ` ${this.illustrationData} `;
    }
}
Illustration.styles = [...Component.styles, ...styles];
__decorate([
    state(),
    __metadata("design:type", HTMLElement)
], Illustration.prototype, "illustrationData", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Illustration.prototype, "name", void 0);
__decorate([
    property({ type: String, attribute: 'aria-label' }),
    __metadata("design:type", Object)
], Illustration.prototype, "ariaLabel", void 0);
__decorate([
    property({ type: String, attribute: 'aria-labelledby' }),
    __metadata("design:type", Object)
], Illustration.prototype, "ariaLabelledby", void 0);
__decorate([
    state(),
    __metadata("design:type", AbortController)
], Illustration.prototype, "abortController", void 0);
export default Illustration;
