// @momentum-design/components/dist/components/statictoggle/statictoggle.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { Component } from '../../models';
import { DisabledMixin } from '../../utils/mixins/DisabledMixin';
import { DEFAULTS, ICON_NAME, ICON_SIZE_IN_REM } from './statictoggle.constants';
import styles from './statictoggle.styles';
/**
 * The StaticToggle component is a decorative, non-interactive toggle switch used for visual
 * presentation. It shares the same styling as the interactive `mdc-toggle` component but does
 * not provide user interaction functionality.
 *
 * This component supports multiple visual states: checked, disabled, readonly, soft-disabled,
 * and size variations (default, compact).
 *
 * ## When to use
 * Use StaticToggle to display toggle states in read-only contexts such as summary views,
 * confirmation screens, or as a building block within custom interactive components. For
 * interactive toggles, use `mdc-toggle` instead.
 *
 * ## Accessibility
 * - StaticToggle is decorative and non-interactive by design
 * - When used within parent components, ensure proper ARIA attributes are provided by the parent
 * - Do not use this as a replacement for interactive toggles in forms
 *
 * @tagname mdc-statictoggle
 *
 * @dependency mdc-icon
 *
 * @cssproperty --mdc-statictoggle-width - Width of the static toggle.
 * @cssproperty --mdc-statictoggle-height - Height of the static toggle.
 * @cssproperty --mdc-statictoggle-border-radius - Border radius of the static toggle.
 * @cssproperty --mdc-statictoggle-border-color - Border color of the static toggle.
 * @cssproperty --mdc-statictoggle-background-color - Background color of the static toggle.
 * @cssproperty --mdc-statictoggle-icon-color - Icon color of the static toggle.
 * @cssproperty --mdc-statictoggle-icon-background-color - Icon background color of the static toggle.
 *
 * @csspart slider - The slider part of the toggle.
 * @csspart toggle-icon - The icon part of the toggle.
 *
 * @slot - Default slot for slotted content (typically used by parent `mdc-toggle` for the checkbox input).
 */
class StaticToggle extends DisabledMixin(Component) {
    constructor() {
        super(...arguments);
        /**
         * Controls the visual state of the toggle.
         * When true, the toggle appears in the "on" or "active" state.
         * When false, the toggle appears in the "off" or "inactive" state.
         * @default false
         */
        this.checked = false;
        /**
         * Specifies the size of the toggle component.
         * - **default**: Standard size with 1.5rem height
         * - **compact**: Smaller size with 1rem height
         * @default default
         */
        this.size = DEFAULTS.SIZE;
        /**
         * When true, the toggle appears in a read-only visual state.
         * This is a decorative property that changes the visual appearance only.
         * @default false
         */
        this.readonly = false;
        /**
         * When true, the toggle appears in a soft-disabled visual state.
         * Soft-disabled elements maintain visual presence but appear non-interactive.
         * This is a decorative property that changes the visual appearance only.
         * @default false
         */
        this.softDisabled = false;
    }
    render() {
        return html `
      <slot></slot>
      <div part="slider">
        <mdc-icon
          name="${this.checked ? ICON_NAME.CHECKED : ICON_NAME.UNCHECKED}"
          part="toggle-icon"
          length-unit="rem"
          size="${ICON_SIZE_IN_REM[this.size]}"
        ></mdc-icon>
      </div>
    `;
    }
}
StaticToggle.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], StaticToggle.prototype, "checked", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], StaticToggle.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], StaticToggle.prototype, "readonly", void 0);
__decorate([
    property({ type: Boolean, attribute: 'soft-disabled', reflect: true }),
    __metadata("design:type", Object)
], StaticToggle.prototype, "softDisabled", void 0);
export default StaticToggle;
