// @momentum-design/components/dist/react/accordion/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/accordion';
import { TAG_NAME } from '../../components/accordion/accordion.constants';
/**
 * An accordion is a vertically stacked component with a header and expandable/collapsible body section.
 * The header displays a prefix icon, header text, optional control slots (leading and trailing), and a dedicated expand/collapse button.
 * Unlike `accordionbutton`, only the expand/collapse button is clickable and not the entire header.
 *
 * ## Header contains
 * - Optional prefix icon
 * - Header text (default H3, customizable via `data-aria-level`)
 * - Leading slot for controls (placed after header text)
 * - Trailing slot for controls (placed before expand/collapse button)
 * - Expand/collapse button (automatically positioned at the end)
 *
 * ## Body contains
 * - Default slot for any content
 *
 * The accordion supports different border styles through the `variant` attribute and different spacing through the `size` attribute.
 * An accordion can be disabled, which prevents all interactions including the expand/collapse button and any slotted controls.
 *
 * ## When to use
 * - Use `mdc-accordion` when you need additional interactive controls (chips, badges, buttons, icons) in the header.
 * - Use `mdc-accordionbutton` if you only need a simple clickable header without extra controls.
 *
 * ## Accessibility
 * - Always provide `open-button-aria-label` and `close-button-aria-label` for screen reader support
 * - Adjust `data-aria-level` based on heading hierarchy in your page
 * - Note: Screen readers may lose focus when toggling if accordion is expanded by default
 *
 * @tagname mdc-accordion
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @slot leading-header-text - The leading header text slot of the accordion on the header section. Placed after the prefix icon.
 * @slot leading-controls - The leading controls slot of the accordion on the header section. Placed after the header text.
 * @slot trailing-controls - The trailing controls slot of the accordion on the header section. Placed before the expand/collapse button.
 * @slot default - The default slot contains the body section of the accordion. User can place anything inside this body slot.
 *
 * @event shown - (React: onShown) This event is triggered when the accordion is toggled (expanded or collapsed).
 *
 * @cssproperty --mdc-accordionbutton-border-color - The border color of the accordion.
 * @cssproperty --mdc-accordionbutton-disabled-color - The disabled text color of the accordion.
 *
 * @csspart body-section - The body section of the accordion.
 * @csspart header-section - The header section of the accordion.
 * @csspart leading-header - The leading header of the accordion.
 * @csspart leading-header-text - The leading header text of the accordion button.
 * @csspart trailing-header - The trailing header of the accordion.
 * @csspart toggle-button - The expand/collapse toggle button of the accordion.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onShown: 'shown',
    },
    displayName: 'Accordion',
});
export default reactWrapper;
