// @momentum-design/components/dist/components/responsivesettingsprovider/responsivesettingsprovider.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import { Provider } from '../../models';
import ResponsiveSettingsContext from './responsiveSettingsContext';
import { DEFAULTS, VALID_MEDIA_VALUES, VALID_POPOVER_POSITIONING_VALUES } from './responsivesettingsprovider.constants';
/**
 * `mdc-responsivesettingsprovider` is a provider component that supplies responsive settings
 * context to its child components.
 *
 * This component does not make any assumptions about how the values are determined. Values can be set
 * based on media queries or other device detection mechanisms.This way consumer can mix and match
 * different settings depending on their target devices and use cases.
 *
 * For example, when the device has table screen size/resolution, but because it is fixed dialog like
 * popovers provide better user experience.
 *
 * ## Responsive settings
 *
 * ### Media
 *
 * Generic media type to enforce responsive behavior in child components.
 * Consumer component can use the media type from the context or
 * just use CSS selector like [media="mobile"] to apply responsive styles.
 *
 * It is "unknown" by default so components can fall back to there default behavior.
 *
 * ### Popover Positioning
 *
 * By default, popovers are positioned close to the trigger element. But on small screens (e.g.: mobile devices),
 * it is often better to show popovers/menus at the center of the screen like dialogs.
 *
 * ### Force Fullscreen Dialog
 *
 * Some components like dialogs can be shown in fullscreen mode on small screens for better user experience.
 *
 * @tagname mdc-responsivesettingsprovider
 */
class ResponsiveSettingsProvider extends Provider {
    constructor() {
        // initialise the context by running the Provider constructor:
        super({
            context: ResponsiveSettingsContext.context,
            initialValue: new ResponsiveSettingsContext({
                media: DEFAULTS.MEDIA,
                popoverPositioning: DEFAULTS.POPOVER_POSITIONING,
                forceFullscreenDialog: DEFAULTS.FORCE_FULLSCREEN_DIALOG,
            }),
        });
        /**
         * Determines whether dialogs should be forced to fullscreen mode.
         */
        this.forceFullscreenDialog = DEFAULTS.FORCE_FULLSCREEN_DIALOG;
        /** @internal */
        this.privatePopoverPositioning = DEFAULTS.POPOVER_POSITIONING;
        /** @internal */
        this.privateMedia = DEFAULTS.MEDIA;
    }
    /**
     * Context object of the ResponsiveSettingsProvider, to be consumed by child components
     */
    static get Context() {
        return ResponsiveSettingsContext.context;
    }
    /**
     * The popover positioning for responsive settings.
     * @default 'float'
     */
    set popoverPositioning(value) {
        if (VALID_POPOVER_POSITIONING_VALUES.includes(value)) {
            this.privatePopoverPositioning = value;
        }
    }
    get popoverPositioning() {
        return this.privatePopoverPositioning;
    }
    /**
     * The media type for responsive settings.
     * @default 'unknown'
     */
    set media(value) {
        if (VALID_MEDIA_VALUES.includes(value)) {
            this.privateMedia = value;
        }
    }
    get media() {
        return this.privateMedia;
    }
    updateContext() {
        var _a, _b, _c;
        if (((_a = this.context.value) === null || _a === void 0 ? void 0 : _a.media) !== this.privateMedia ||
            ((_b = this.context.value) === null || _b === void 0 ? void 0 : _b.popoverPositioning) !== this.privatePopoverPositioning ||
            ((_c = this.context.value) === null || _c === void 0 ? void 0 : _c.forceFullscreenDialog) !== this.forceFullscreenDialog) {
            this.context.value.media = this.media;
            this.context.value.popoverPositioning = this.popoverPositioning;
            this.context.value.forceFullscreenDialog = this.forceFullscreenDialog;
            this.context.updateObservers();
        }
    }
}
__decorate([
    property({ type: Boolean, attribute: 'force-fullscreen-dialog', reflect: true }),
    __metadata("design:type", Boolean)
], ResponsiveSettingsProvider.prototype, "forceFullscreenDialog", void 0);
__decorate([
    property({ type: String, attribute: 'popover-positioning', reflect: true }),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [String])
], ResponsiveSettingsProvider.prototype, "popoverPositioning", null);
__decorate([
    property({ type: String, attribute: 'media', reflect: true }),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [String])
], ResponsiveSettingsProvider.prototype, "media", null);
export default ResponsiveSettingsProvider;
