// @momentum-design/components/dist/components/menubar/menubar.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { queryAssignedElements } from 'lit/decorators.js';
import { Component } from '../../models';
import { ROLE } from '../../utils/roles';
import { POPOVER_PLACEMENT } from '../popover/popover.constants';
import { TAG_NAME as MENUPOPOVER_TAGNAME } from '../menupopover/menupopover.constants';
import { TAG_NAME as MENUSECTION_TAGNAME } from '../menusection/menusection.constants';
import { TAG_NAME as SIDENAV_TAGNAME } from '../sidenavigation/sidenavigation.constants';
import { KEYS } from '../../utils/keys';
import { popoverStack } from '../popover/popover.stack';
import { DEFAULTS, TAG_NAME as MENUBAR_TAGNAME } from './menubar.constants';
import styles from './menubar.styles';
/**
 * Menubar is a navigational menu component that provides a vertical fixed list of menu items,
 * with support for nested submenus and keyboard navigation. It serves as a container
 * for menu items and manages their interaction patterns, including:
 * - Keyboard navigation (Arrow keys, Home, End)
 * - Menu item activation (Enter/Space)
 * - Submenu toggling (Right/Left arrow keys)
 * - Focus management
 * - Integration with MenuPopover for nested menus
 *
 * A menubar will contain a set of menu items and their associated popovers.
 * Each menu item can have a popover for nested menus.
 *
 * The component automatically handles ARIA attributes and follows WAI-ARIA menu design patterns.
 * It works in conjunction with `mdc-menuitem` and `mdc-menupopover` to create accessible menu structures.
 *
 * **Note:** A menubar contains three types of menu items, including menuitem, menuitemradio and menuitemcheckbox. These menu items may optionally be nested in one or more group containers. Groups or items may optionally by separated with separator elements.
 *
 * `mdc-menubar` contains a group that wraps all its chilren passed within the default slot. This has been added to ensure the right accessibility behavior while using screen readers.
 *
 * @tagname mdc-menubar
 * @slot default - Contains the menu items and their associated popovers
 */
class MenuBar extends Component {
    constructor() {
        super();
        this.addEventListener('click', this.handleClick.bind(this));
        this.addEventListener('keydown', this.handleKeyDown.bind(this));
    }
    async connectedCallback() {
        var _a;
        super.connectedCallback();
        this.role = ROLE.MENUBAR;
        this.ariaOrientation = DEFAULTS.ORIENTATION;
        await this.updateComplete;
        // to make sure menusection dividers have the correct divider variant
        (_a = this.menusections) === null || _a === void 0 ? void 0 : _a.forEach(section => {
            section.setAttribute('divider-variant', 'gradient');
        });
    }
    /**
     * Returns all menuitem elements, including those nested inside menusection.
     */
    get menuItems() {
        var _a, _b;
        const slot = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('slot');
        const assigned = (_b = slot === null || slot === void 0 ? void 0 : slot.assignedElements({ flatten: true })) !== null && _b !== void 0 ? _b : [];
        const items = [];
        const collect = (el) => {
            if (el.role === ROLE.MENUITEM && !el.hasAttribute('disabled')) {
                items.push(el);
            }
            else if (el.tagName.toLowerCase() === MENUSECTION_TAGNAME) {
                Array.from(el.children).forEach(collect);
            }
        };
        assigned.forEach(collect);
        return items;
    }
    /** @internal */
    getVisibleSubMenusOfItems(items) {
        const subMenus = [];
        if (!items || items.length === 0)
            return subMenus;
        items.forEach(item => {
            var _a;
            const id = item.getAttribute('id');
            if (!id)
                return;
            const submenu = (_a = document.querySelector) === null || _a === void 0 ? void 0 : _a.call(document, `${MENUPOPOVER_TAGNAME}[triggerid="${id}"]`);
            if (submenu && submenu.visible) {
                subMenus.push(submenu);
            }
        });
        return subMenus;
    }
    /**
     * Returns all menupopovers, including direct and slotted ones .
     */
    getAllPopovers() {
        var _a, _b;
        if (!this.menuItems.length)
            return [];
        const root = (_b = (_a = this.menuItems[0]) === null || _a === void 0 ? void 0 : _a.closest(SIDENAV_TAGNAME)) !== null && _b !== void 0 ? _b : this;
        if (!root)
            return [];
        const popovers = [];
        const searchPopovers = (node) => {
            // Check direct popovers
            node.querySelectorAll(`${MENUPOPOVER_TAGNAME}`).forEach(el => {
                popovers.push(el);
            });
            // Handle slots
            node.querySelectorAll('slot').forEach(slot => {
                const assigned = slot.assignedElements({ flatten: true });
                assigned.forEach(searchPopovers);
            });
        };
        searchPopovers(root);
        return popovers;
    }
    update(changedProperties) {
        super.update(changedProperties);
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.resetTabIndexAndSetActiveTabIndex(this.menuItems);
        this.updatePopoverPlacement();
    }
    /**
     * Closes all other submenus on the top level.
     * This method is used to ensure that only one topolevel submenu is open at a time.
     * It finds all other menu items with submenus and closes their submenus.
     * @param target - The target menu item that was clicked.
     */
    handleClick(event) {
        const target = event.target;
        if (!target || !this.isTopLevelMenuItem(target))
            return;
        const otherMenuItemsOnTop = this.menuItems.filter(item => item !== target);
        const otherOpenSubMenus = this.getVisibleSubMenusOfItems(otherMenuItemsOnTop);
        otherOpenSubMenus.forEach(subMenu => {
            subMenu.hide();
        });
    }
    /**
     * Resets all list items tabindex to -1 and sets the tabindex of the
     * element at the given index to 0, effectively setting the active
     * element. This is used when navigating the list via keyboard.
     *
     * @param newIndex - The index of the new active element in the list.
     */
    resetTabIndexAndSetActiveTabIndex(menuItems) {
        menuItems.forEach((node, index) => {
            const newTabindex = index === 0 ? '0' : '-1';
            node === null || node === void 0 ? void 0 : node.setAttribute('tabindex', newTabindex);
        });
    }
    getCurrentIndex(target) {
        return this.menuItems.findIndex(item => item === target || item === target.parentElement);
    }
    getSubmenu(triggerId) {
        if (!triggerId)
            return undefined;
        return this.getAllPopovers().find(popover => popover.getAttribute('triggerid') === triggerId);
    }
    showSubmenu(triggerId) {
        if (!triggerId)
            return;
        const triggerElement = this.menuItems.find(item => item.getAttribute('id') === triggerId);
        if (!triggerElement)
            return;
        const submenu = this.getSubmenu(triggerId);
        if (submenu)
            submenu.show();
    }
    updatePopoverPlacement() {
        const placement = POPOVER_PLACEMENT.RIGHT_START;
        this.getAllPopovers().forEach(popover => {
            popover.setAttribute('placement', placement);
        });
    }
    updateTabIndexAndFocus(menuItems, currentIndex, newIndex) {
        var _a, _b, _c;
        if (newIndex < 0 || currentIndex < 0)
            return;
        if (currentIndex !== newIndex) {
            (_a = menuItems[currentIndex]) === null || _a === void 0 ? void 0 : _a.setAttribute('tabindex', '-1');
            (_b = menuItems[newIndex]) === null || _b === void 0 ? void 0 : _b.setAttribute('tabindex', '0');
        }
        (_c = menuItems[newIndex]) === null || _c === void 0 ? void 0 : _c.focus();
    }
    navigateToMenuItem(currentIndex, direction, shouldOpenSubmenu = false) {
        var _a;
        const { length } = this.menuItems;
        if (length === 0)
            return;
        let newIndex = currentIndex;
        newIndex = direction === 'prev' ? (currentIndex - 1 + length) % length : (currentIndex + 1) % length;
        this.updateTabIndexAndFocus(this.menuItems, currentIndex, newIndex);
        if (shouldOpenSubmenu) {
            const triggerId = (_a = this.menuItems[newIndex]) === null || _a === void 0 ? void 0 : _a.getAttribute('id');
            if (this.getSubmenu(triggerId) && !this.menuItems[newIndex].hasAttribute('soft-disabled')) {
                this.showSubmenu(triggerId);
            }
        }
    }
    getKeyWithDirectionFix(originalKey) {
        const isRtl = window.getComputedStyle(this).direction === 'rtl';
        if (!isRtl)
            return originalKey;
        if (originalKey === KEYS.ARROW_LEFT)
            return KEYS.ARROW_RIGHT;
        if (originalKey === KEYS.ARROW_RIGHT)
            return KEYS.ARROW_LEFT;
        return originalKey;
    }
    /**
     * Determines if a menuitem is a top-level menuitem (direct child of menubar or child of menusection whose parent is menubar)
     */
    isTopLevelMenuItem(element) {
        var _a;
        const parent = element.parentElement;
        if (!parent || element.role !== ROLE.MENUITEM)
            return false;
        const parentTag = parent.tagName.toLowerCase();
        const grandParentTag = (_a = parent.parentElement) === null || _a === void 0 ? void 0 : _a.tagName.toLowerCase();
        if (parentTag === MENUBAR_TAGNAME || parentTag === SIDENAV_TAGNAME)
            return true;
        if (parentTag === MENUSECTION_TAGNAME &&
            (grandParentTag === MENUBAR_TAGNAME || grandParentTag === SIDENAV_TAGNAME)) {
            return true;
        }
        return false;
    }
    isNestedMenuItem(element) {
        return !!element.closest(MENUPOPOVER_TAGNAME) && element.role === ROLE.MENUITEM;
    }
    async closeAllMenuPopovers() {
        const popovers = [];
        while (popoverStack.peek()) {
            const popover = popoverStack.pop();
            if (popover) {
                popover.hide();
                popovers.push(popover);
            }
        }
        await Promise.all(popovers.map(popover => popover.updateComplete));
    }
    async crossMenubarNavigationOnLeft(element) {
        const isMenuItem = element.role === ROLE.MENUITEM;
        if (isMenuItem) {
            const parentPopover = element.closest(MENUPOPOVER_TAGNAME);
            const triggerId = parentPopover === null || parentPopover === void 0 ? void 0 : parentPopover.getAttribute('triggerid');
            const triggerMenuItem = this.menuItems.find(item => item.getAttribute('id') === triggerId);
            if (triggerMenuItem) {
                if (this.isTopLevelMenuItem(triggerMenuItem)) {
                    parentPopover === null || parentPopover === void 0 ? void 0 : parentPopover.hide();
                }
                await (parentPopover === null || parentPopover === void 0 ? void 0 : parentPopover.updateComplete);
                const parentMenuItemIndex = this.getCurrentIndex(triggerMenuItem);
                this.navigateToMenuItem(parentMenuItemIndex, 'prev', true);
            }
        }
    }
    async crossMenubarNavigationOnRight(element) {
        if (this.isTopLevelMenuItem(element) && this.getSubmenu(element.id) && !element.hasAttribute('soft-disabled')) {
            this.showSubmenu(element.id);
        }
        else if (this.isNestedMenuItem(element) && !this.getSubmenu(element.id)) {
            await this.closeAllMenuPopovers();
            const parentIndex = this.getParentMenuItemIndex(element);
            if (parentIndex >= 0)
                this.navigateToMenuItem(parentIndex, 'next', true);
        }
    }
    getParentMenuItemIndex(element) {
        let parent = element.parentElement;
        while (parent) {
            if (parent.tagName.toLowerCase() === MENUPOPOVER_TAGNAME) {
                const triggerId = parent.getAttribute('triggerid');
                const triggerElement = this.menuItems.find(item => item.getAttribute('id') === triggerId);
                if (triggerElement) {
                    if (this.isTopLevelMenuItem(triggerElement)) {
                        return this.menuItems.findIndex(item => item === triggerElement);
                    }
                    return this.getParentMenuItemIndex(triggerElement);
                }
            }
            parent = parent.parentElement;
        }
        return -1;
    }
    async handleKeyDown(event) {
        const currentIndex = this.getCurrentIndex(event.target);
        const key = this.getKeyWithDirectionFix(event.key);
        switch (key) {
            case KEYS.HOME:
                this.updateTabIndexAndFocus(this.menuItems, currentIndex, 0);
                break;
            case KEYS.END:
                this.updateTabIndexAndFocus(this.menuItems, currentIndex, this.menuItems.length - 1);
                break;
            case KEYS.ARROW_LEFT: {
                const element = currentIndex >= 0 ? this.menuItems[currentIndex] : event.target;
                await this.crossMenubarNavigationOnLeft(element);
                break;
            }
            case KEYS.ARROW_RIGHT: {
                const element = currentIndex >= 0 ? this.menuItems[currentIndex] : event.target;
                await this.crossMenubarNavigationOnRight(element);
                break;
            }
            case KEYS.ARROW_UP: {
                this.navigateToMenuItem(currentIndex, 'prev');
                event.preventDefault();
                break;
            }
            case KEYS.ARROW_DOWN: {
                this.navigateToMenuItem(currentIndex, 'next');
                event.preventDefault();
                break;
            }
            default:
                break;
        }
    }
    render() {
        return html `<slot role="${ROLE.GROUP}"></slot>`;
    }
}
MenuBar.styles = [...Component.styles, ...styles];
__decorate([
    queryAssignedElements({ selector: 'mdc-menusection', flatten: true }),
    __metadata("design:type", Array)
], MenuBar.prototype, "menusections", void 0);
export default MenuBar;
