// @momentum-design/components/dist/components/formfieldwrapper/formfieldwrapper.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { v4 as uuidv4 } from 'uuid';
import { ifDefined } from 'lit/directives/if-defined.js';
import { Component } from '../../models';
import { DisabledMixin } from '../../utils/mixins/DisabledMixin';
import { BUTTON_VARIANTS } from '../button/button.constants';
import { DEFAULTS, MDC_TEXT_OPTIONS } from './formfieldwrapper.constants';
import styles from './formfieldwrapper.styles';
import { getHelperIcon } from './formfieldwrapper.utils';
/**
 * formfieldwrapper is a component that contains the label and helper/validation text
 *  that can be configured in various ways to suit different use cases (most of the input related components).
 * It is used as an internal component and is not intended to be used directly by consumers.
 *
 * @tagname mdc-formfieldwrapper
 *
 * @dependency mdc-text
 * @dependency mdc-icon
 * @dependency mdc-button
 * @dependency mdc-toggletip
 *
 *
 */
class FormfieldWrapper extends DisabledMixin(Component) {
    constructor() {
        super(...arguments);
        /**
        * Indicates the unique identifier for the native input element.
        * Required for acccessibility.
        * @internal
        */
        this.inputId = `mdc-el-${uuidv4()}`;
        /**
         * The required attribute to indicate that the input field is required.
         * It is used to append a required indicator (*) to the label.
         * @default false
         */
        this.required = false;
        /**
         * The type of help text. It can be 'default', 'error', 'warning', 'success', 'priority'.
         */
        this.helpTextType = DEFAULTS.VALIDATION;
        /**
         * The placement of the toggletip that is displayed when the info icon is hovered.
         * @default 'top'
         */
        this.toggletipPlacement = DEFAULTS.TOGGLETIP_PLACEMENT;
        /** @internal */
        this.shouldRenderLabel = true;
    }
    /**
     * creates the label element when the label property is set.
     * id is used to link the label with the input field.
     * @returns void
     */
    renderLabelElement() {
        if (!this.label) {
            return nothing;
        }
        return this.shouldRenderLabel
            ? html `<label for="${this.inputId}" id="${DEFAULTS.HEADING_ID}" class="mdc-label" part="label">${this.label}</label>`
            : html ` <mdc-text
          id="${DEFAULTS.HEADING_ID}"
          tagname="${MDC_TEXT_OPTIONS.TAGNAME}"
          type="${MDC_TEXT_OPTIONS.HEADER_TYPE}"
          part="label"
          >${this.label}</mdc-text
        >`;
    }
    /**
     * creates the helpertext icon based on the helpTextType for validation.
     * If the helpTextType is not set, it defaults to 'default' and it doesn't display any icon.
     * @returns void
     */
    renderHelpTextIcon() {
        if (!this.helpText) {
            return nothing;
        }
        const helperIcon = getHelperIcon(this.helpTextType || DEFAULTS.VALIDATION);
        if (helperIcon && !this.disabled) {
            return html `<mdc-icon part="helper-icon" size="1" length-unit="rem" name=${helperIcon}></mdc-icon>`;
        }
        return nothing;
    }
    /**
     * creates the helper text component when the helpertext value is set.
     * It is also used to display the validation message based on the helpTextType.
     * @returns void
     */
    renderHelpText() {
        if (!this.helpText) {
            return nothing;
        }
        return html `
      <mdc-text
        id="${DEFAULTS.HELPER_TEXT_ID}"
        part="help-text"
        tagname=${MDC_TEXT_OPTIONS.TAGNAME}
        type=${MDC_TEXT_OPTIONS.TYPE}
      >
        ${this.helpText}
      </mdc-text>
    `;
    }
    /**
     * renders the mdc-label-text container that contains the label and labelInfoToggleTip.
     * @returns void
     */
    renderLabel() {
        if (!this.label)
            return nothing;
        return html `<div class="mdc-label-text" part="label-text">
      <slot name="label">${this.renderLabelElement()}</slot>
      ${this.required ? html `<span part="required-indicator">*</span>` : nothing}
      ${this.toggletipText
            ? html ` <mdc-button
              part="info-icon-btn"
              prefix-icon="${DEFAULTS.INFO_ICON}"
              size="${DEFAULTS.ICON_SIZE}"
              variant="${BUTTON_VARIANTS.TERTIARY}"
              aria-label="${ifDefined(this.infoIconAriaLabel)}"
              id="info-icon-id"
            ></mdc-button>
            <mdc-toggletip
              part="label-toggletip"
              triggerid="info-icon-id"
              id="label-toggletip-id"
              placement="${this.toggletipPlacement}"
              show-arrow
              >${this.toggletipText}</mdc-toggletip
            >`
            : nothing}
    </div>`;
    }
    /**
     * renders the mdc-help-text container that contains the helpertext icon and helpertext.
     * @returns void
     */
    renderHelperText() {
        if (!this.helpText) {
            return nothing;
        }
        return html `<div class="mdc-help-text" part="help-text">
      <slot name="help-icon">${this.renderHelpTextIcon()}</slot>
      <slot name="help-text">${this.renderHelpText()}</slot>
    </div>`;
    }
}
FormfieldWrapper.styles = [...Component.styles, ...styles];
__decorate([
    property({ reflect: true, type: String }),
    __metadata("design:type", String)
], FormfieldWrapper.prototype, "label", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'required' }),
    __metadata("design:type", Object)
], FormfieldWrapper.prototype, "required", void 0);
__decorate([
    property({ reflect: true, attribute: 'help-text-type' }),
    __metadata("design:type", String)
], FormfieldWrapper.prototype, "helpTextType", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'help-text' }),
    __metadata("design:type", String)
], FormfieldWrapper.prototype, "helpText", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'toggletip-text' }),
    __metadata("design:type", String)
], FormfieldWrapper.prototype, "toggletipText", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'toggletip-placement' }),
    __metadata("design:type", String)
], FormfieldWrapper.prototype, "toggletipPlacement", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'info-icon-aria-label' }),
    __metadata("design:type", String)
], FormfieldWrapper.prototype, "infoIconAriaLabel", void 0);
export default FormfieldWrapper;
