// @momentum-design/components/dist/react/stepperitem/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/stepperitem';
import { TAG_NAME } from '../../components/stepperitem/stepperitem.constants';
/**
 * stepperitem component is used to represent a single step in a stepper component. It is used within a `mdc-stepper` component to indicate the current step in a process.
 * It can have different statuses such as `completed`, `current`, `incomplete`, `error-current`, and `error-incomplete`.
 * The component supports various visual styles and can be customized with labels, help text, and step numbers.
 *
 * This is an uncontrolled component, meaning it does not manage its own state. Instead, it relies on the consumer's to manage the state of each step.
 * Make sure to set `aria-current="step"` on the current stepper item. It is applicable only when status is `current` or `error-current`. This ensures accessibility for the stepper component. Only one stepper item should have this attribute at a time.
 *
 * Additionally, make use of `aria-label` to provide a descriptive detail about the stepper item, especially for screen readers. If this aria-label is not set, it would read out only the label text and doesn't provide enough context for the user.
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @tagname mdc-stepperitem
 *
 * @event click - (React: onClick) This event is dispatched when the stepperitem is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the stepperitem.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the stepperitem.
 *
 * @csspart status-container - The container for the status icon or step number.
 * @csspart label-container - The container for the label and help text.
 * @csspart help-text-container - The container for the help text and error icon when applicable.
 * @csspart status-icon - The icon representing the status of the stepper item.
 * @csspart step-number - The text representing the step number.
 * @csspart label - The text representing the label of the stepper item.
 * @csspart help-text - The text providing additional information about the stepper item.
 * @csspart help-icon - The icon representing an error in the stepper item.
 *
 * @cssproperty --mdc-stepperitem-status-container-background - The background color of the status container.
 * @cssproperty --mdc-stepperitem-status-container-border-color - The border color of the status container.
 * @cssproperty --mdc-stepperitem-label-color - The color of the label text.
 * @cssproperty --mdc-stepperitem-help-text-color - The color of the optional label text.
 * @cssproperty --mdc-stepperitem-label-container-background - The background color of the label container.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onKeyUp: 'keyup',
    },
    displayName: 'StepperItem',
});
export default reactWrapper;
