// @momentum-design/components/dist/components/radiogroup/radiogroup.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import FormfieldGroup from '../formfieldgroup';
import { TAG_NAME as RADIO_TAGNAME } from '../radio/radio.constants';
import { TAG_NAME as CARD_RADIO_TAGNAME } from '../cardradio/cardradio.constants';
const isRadio = new RegExp(`^(${RADIO_TAGNAME}|${CARD_RADIO_TAGNAME})$`, 'i');
/**
 * `mdc-radiogroup` - This is the wrapper component for radio buttons which are grouped together.
 * It can have a header text and a description. It enables users to select a single option from a set of options.
 * It is often used in forms, settings, and selection in lists. It automatically group the radio buttons inside it.
 *
 * @tagname mdc-radiogroup
 *
 */
class RadioGroup extends FormfieldGroup {
    constructor() {
        super();
        /**
         * Name of the radio group.
         * They are used to group elements in a form together.
         * @default ''
         */
        this.name = '';
        // This is used to set the role of the component as `radiogroup`.
        this.isRadio = true;
    }
    connectedCallback() {
        super.connectedCallback();
        // toggletipPlacement and toggletipStrategy are set to undefined as they are not used by radiogroup.
        this.toggletipPlacement = undefined;
        this.toggletipStrategy = undefined;
    }
    /**
     * Using name property to group the radio buttons together.
     * If the name is not set, it will be set to the name of the radio group.
     */
    firstUpdated() {
        var _a;
        Array.from(((_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelectorAll('slot')) || [])
            .flatMap(slot => slot.assignedElements({ flatten: true }))
            .filter(el => isRadio.test(el.tagName))
            .forEach(radio => {
            if (!radio.hasAttribute('name')) {
                radio.setAttribute('name', this.name);
            }
            radio.toggleAttribute('required', this.required);
        });
    }
}
RadioGroup.styles = [...FormfieldGroup.styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", Object)
], RadioGroup.prototype, "name", void 0);
export default RadioGroup;
