// @momentum-design/components/dist/components/list/list.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { Component } from '../../models';
import { ROLE } from '../../utils/roles';
import { ListNavigationMixin } from '../../utils/mixins/ListNavigationMixin';
import { TAG_NAME as LISTITEM_TAGNAME } from '../listitem/listitem.constants';
import { ElementStore } from '../../utils/controllers/ElementStore';
import { CaptureDestroyEventForChildElement } from '../../utils/mixins/lifecycle/CaptureDestroyEventForChildElement';
import { LIFE_CYCLE_EVENTS } from '../../utils/mixins/lifecycle/lifecycle.contants';
import { ACTIONS } from '../../utils/mixins/KeyToActionMixin';
import styles from './list.styles';
import { DEFAULTS } from './list.constants';
/**
 * mdc-list component is used to display a group of list items. It is used as a container to wrap other list items.
 *
 * To add a header to the list, use the `mdc-listheader` component and place it in the `list-header` slot.
 * `mdc-listitem` components can be placed in the default slot.
 *
 * @tagname mdc-list
 *
 * @slot default - This is a default/unnamed slot, where listitems can be placed.
 * @slot list-header - This slot is used to pass a header for the list, which can be a `mdc-listheader` component.
 *
 * @csspart container - The container slot around the list items
 */
class List extends ListNavigationMixin(CaptureDestroyEventForChildElement(Component)) {
    constructor() {
        super();
        /**
         * Whether to loop navigation when reaching the end of the list.
         * If 'true', pressing the down arrow on the last item will focus the first item,
         * and pressing the up arrow on the first item will focus the last item.
         * If 'false', navigation will stop at the first or last item.
         *
         * @default 'true'
         */
        this.loop = DEFAULTS.LOOP;
        /**
         * The index of the item that should receive focus when the list is first rendered.
         * If the index is out of bounds, the focused element will be clamped to the nearest valid index.
         *
         * @default 0
         */
        this.initialFocus = DEFAULTS.INITIAL_FOCUS;
        /**
         * The orientation of the list.
         * Controls the Flexbox direction and the direction of keyboard navigation:
         * - 'vertical': Up/Down arrow keys navigate between items
         * - 'horizontal': Left/Right arrow keys navigate between items
         *
         * @default 'vertical'
         */
        this.orientation = DEFAULTS.ORIENTATION;
        /** @internal */
        this.focusWithin = false;
        /** @internal */
        this.handleModifiedEvent = (event) => {
            const item = event.target;
            switch (event.detail.change) {
                case 'enabled':
                    this.itemsStore.add(item);
                    break;
                case 'disabled':
                    this.itemsStore.delete(item);
                    break;
                default:
                    break;
            }
        };
        this.addEventListener(LIFE_CYCLE_EVENTS.MODIFIED, this.handleModifiedEvent);
        this.addEventListener('keydown', this.handleKeyDown);
        this.addEventListener('focusin', this.handleFocusEvent);
        this.addEventListener('focusout', this.handleFocusEvent);
        // This must be initialized after the destroyed event listener
        // to keep the element in the itemStore in order to move the focus correctly
        this.itemsStore = new ElementStore(this, {
            isValidItem: this.isValidItem,
            onStoreUpdate: this.onElementStoreUpdate.bind(this),
        });
    }
    connectedCallback() {
        super.connectedCallback();
        this.role = ROLE.LIST;
    }
    /**
     * @internal
     */
    get navItems() {
        return this.itemsStore.items;
    }
    onElementStoreUpdate(item, changeType, index) {
        if (changeType === 'added') {
            // Update the tabIndex of the list items when a new item is added.
            item.tabIndex = -1;
            if (this.navItems.length === 0) {
                // If this is the first item added, set its tabIndex to 0 to make it focusable
                item.tabIndex = 0;
            }
        }
        else if (changeType === 'removed' && item.tabIndex === 0) {
            let newIndex = index + 1;
            if (newIndex >= this.navItems.length) {
                newIndex = index - 1;
            }
            this.resetTabIndexes(newIndex, this.focusWithin);
        }
    }
    /** @internal */
    handleKeyDown(event) {
        const action = this.getActionForKeyEvent(event);
        if (action === ACTIONS.SPACE && this.navItems.find(item => item === event.target)) {
            this.keyDownEventHandled();
            event.stopPropagation();
            event.preventDefault();
        }
    }
    /** @internal */
    handleFocusEvent(event) {
        // If previously focused element is being removed from the DOM, ignore the focusout event
        if (!(event.type === 'focusout' && event.relatedTarget === null)) {
            this.focusWithin = event.type === 'focusin';
        }
    }
    /** @internal */
    isValidItem(item) {
        return item.matches(`${LISTITEM_TAGNAME}:not([disabled])`);
    }
    render() {
        return html `
      <slot name="list-header"></slot>
      <!-- make the container slot role presentation to keep it ignored in a11y tree -->
      <slot part="container" role="presentation"></slot>
    `;
    }
}
List.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], List.prototype, "loop", void 0);
__decorate([
    property({ type: Number, reflect: true, attribute: 'initial-focus' }),
    __metadata("design:type", Number)
], List.prototype, "initialFocus", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], List.prototype, "orientation", void 0);
export default List;
