// @momentum-design/components/dist/components/listitem/listitem.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { nothing, html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { Component } from '../../models';
import { KEYS } from '../../utils/keys';
import { DisabledMixin } from '../../utils/mixins/DisabledMixin';
import { TabIndexMixin } from '../../utils/mixins/TabIndexMixin';
import { ROLE } from '../../utils/roles';
import { TYPE, VALID_TEXT_TAGS } from '../text/text.constants';
import { TAG_NAME as TOOLTIP_TAG_NAME } from '../tooltip/tooltip.constants';
import { DEFAULTS } from './listitem.constants';
import styles from './listitem.styles';
import { generateListItemId, generateTooltipId } from './listitem.utils';
import { ListItemEventManager } from './listitem.events';
/**
 * mdc-listitem component is used to display a label with different types of controls.
 * There can be three types of controls, a radio button, a checkbox on the
 * leading side and a toggle on the trailing side.
 * The list item can contain an avatar on the leading side and a badge on the trailing side.
 * Additionally, the list item can contain a side header and a subline text.
 *
 * The leading and trailing slots can be used to display controls and text. <br/>
 * Based on the leading/trailing slot, the position of the controls and text can be adjusted. <br/>
 * Please use mdc-list as a parent element even when there is only listitem for a11y purpose.
 *
 * By providing the tooltip-text attribute, a tooltip will be displayed on hover of the listitem.
 * The placement of the tooltip can be adjusted using the tooltip-placement attribute.
 * This will be helpful when the listitem text is truncated or
 * when you want to display additional information about the listitem.
 *
 * @tagname mdc-listitem
 *
 * @dependency mdc-text
 * @dependency mdc-tooltip
 *
 * @slot leading-controls - slot for list item controls to appear of leading end.
 * @slot leading-text-primary-label - slot for list item primary label.
 * @slot leading-text-secondary-label - slot for list item secondary label.
 * @slot leading-text-tertiary-label - slot for list item tertiary label.
 * @slot trailing-text-side-header - slot for list item side header text.
 * @slot trailing-text-subline - slot for list item subline text.
 * @slot trailing-controls - slot for list item controls to appear of trailing end.
 * @slot content - content slot can be used to override the content completely. Be aware that
 * this will override the default content of the list item.
 *
 * @cssproperty --mdc-listitem-default-background-color - Allows customization of the default background color.
 * @cssproperty --mdc-listitem-background-color-hover - Allows customization of the background color on hover.
 * @cssproperty --mdc-listitem-background-color-active - Allows customization of the background color when pressed.
 * @cssproperty --mdc-listitem-primary-label-color
 *  - Allows customization of the primary label, side header and subline text slot color.
 * @cssproperty --mdc-listitem-secondary-label-color
 *  - Allows customization of the secondary and tertiary label text slot color.
 * @cssproperty --mdc-listitem-disabled-color - Allows customization of the disabled color.
 * @cssproperty --mdc-listitem-column-gap - Allows customization of column gap.
 * @cssproperty --mdc-listitem-padding-left-right - Allows customization of padding left and right.
 * @cssproperty --mdc-listitem-padding-top-bottom - Allows customization of padding top and bottom.
 * @cssproperty --mdc-listitem-cursor - Allows customization of the cursor.
 * @cssproperty --mdc-listitem-width - Allows customization of the width of the list item.
 * @cssproperty --mdc-listitem-height - Allows customization of the height of the list item.
 *
 * @event click - (React: onClick) This event is dispatched when the listitem is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the listitem.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the listitem.
 * @event focus - (React: onFocus) This event is dispatched when the listitem receives focus.
 * @event enabled - (React: onEnabled) This event is dispatched after the listitem is enabled
 * @event disabled - (React: onDisabled) This event is dispatched after the listitem is disabled
 * @event created - (React: onCreated) This event is dispatched after the listitem is created (added to the DOM)
 * @event destroyed - (React: onDestroyed) This event is dispatched after the listitem is destroyed (removed from the DOM)
 */
class ListItem extends DisabledMixin(TabIndexMixin(Component)) {
    constructor() {
        super();
        /**
         * The variant of the list item. It can be a pill, rectangle or a full-width.
         * @default 'full-width'
         */
        this.variant = DEFAULTS.VARIANT;
        /**
         * The tooltip placement of the list item. If the tooltip text is present,
         * then this tooltip placement will be applied.
         * @default 'top'
         */
        this.tooltipPlacement = DEFAULTS.TOOLTIP_PLACEMENT;
        this.addEventListener('keydown', this.handleKeyDown.bind(this));
        this.addEventListener('focusin', this.displayTooltipForLongText.bind(this));
        this.addEventListener('mouseenter', this.displayTooltipForLongText.bind(this));
        this.addEventListener('focusout', this.hideTooltipOnLeave.bind(this));
        this.addEventListener('mouseout', this.hideTooltipOnLeave.bind(this));
        this.addEventListener('click', this.handleClick.bind(this));
    }
    connectedCallback() {
        super.connectedCallback();
        this.role = this.role || ROLE.LISTITEM;
        // Add a unique id to the listitem if it does not have one.
        this.id = this.id || generateListItemId();
        ListItemEventManager.onCreatedListItem(this);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        ListItemEventManager.onDestroyedListItem(this);
    }
    /**
     * Fires the click event when the enter or space key is pressed down.
     * This behavior is similar to a button click and key interaction.
     * @param event - The keyboard event triggered when a key is pressed down.
     */
    handleKeyDown(event) {
        if (event.key === KEYS.ENTER || event.key === KEYS.SPACE) {
            this.triggerClickEvent();
            event.preventDefault();
        }
    }
    /**
     * Triggers a click event on the list item.
     */
    triggerClickEvent() {
        const clickEvent = new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
            view: window,
        });
        this.dispatchEvent(clickEvent);
    }
    /**
     * Handles the click event on the list item.
     * If the tooltip is open, it has to be closed first.
     */
    handleClick(event) {
        if (this.disabled) {
            // when disabled, prevent the click event from propagating
            // and from firing on the host (immediate)
            event.stopImmediatePropagation();
            event.preventDefault();
            return;
        }
        // If the tooltip is open, it has to be closed first.
        this.hideTooltipOnLeave();
    }
    /**
     * Display a tooltip for the listitem.
     * Create the tooltip programmatically after the nearest parent element.
     */
    displayTooltipForLongText() {
        var _a, _b;
        if (!this.tooltipText) {
            return;
        }
        // Remove any existing tooltip.
        this.hideTooltipOnLeave();
        // Create tooltip for the listitem element.
        const tooltip = document.createElement(TOOLTIP_TAG_NAME);
        tooltip.id = generateTooltipId();
        tooltip.textContent = this.tooltipText;
        tooltip.setAttribute('triggerid', this.id);
        tooltip.setAttribute('placement', this.tooltipPlacement);
        tooltip.setAttribute('visible', '');
        tooltip.setAttribute('show-arrow', '');
        // Set the slot attribute if the parent element has a slot.
        if ((_a = this.parentElement) === null || _a === void 0 ? void 0 : _a.hasAttribute('slot')) {
            tooltip.setAttribute('slot', this.parentElement.getAttribute('slot') || '');
        }
        // Attach the tooltip programmatically after the nearest parent element.
        (_b = this.parentElement) === null || _b === void 0 ? void 0 : _b.after(tooltip);
    }
    /**
     * Removes the dynamically created tooltip for long text label on focus or mouse leave.
     * This is triggered on focusout and mouseout events.
     */
    hideTooltipOnLeave() {
        const existingTooltip = document.querySelector(`${TOOLTIP_TAG_NAME}[triggerid="${this.id}"]`);
        if (existingTooltip) {
            existingTooltip.remove();
        }
    }
    /**
     * Generates a template for a text slot with the specified content.
     *
     * @param slotName - The name of the slot to be used.
     * @param type - The type of the text element.
     * @param content - The text content to be displayed within the slot.
     * @returns A TemplateResult containing a slot with an `mdc-text` element of type BODY_SMALL_REGULAR.
     */
    getText(slotName, type, content) {
        return html `
      <slot name="${slotName}">
        ${content
            ? html `<mdc-text part="${slotName}" type="${type}" tagname="${VALID_TEXT_TAGS.SPAN}">${content}</mdc-text>`
            : nothing}
      </slot>
    `;
    }
    /**
     * Disable or enable all slotted elements in the leading and trailing slots.
     * This is useful when the list item is disabled, to prevent the user from interacting with the controls.
     * @param disabled - Whether to disable or enable the controls.
     */
    disableSlottedChildren(disabled = false) {
        [...this.leadingControlsSlot, ...this.trailingControlsSlot].forEach(element => {
            if (disabled) {
                element.setAttribute('disabled', '');
                ListItemEventManager.onDisableListItem(this);
            }
            else {
                element.removeAttribute('disabled');
                ListItemEventManager.onEnableListItem(this);
            }
        });
        // Set the aria-disabled attribute to indicate that the list item is disabled.
        this.setAttribute('aria-disabled', `${disabled}`);
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('disabled')) {
            this.tabIndex = this.disabled ? -1 : 0;
            this.disableSlottedChildren(this.disabled);
        }
        if (changedProperties.has('softDisabled')) {
            this.disableSlottedChildren(this.softDisabled);
        }
    }
    /**
     * Renders the trailing controls slot.
     * @returns A template for the trailing controls slot.
     */
    renderTrailingControls() {
        return html `<slot
      name="trailing-controls"
      @click=${this.stopEventPropagation}
      @keyup=${this.stopEventPropagation}
      @keydown=${this.stopEventPropagation}
    ></slot>`;
    }
    /**
     * Renders the leading controls slot.
     * @returns A template for the leading controls slot.
     */
    renderLeadingControls() {
        return html `<slot
      name="leading-controls"
      @click=${this.stopEventPropagation}
      @keyup=${this.stopEventPropagation}
      @keydown=${this.stopEventPropagation}
    ></slot>`;
    }
    /**
     * Stops the click event from propagating to parent elements. In case of keyboard events,
     * it stops the propagation for Enter and Space keys.
     * This is useful when the list item contains controls that
     * should not trigger the click event on the list item itself.
     * @param event - The mouse event triggered when a click occurs.
     */
    stopEventPropagation(event) {
        if ((event instanceof KeyboardEvent && (event.key === KEYS.ENTER || event.key === KEYS.SPACE)) ||
            event instanceof MouseEvent) {
            event.stopPropagation();
        }
    }
    render() {
        return html `
      <slot name="content">
        <div part="leading">
          ${this.renderLeadingControls()}
          <div part="leading-text">
            ${this.getText('leading-text-primary-label', TYPE.BODY_MIDSIZE_REGULAR, this.label)}
            ${this.getText('leading-text-secondary-label', TYPE.BODY_SMALL_REGULAR, this.secondaryLabel)}
            ${this.getText('leading-text-tertiary-label', TYPE.BODY_SMALL_REGULAR, this.tertiaryLabel)}
          </div>
        </div>
        <div part="trailing">
          <div part="trailing-text">
            ${this.getText('trailing-text-side-header', TYPE.BODY_MIDSIZE_REGULAR, this.sideHeaderText)}
            ${this.getText('trailing-text-subline', TYPE.BODY_SMALL_REGULAR, this.sublineText)}
          </div>
          ${this.renderTrailingControls()}
        </div>
      </slot>
    `;
    }
}
ListItem.styles = [...Component.styles, ...styles];
__decorate([
    queryAssignedElements({ slot: 'leading-controls' }),
    __metadata("design:type", Array)
], ListItem.prototype, "leadingControlsSlot", void 0);
__decorate([
    queryAssignedElements({ slot: 'trailing-controls' }),
    __metadata("design:type", Array)
], ListItem.prototype, "trailingControlsSlot", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], ListItem.prototype, "variant", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], ListItem.prototype, "label", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'secondary-label' }),
    __metadata("design:type", String)
], ListItem.prototype, "secondaryLabel", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'tertiary-label' }),
    __metadata("design:type", String)
], ListItem.prototype, "tertiaryLabel", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'side-header-text' }),
    __metadata("design:type", String)
], ListItem.prototype, "sideHeaderText", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'subline-text' }),
    __metadata("design:type", String)
], ListItem.prototype, "sublineText", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'soft-disabled' }),
    __metadata("design:type", Boolean)
], ListItem.prototype, "softDisabled", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'tooltip-text' }),
    __metadata("design:type", String)
], ListItem.prototype, "tooltipText", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'tooltip-placement' }),
    __metadata("design:type", String)
], ListItem.prototype, "tooltipPlacement", void 0);
export default ListItem;
