// @momentum-design/components/dist/components/accordiongroup/accordiongroup.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { TAG_NAME as ACCORDION_TAGNAME } from '../accordion/accordion.constants';
import { TAG_NAME as ACCORDIONBUTTON_TAGNAME } from '../accordionbutton/accordionbutton.constants';
import { Component } from '../../models';
import { DEFAULTS } from './accordiongroup.constants';
import styles from './accordiongroup.styles';
/**
 * An accordion group is a vertically stacked set of interactive headings that each contain a header and body content.
 * Each heading of the accordion acts as a control that enable users to expand or hide their associated body sections of content.
 * Accordions are commonly used to reduce the need to scroll when presenting multiple sections of content on a single page.
 *
 * - Default Slot: The accordion group component only accepts, `accordion` and `accordionbutton` components as the children, rest are ignored.
 *
 * There are three types of variants:
 * - Stacked - Each accordion will have a gap of 1.5rem (24px).
 * - Borderless - Each accordion will not have any border and the group will also not have any border.
 * - Contained - Each accordion will have no gap in between them and the border of the entire accordiongroup will be continuous.
 *
 * There are two types of sizes:
 * - Small: Small size has a padding of 1rem (16px) for both heading and body sections.
 * - Large: Large size has a padding of 1.5rem (24px) for both heading and body sections.
 *
 * The variant and size will be applied to all accordions inside this accordion group.
 * To show/expand more than one accordion at any given time, then set `allow-multiple` to `true`. By default, it's `false`.
 *
 * If you don't need any controls on your accordion heading, then it's advised to use `accordionbutton` component.
 *
 * If the first accordion of the accordion group is expanded by default, then the screen reader might loose focus when toggling the visibilty of the first accordion.
 *
 * @tagname mdc-accordiongroup
 *
 * @slot default - The default slot can contain the `accordion` or `accordionbutton` components.
 *
 * @cssproperty --mdc-accordiongroup-border-color - The border color of the entire accordiongroup
 */
class AccordionGroup extends Component {
    constructor() {
        super();
        /**
         * The size of the accordion item.
         * @default 'small'
         */
        this.size = DEFAULTS.SIZE;
        /**
         * The variant of the accordion item. Based on the variant, the styling of the accordion gets changed.
         * @default 'stacked'
         */
        this.variant = DEFAULTS.VARIANT;
        /**
         * If true, multiple accordion items can be visible at the same time.
         * @default false
         */
        this.allowMultiple = false;
        this.addEventListener('shown', this.handleAccordionExpanded);
    }
    /**
     * Handles the 'shown' event for accordion items.
     * If `allowMultiple` is false, ensures that only one accordion item
     * remains expanded by collapsing all other expanded items when a new item is expanded.
     *
     * @param event - The event object from the 'shown' event.
     */
    handleAccordionExpanded(event) {
        if (this.allowMultiple)
            return;
        [...this.accordionItems, ...this.accordionButtonItems].forEach(accordionItem => {
            if (accordionItem !== event.target && accordionItem.hasAttribute('expanded')) {
                accordionItem.toggleAttribute('expanded');
            }
        });
    }
    /**
     * Sets the given attribute on all child accordion or accordionbutton components.
     *
     * @param attributeName - The name of the attribute to set.
     * @param attributeValue - The value to set the attribute to.
     */
    setChildrenAccordionAttributes(attributeName, attributeValue) {
        [...this.accordionItems].forEach(accordion => {
            accordion.setAttribute(attributeName, attributeValue);
        });
        [...this.accordionButtonItems].forEach(accordion => {
            accordion.setAttribute(attributeName, attributeValue);
        });
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('size')) {
            if (!this.size)
                this.size = DEFAULTS.SIZE;
            this.setChildrenAccordionAttributes('size', this.size);
        }
        if (changedProperties.has('variant')) {
            if (!this.variant)
                this.variant = DEFAULTS.VARIANT;
            this.setChildrenAccordionAttributes('variant', this.variant);
        }
    }
    render() {
        return html ` <slot></slot> `;
    }
}
AccordionGroup.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], AccordionGroup.prototype, "size", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], AccordionGroup.prototype, "variant", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'allow-multiple' }),
    __metadata("design:type", Object)
], AccordionGroup.prototype, "allowMultiple", void 0);
__decorate([
    queryAssignedElements({ selector: ACCORDION_TAGNAME }),
    __metadata("design:type", Array)
], AccordionGroup.prototype, "accordionItems", void 0);
__decorate([
    queryAssignedElements({ selector: ACCORDIONBUTTON_TAGNAME }),
    __metadata("design:type", Array)
], AccordionGroup.prototype, "accordionButtonItems", void 0);
export default AccordionGroup;
