// @momentum-design/components/dist/components/accordiongroup/accordiongroup.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { TAG_NAME as ACCORDION_TAGNAME } from '../accordion/accordion.constants';
import { TAG_NAME as ACCORDIONBUTTON_TAGNAME } from '../accordionbutton/accordionbutton.constants';
import { Component } from '../../models';
import { DEFAULTS } from './accordiongroup.constants';
import styles from './accordiongroup.styles';
/**
 * An accordion group is a container that manages multiple accordion or accordionbutton components as a unified set.
 * It controls the visual styling, spacing, and expansion behavior of all child accordions.
 *
 * The group applies consistent `variant` and `size` attributes to all children automatically.
 * By default, expanding one accordion collapses others (`allow-multiple` is false). Set `allow-multiple` to true to allow multiple expanded items.
 *
 * ## Accepted children
 * - Use `mdc-accordionbutton` children for simple clickable headers.
 * - Use `mdc-accordion` children when you need additional controls (chips, badges, icons) in the headers.
 * - Other elements in the slot are ignored.
 *
 * ## Accessibility
 * - Note: Screen readers may lose focus when toggling if the first accordion is expanded by default.
 *
 * @tagname mdc-accordiongroup
 *
 * @slot default - The default slot can contain the `accordion` or `accordionbutton` components.
 *
 * @cssproperty --mdc-accordiongroup-items-border-color - The border color applied to all child accordion items and their separators.
 */
class AccordionGroup extends Component {
    constructor() {
        super();
        /**
         * Controls the spacing and padding for all accordion items in the group.
         * This size is applied to all child accordions.
         * - Small provides 1rem (16px) padding, large provides 1.5rem (24px) padding.
         * @default 'small'
         */
        this.size = DEFAULTS.SIZE;
        /**
         * Visual style variant applied to all accordions in the group.
         * - Stacked: accordions have 1.5rem gap between them.
         * - Borderless: removes all borders from accordions and group.
         * - Contained: accordions have continuous borders with no gaps.
         * @default 'stacked'
         */
        this.variant = DEFAULTS.VARIANT;
        /**
         * Controls whether multiple accordions can be expanded simultaneously.
         * - When false, expanding one accordion automatically collapses others.
         * - When true, multiple accordions can remain open at the same time.
         * @default false
         */
        this.allowMultiple = false;
        this.addEventListener('shown', this.handleAccordionExpanded);
    }
    /**
     * Handles the 'shown' event for accordion items.
     * If `allowMultiple` is false, ensures that only one accordion item
     * remains expanded by collapsing all other expanded items when a new item is expanded.
     *
     * @param event - The event object from the 'shown' event.
     */
    handleAccordionExpanded(event) {
        if (this.allowMultiple)
            return;
        [...this.accordionItems, ...this.accordionButtonItems].forEach(accordionItem => {
            if (accordionItem !== event.target && accordionItem.hasAttribute('expanded')) {
                accordionItem.toggleAttribute('expanded');
            }
        });
    }
    /**
     * Sets the given attribute on all child accordion or accordionbutton components.
     *
     * @param attributeName - The name of the attribute to set.
     * @param attributeValue - The value to set the attribute to.
     */
    setChildrenAccordionAttributes(attributeName, attributeValue) {
        [...this.accordionItems].forEach(accordion => {
            accordion.setAttribute(attributeName, attributeValue);
        });
        [...this.accordionButtonItems].forEach(accordion => {
            accordion.setAttribute(attributeName, attributeValue);
        });
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('size')) {
            if (!this.size)
                this.size = DEFAULTS.SIZE;
            this.setChildrenAccordionAttributes('size', this.size);
        }
        if (changedProperties.has('variant')) {
            if (!this.variant)
                this.variant = DEFAULTS.VARIANT;
            this.setChildrenAccordionAttributes('variant', this.variant);
        }
    }
    render() {
        return html ` <slot></slot> `;
    }
}
AccordionGroup.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], AccordionGroup.prototype, "size", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], AccordionGroup.prototype, "variant", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'allow-multiple' }),
    __metadata("design:type", Object)
], AccordionGroup.prototype, "allowMultiple", void 0);
__decorate([
    queryAssignedElements({ selector: ACCORDION_TAGNAME }),
    __metadata("design:type", Array)
], AccordionGroup.prototype, "accordionItems", void 0);
__decorate([
    queryAssignedElements({ selector: ACCORDIONBUTTON_TAGNAME }),
    __metadata("design:type", Array)
], AccordionGroup.prototype, "accordionButtonItems", void 0);
export default AccordionGroup;
