// @momentum-design/components/dist/react/radio/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/radio';
import { TAG_NAME } from '../../components/radio/radio.constants';
/**
 * The Radio component allows users to select a single option from a group of mutually exclusive choices.
 * Unlike checkboxes which allow multiple selections, radio buttons ensure only one option can be selected
 * at a time within the same group. These are commonly used in forms, surveys, and settings where users
 * need to make a single selection from multiple options.
 *
 * To create a group of radio buttons, use the `mdc-radiogroup` component or ensure all radio buttons
 * share the same `name` attribute.
 *
 * ## Validation
 *
 * Radio component support native form validation. But it does not have default validation message.
 * Also, `required` attribute does not render indicator (red asterisk) for the radio component.
 *
 * The recommended way to show validation message for radio groups is to wrap the `mdc-radio` with `mdc-radiogroup`
 * and set the `help-text` of the `mdc-radiogroup` based on its validation state.
 *
 * Alternatively you can also set the `validation-message` attribute of the `mdc-radio`. This message will appear
 * in a native tooltip when the radio is checked and invalid.
 *
 * ## Accessibility
 *
 * - Provide clear labels that describe each option
 * - Use `data-aria-label` when a visual label is not present
 * - Keyboard navigation: Arrow keys to move between options, Space to select, Tab to navigate groups, Enter to submit form
 * - Group related radio buttons using the same `name` attribute or `mdc-radiogroup` component
 *
 * ## Styling
 *
 * Use the `radio-indicator` part to apply custom styles to the radio visual element.
 * This part exposes the underlying [StaticRadio](?path=/docs/components-decorator-staticradio--docs) component for advanced styling.
 *
 * The `indicator` slot allows replacing the default radio circle with a custom element.
 * When a custom indicator is slotted, the component automatically adds the `mdc-focus-ring`
 * class to the host element. This shifts the focus ring from the default static radio to the
 * entire host element, ensuring keyboard focus remains visible.
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-staticradio
 * @dependency mdc-toggletip
 *
 * @tagname mdc-radio
 *
 * @event input - (React: onInput) Event that gets dispatched when the radio state changes (before the change event).
 * @event change - (React: onChange) Event that gets dispatched when the radio state changes (after the input event).
 * @event focus - (React: onFocus) Event that gets dispatched when the radio receives focus.
 *
 * @csspart label - The label element.
 * @csspart label-text - The container for the label and required indicator elements.
 * @csspart radio-indicator - The staticradio that provides the visual radio appearance.
 *
 * @slot indicator - Slot for the radio indicator element. If not provided, a default styled radio will be rendered.
 * @slot label - Slot for the label of the radio.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onInput: 'input',
        onChange: 'change',
        onFocus: 'focus',
    },
    displayName: 'Radio',
});
export default reactWrapper;
