// @momentum-design/components/dist/react/buttonlink/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/buttonlink';
import { TAG_NAME } from '../../components/buttonlink/buttonlink.constants';
/**
 * `mdc-buttonlink` combines the functional behavior of `mdc-linksimple` with the visual and structural
 * features of `mdc-button`. This includes support for variants, sizing, and optional
 * prefix and postfix icons via slots.
 *
 * ### Features:
 * - Behaves like an link while visually resembling a button.
 * - Supports slots for `prefix-icon` and `postfix-icon`.
 * - Customizable size, color, and variant through attributes.
 * - Inherits accessibility and keyboard interaction support from `mdc-linksimple`.
 *
 * @dependency mdc-icon
 *
 * @tagname mdc-buttonlink
 *
 * @event click - (React: onClick) Fired when the user activates the buttonLink using a mouse or assistive technology.
 * @event keydown - (React: onKeyDown) Fired when the user presses a key while the buttonLink has focus.
 * @event focus - (React: onFocus) Fired when the buttonLink receives keyboard or mouse focus.
 * @event blur - (React: onBlur) Fired when the buttonLink loses keyboard or mouse focus.
 *
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onFocus: 'focus',
        onBlur: 'blur',
    },
    displayName: 'ButtonLink',
});
export default reactWrapper;
