// @momentum-design/components/dist/react/buttonlink/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/buttonlink';
import { TAG_NAME } from '../../components/buttonlink/buttonlink.constants';
/**
 * `mdc-buttonlink` combines the functional behavior of `mdc-linksimple` with the visual and structural
 * features of `mdc-button`. This includes support for variants, sizing, and optional
 * prefix and postfix icons.
 *
 * ### Features:
 * - Behaves like a link while visually resembling a button.
 * - Customizable size, color, and variant through attributes.
 * - Supports prefix and postfix icons.
 * - Inherits accessibility and keyboard interaction support from `mdc-linksimple`.
 *
 * @dependency mdc-icon
 *
 * @tagname mdc-buttonlink
 *
 * @slot default - The default slot for buttonlink text content.
 *
 * @event click - (React: onClick) Fired when the user activates the buttonLink using a mouse or assistive technology.
 * @event keydown - (React: onKeyDown) Fired when the user presses a key while the buttonLink has focus.
 * @event focus - (React: onFocus) Fired when the buttonLink receives keyboard or mouse focus.
 * @event blur - (React: onBlur) Fired when the buttonLink loses keyboard or mouse focus.
 *
 * @csspart anchor - The anchor element that wraps the buttonlink content.
 * @csspart prefix-icon - The prefix icon element.
 * @csspart button-text - The slot containing the buttonlink text.
 * @csspart postfix-icon - The postfix icon element.
 *
 * @cssproperty --mdc-button-height - Height of the buttonlink
 * @cssproperty --mdc-button-background - Background color of the buttonlink
 * @cssproperty --mdc-button-border-color - Border color of the buttonlink
 * @cssproperty --mdc-button-text-color - Text color of the buttonlink
 * @cssproperty --mdc-button-line-height - Line height of the buttonlink text
 * @cssproperty --mdc-button-prefix-icon-size - Size of the prefix icon
 * @cssproperty --mdc-button-postfix-icon-size - Size of the postfix icon
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onFocus: 'focus',
        onBlur: 'blur',
    },
    displayName: 'ButtonLink',
});
export default reactWrapper;
