// @momentum-design/components/dist/utils/virtualIndexArray.js
/**
 * A wrapper around an array that applies a "virtual" index to the index when accessing items.
 * This is useful when the array is just a view of the real data and the real data has gaps in the indexes.
 *
 * @example
 * ```ts
 * const originalArray = [{realIndex: 2, value: 10}, {realIndex: 3, value: 20}];
 * const offsetArray = new OffsetArray(originalArray, (item) => item.realIndex);
 *
 * console.log(offsetArray.at(0)); // Output: undefined
 * console.log(offsetArray.at(2).value); // Output: 10
 * ```
 */
export class VirtualIndexArray {
    constructor(items, getIndex, length) {
        this.getIndex = () => 0;
        this.items = items;
        this.getIndex = getIndex;
        this.lengthFn = length;
    }
    get length() {
        return this.lengthFn();
    }
    at(index) {
        return this.items.find(item => this.getIndex(item) === index);
    }
    map(cb, thisArg) {
        return this.items.map((value, _index, array) => cb.call(thisArg, value, this.getIndex(value), array));
    }
    forEach(cb, thisArg) {
        this.items.forEach((value, _index, array) => {
            cb.call(thisArg, value, this.getIndex(value), array);
        });
    }
    findIndex(predicate, thisArg) {
        const item = this.items.find((value, _index, obj) => predicate.call(thisArg, value, this.getIndex(value), obj));
        return !item ? -1 : this.getIndex(item);
    }
    find(predicate, thisArg) {
        return this.items.find((value, _index, obj) => predicate.call(thisArg, value, this.getIndex(value), obj));
    }
}
