// @momentum-design/components/dist/components/filterchip/filterchip.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import Chip from '../chip/chip.component';
import styles from './filterchip.styles';
import { DEFAULTS } from './filterchip.constants';
/**
 * mdc-filterchip component is an interactive chip that consumers can use to select or deselect.
 * They can be found with lists or tables as quick filters.
 *
 * This component is built on top of the mdc-chip component.
 *
 * @tagname mdc-filterchip
 *
 */
class FilterChip extends Chip {
    constructor() {
        super(...arguments);
        /**
         * The selected state of the filterchip.
         *
         * @default false
         */
        this.selected = false;
    }
    connectedCallback() {
        super.connectedCallback();
        this.color = DEFAULTS.COLOR;
    }
    /**
     * Sets the selected state of the filterchip along with the icon.
     * aria-pressed attribute is also set to true or false.
     * @param selected - The selected state of the filterchip.
     */
    setSelected(selected) {
        if (selected) {
            this.iconName = DEFAULTS.ICON_NAME;
            this.setAttribute('aria-pressed', 'true');
        }
        else {
            this.iconName = undefined;
            this.setAttribute('aria-pressed', 'false');
        }
    }
    executeAction() {
        this.selected = !this.selected;
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('selected')) {
            this.setSelected(this.selected);
        }
    }
}
FilterChip.styles = [...Chip.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], FilterChip.prototype, "selected", void 0);
export default FilterChip;
