// @momentum-design/components/dist/react/verticaltablist/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/verticaltablist';
import { TAG_NAME } from '../../components/verticaltablist/verticaltablist.constants';
/**
 * `mdc-verticaltablist` is a vertical tab list component that extends the List component.
 * It organizes tabs vertically and follows the same WCAG standards as the horizontal tablist.
 *
 * Children of the vertical tab list are `mdc-tab` elements, sent to the default slot.
 * Currently only "line" variant of the tab is supported, setting other variants doesn't have any effect.
 * The variant is automatically set on the tab when it is added to the vertical tab list, so there is no need to set it manually.
 *
 * The tabs can be navigated using the up/down arrow keys, and selected by clicking,
 * or pressing the Enter and Space keys.
 *
 * **Implicit accessibility rules**
 *
 * - The element that serves as the container for the set of tabs has role `tablist`.
 * - Each element that serves as a tab has role `tab` and is contained within the element with role `tablist` (Tab components passed in to default slot).
 * - The active tab element has the state `aria-selected` set to `true`
 *   and all other tab elements have it set to `false`.
 *
 * **Accessibility notes for consuming (have to be explicitly set when you consume the component)**
 *
 * - Each element that contains the `content panel` for a `tab` has role `tabpanel`.
 * - The `tablist` element needs to have a label provided by `data-aria-label`.
 * - Each element with role `tab` has the property `aria-controls`
 *   that should refer to its associated `tabpanel` element.
 * - The vertical orientation is conveyed through the aria-orientation property on the `tablist` element, which is set to `vertical` by the component.
 * - Each element with role `tabpanel` has the property `aria-labelledby` referring to its associated `tab` element.
 *
 * @tagname mdc-verticaltablist
 *
 * @dependency mdc-tab
 *
 * @event change - (React: onChange) This event is dispatched when the tab is selected.
 * Event that fires when user changes the active tab.
 *     The function sent as the argument will receive the fired event
 *      and the new tab id can be fetched from event.detail.tabId.
 *
 *    `(event: CustomEvent) => handleTabChange(event.detail.tabId);`
 *
 * @slot Default slot for mdc-tab elements.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change',
    },
    displayName: 'Verticaltablist',
});
export default reactWrapper;
