// @momentum-design/components/dist/components/listbox/listbox.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, state } from 'lit/decorators.js';
import List from '../list';
import { TAG_NAME as OPTION_TAGNAME } from '../option/option.constants';
import { ROLE } from '../../utils/roles';
import { CaptureDestroyEventForChildElement } from '../../utils/mixins/lifecycle/CaptureDestroyEventForChildElement';
import { ListNavigationMixin } from '../../utils/mixins/ListNavigationMixin';
import { Component } from '../../models';
import { ElementStore } from '../../utils/controllers/ElementStore';
import styles from './listbox.styles';
/**
 * listbox component presents a list of options and allows a user to select one or more of them.
 *
 * When `multiple` is enabled, clicking/pressing Enter/Space on an option toggles its selection
 * instead of replacing the current selection.
 *
 * Notes:
 * - This is a standalone listbox component. Select has its own mdc-selectlistbox component.
 * - this component has name and value attributes and also emits change event,
 *   but it is not a form control (yet).
 *
 * @dependency mdc-list
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-option
 * @dependency mdc-optgroup
 *
 * @tagname mdc-listbox
 *
 * @cssproperty --mdc-listbox-max-height - max height of the listbox
 *
 * @slot default - This is a default/unnamed slot, where options and optgroups are placed
 *
 * @csspart container - The container of the listbox
 *
 * @event change - (React: onChange) This event is emitted when the selected item changed
 */
class ListBox extends ListNavigationMixin(CaptureDestroyEventForChildElement(Component)) {
    constructor() {
        super();
        /**
         * According to WCAG this is the expected behavior for listbox
         * https://www.w3.org/WAI/ARIA/apg/practices/listbox
         * @internal
         */
        this.loop = 'false';
        /**
         * The name attribute is used to identify the listbox
         */
        this.name = undefined;
        /**
         * The value attribute is used to represent the last selected option's value in the listbox.
         */
        this.value = undefined;
        /**
         * When true, multiple options can be selected.
         * @default false
         */
        this.multiple = false;
        /** @internal */
        this.itemsStore = new ElementStore(this, {
            isValidItem: this.isValidItem,
        });
        /** @internal */
        this.handleDestroyEvent = () => {
            this.handleNoSelection();
        };
        this.addEventListener('click', this.handleClick);
        this.addEventListener('modified', this.handleModifiedEvent);
        this.addEventListener('destroyed', this.handleDestroyEvent);
    }
    connectedCallback() {
        super.connectedCallback();
        this.role = ROLE.LISTBOX;
    }
    /** @internal */
    handleModifiedEvent(event) {
        const item = event.target;
        switch (event.detail.change) {
            case 'enabled':
                this.itemsStore.add(item);
                break;
            case 'disabled':
                this.itemsStore.delete(item);
                break;
            case 'selected':
                if (this.multiple) {
                    this.syncValueToFirstSelected();
                }
                else {
                    this.setSelectedOption(item, false, false);
                }
                break;
            case 'unselected':
                if (this.multiple) {
                    this.syncValueToFirstSelected();
                }
                else {
                    this.handleNoSelection();
                }
                break;
            default:
                break;
        }
    }
    /**
     * If nothing is selected anymore, clear the value
     * @internal
     */
    handleNoSelection() {
        const selectedOption = this.getFirstSelectedOption();
        if (!selectedOption) {
            this.selectedOption = undefined;
            this.value = undefined;
        }
    }
    /** @internal */
    get navItems() {
        return this.itemsStore.items;
    }
    /** @internal */
    isValidItem(item) {
        return item.matches(`${OPTION_TAGNAME}:not([disabled])`);
    }
    /** @internal */
    handleClick(event) {
        const target = event.target;
        const option = target.closest(OPTION_TAGNAME);
        if (!option || !this.isValidItem(option))
            return;
        if (this.multiple) {
            this.toggleOptionSelection(option);
        }
        else {
            this.setSelectedOption(option);
        }
    }
    /**
     * Toggles the selection state of an option (for multiselect mode).
     * @internal
     */
    toggleOptionSelection(option) {
        if (option.disabled || option.softDisabled)
            return;
        const isCurrentlySelected = option.hasAttribute('selected');
        option.toggleAttribute('selected', !isCurrentlySelected);
        this.syncValueToFirstSelected();
        this.fireEvents();
    }
    /** @internal */
    getFirstSelectedOption() {
        return this.itemsStore.items.find(el => el.matches('[selected]'));
    }
    /**
     * Syncs value and selectedOption to reflect the first selected option (when multiple)
     * @internal
     */
    syncValueToFirstSelected() {
        const firstSelected = this.getFirstSelectedOption();
        this.value = firstSelected === null || firstSelected === void 0 ? void 0 : firstSelected.value;
        this.selectedOption = firstSelected !== null && firstSelected !== void 0 ? firstSelected : null;
    }
    /** @internal */
    getSelectedValues() {
        return this.itemsStore.items.reduce((acc, option) => {
            const { value } = option;
            if (option.hasAttribute('selected') && value !== undefined) {
                acc.push(value);
            }
            return acc;
        }, []);
    }
    /**
     * Override to focus the first selected option per W3C APG.
     * This applies to both single-select and multi-select listboxes.
     * @internal
     */
    setInitialFocus() {
        const firstSelected = this.getFirstSelectedOption();
        if (firstSelected) {
            const index = this.itemsStore.items.indexOf(firstSelected);
            if (index !== -1) {
                this.resetTabIndexAndSetFocus(index, undefined, false);
                return;
            }
        }
        super.setInitialFocus();
    }
    /**
     * Handles the updated lifecycle event.
     *
     * @param changedProperties - The properties that have changed since the last update.
     */
    updated(changedProperties) {
        if (changedProperties.has('multiple')) {
            if (this.multiple) {
                this.setAttribute('aria-multiselectable', 'true');
            }
            else {
                this.removeAttribute('aria-multiselectable');
            }
        }
        if (changedProperties.has('value') && !this.multiple) {
            const newSelectedOption = this.itemsStore.items.find(option => option.value === this.value);
            if (newSelectedOption) {
                this.setSelectedOption(newSelectedOption, false);
            }
            else {
                this.setSelectedOption(null, false);
            }
        }
    }
    /**
     * When an option is selected, this method updates local and options' DOM state and fires the change event.
     *
     * @param option - The option element in DOM which gets selected.
     * @param fireEvent - A boolean flag to indicate whether to fire the change event or not.
     * @param updateOptions - Whether update the other options or not
     */
    setSelectedOption(option, fireEvent = true, updateOptions = true) {
        if ((option === null || option === void 0 ? void 0 : option.disabled) || (option === null || option === void 0 ? void 0 : option.softDisabled))
            return;
        if (updateOptions) {
            this.updateSelectedInChildOptions(option);
        }
        // set the selected option in the component state
        this.selectedOption = option;
        this.value = option === null || option === void 0 ? void 0 : option.value;
        if (fireEvent) {
            this.fireEvents();
        }
    }
    /**
     * Update all options selection state in the DDM.
     *
     * @param selectedOption - The option which gets selected
     */
    updateSelectedInChildOptions(selectedOption) {
        this.itemsStore.items.forEach(option => option.removeAttribute('selected'));
        selectedOption === null || selectedOption === void 0 ? void 0 : selectedOption.toggleAttribute('selected', true);
    }
    /**
     * Dispatch change event when selection changes.
     */
    fireEvents() {
        this.dispatchEvent(new CustomEvent('change', {
            detail: {
                value: this.value,
                selectedValues: this.getSelectedValues(),
            },
            composed: true,
            bubbles: true,
        }));
    }
    render() {
        return html `<slot part="container"></slot>`;
    }
}
ListBox.styles = [...List.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", Object)
], ListBox.prototype, "name", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", Object)
], ListBox.prototype, "value", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], ListBox.prototype, "multiple", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], ListBox.prototype, "selectedOption", void 0);
export default ListBox;
