// @momentum-design/components/dist/components/tablist/tablist.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property, query, queryAssignedElements, state } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { Component } from '../../models';
import Tab from '../tab/tab.component';
import Button from '../button/button.component';
import { ROLE } from '../../utils/roles';
import styles from './tablist.styles';
import { ARROW_BUTTON_DIRECTION, KEYCODES } from './tablist.constants';
import { getFirstTab, getLastTab, getNextTab, getPreviousTab, findTab, getActiveTab } from './tablist.utils';
/**
 * Tab list organizes tabs into a container.
 *
 * Children of the tab list are `mdc-tab` elements, sent to the default slot.
 *
 * The tabs can be navigated using the left/right arrow keys, and selected by clicking,
 *  or pressing the Enter and Space keys.
 *
 * **Implicit accessibility rules**
 *
 * - The element that serves as the container for the set of tabs has role `tablist`.
 * - Each element that serves as a tab has role `tab` and is contained within the element with role `tablist`.
 * - The active tab element has the state `aria-selected` set to `true`
 *   and all other tab elements have it set to `false`.
 *
 *
 * **Accessibility notes for consuming (have to be explicitly set when you consume the component)**
 *
 * - Each element that contains the `content panel` for a `tab` has role `tabpanel`.
 * - The `tablist` element needs to have a label provided by `data-aria-label`.
 * - Each element with role `tab` has the property `aria-controls`
 *  that should refer to its associated `tabpanel` element.
 * - Each element with role `tabpanel` has the property `aria-labelledby` referring to its associated `tab` element.
 * - If a `tab` element has a popup menu, it needs to have the property `aria-haspopup` set to either `menu` or `true`.
 *
 * @tagname mdc-tablist
 *
 * @dependency mdc-tab
 * @dependency mdc-button
 *
 * @event change - (React: onChange) This event is dispatched when the tab is selected.
 * Event that fires when user changes the active tab.
 *     The function sent as the argument will receive the fired event
 *      and the new tab id can be fetched from event.detail.tabId.
 *
 *    `(event: CustomEvent) => handleTabChange(event.detail.tabId);`
 *
 * @slot Default slot for mdc-tab elements.
 *
 * @cssproperty --mdc-tablist-gap - Gap between tabs
 * @cssproperty --mdc-tablist-width - Width of the tablist
 * @cssproperty  --mdc-tablist-arrow-button-margin - Margin value for the arrow button
 */
class TabList extends Component {
    /**
     * @internal
     */
    isRtl() {
        var _a;
        return (((_a = document.querySelector('html')) === null || _a === void 0 ? void 0 : _a.getAttribute('dir')) === 'rtl' || window.getComputedStyle(this).direction === 'rtl');
    }
    constructor() {
        super();
        /**
         * Instead of using 'left' and 'right' for notation of the arrow buttons, we use 'forward' and 'backward'.
         * Since the elements of the tablist gets flipped in RTL layouts, the arrow buttons also flip.
         * For both RTL and LTR layouts, the forward arrow button moves the tabs in the forward direction.
         * i.e. the forward arrow button moves the tabs to the right in LTR layouts and to the left in RTL layouts.
         */
        /**
         * @internal
         */
        this.showForwardArrowButton = false;
        /**
         * @internal
         */
        this.showBackwardArrowButton = false;
        this.addEventListener('keydown', this.handleKeydown.bind(this));
        // Reason for assertion below: https://github.com/microsoft/TypeScript/issues/28357
        this.addEventListener('activechange', this.handleNestedTabActiveChange);
    }
    /**
     * Set the initial active tab after firstUpdated,
     * since this.tabs (=slot) are only available after
     * the first update
     *
     * Also set up the event listeners.
     */
    async firstUpdated() {
        var _a, _b, _c;
        if (!this.tabs) {
            return;
        }
        /**
         * If there are no tabs, skip the initialization of the active tab and the event listeners.
         * Then throw an error.
         */
        if (Array.isArray(this.tabs) && this.tabs.length === 0) {
            if (this.onerror) {
                this.onerror('The tablist component must have at least one child tab');
            }
            return;
        }
        const tabIds = this.tabs.map(tab => tab.tabId);
        if (new Set(tabIds).size !== this.tabs.length) {
            if (this.onerror) {
                this.onerror('The tabs inside the tab list must have unique tab ids');
            }
        }
        const resizeObserver = new ResizeObserver(async () => {
            await this.handleArrowButtonVisibility();
        });
        resizeObserver.observe(this);
        if (!this.activeTabId) {
            this.activeTabId = (_a = getFirstTab(this.tabs)) === null || _a === void 0 ? void 0 : _a.tabId;
        }
        (_b = this.tabsContainer) === null || _b === void 0 ? void 0 : _b.addEventListener('focusin', this.handleFocus.bind(this));
        (_c = this.tabsContainer) === null || _c === void 0 ? void 0 : _c.addEventListener('mousedown', this.handleMousedown.bind(this));
    }
    disconnectedCallback() {
        var _a, _b;
        super.disconnectedCallback();
        (_a = this.tabsContainer) === null || _a === void 0 ? void 0 : _a.removeEventListener('focusin', this.handleFocus);
        (_b = this.tabsContainer) === null || _b === void 0 ? void 0 : _b.removeEventListener('mousedown', this.handleMousedown);
    }
    /**
     * Observe the tablist element for changes in the activeTabId attribute.
     * Find the new tab with the new activeTabId.
     * If the new tab is not found, then do nothing.
     *
     * If the new tab exists:
     * Set the new tab as active.
     *
     * Fire the tab change event.
     *
     * @param changedProperties - Map of changed properties with old values.
     */
    async update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('activeTabId')) {
            if (!this.tabs || !this.activeTabId) {
                return;
            }
            const newTab = findTab(this.tabs, this.activeTabId);
            if (!(newTab instanceof Tab)) {
                return;
            }
            this.setActiveTab(newTab);
            /**
             * If the previous activeTabId was not undefined, focus the new tab.
             *
             * Otherwise, reset the tabindex of all tabs and set the new tabindex.
             */
            if (changedProperties.get('activeTabId')) {
                this.fireTabChangeEvent(newTab);
                await this.focusTab(newTab);
            }
            else {
                this.resetTabIndexAndSetNewTabIndex(newTab);
            }
        }
    }
    /**
     * Dispatch the change event.
     *
     * @internal
     *
     * @param newTab - the new tab that is active.
     */
    fireTabChangeEvent(newTab) {
        const event = new CustomEvent('change', {
            detail: { tabId: newTab.tabId },
        });
        this.dispatchEvent(event);
    }
    /**
     * When the tablist receives focus, then focus the active tab.
     *
     * @internal
     *
     * @param event - Focus event.
     */
    async handleFocus(event) {
        /**
         * If the element losing focus is a tab, do nothing.
         * This also covers the case when previous focus was on a tab that belongs to another tablist.
         */
        if (event.relatedTarget instanceof Tab) {
            return;
        }
        /**
         * If the element gaining focus is not a tab, do nothing.
         */
        if (!(event.target instanceof Tab)) {
            return;
        }
        const activeTab = getActiveTab(this.tabs || []);
        if (!(activeTab instanceof Tab)) {
            return;
        }
        await this.focusTab(activeTab);
    }
    /**
     * Prevent the mousedown event from triggering a focus event before the click event.
     *
     * @internal
     *
     * @param event - Mouse event.
     */
    handleMousedown(event) {
        if (!(event.target instanceof Tab)) {
            return;
        }
        event.preventDefault();
    }
    /**
     * Handle the tab active change event fired from the nested tab.
     *
     * @internal
     *
     * @param event - Custom Event fired from the nested tab.
     */
    async handleNestedTabActiveChange(event) {
        event.stopPropagation();
        const tab = event.target;
        if (!(tab instanceof Tab)) {
            return;
        }
        this.setActiveTab(tab);
        await this.focusTab(tab);
        this.activeTabId = tab.tabId;
    }
    /**
     * Resets all tabs' tabindex to -1 and sets the tabindex of the
     * given tab to 0.
     * This is used when navigating the tabs via keyboard.
     *
     * @internal
     *
     * @param tabId - The id of the new active tab in the tabs.
     */
    resetTabIndexAndSetNewTabIndex(newTab) {
        var _a;
        (_a = this.tabs) === null || _a === void 0 ? void 0 : _a.forEach(tab => {
            tab.setAttribute('tabindex', tab === newTab ? '0' : '-1');
        });
    }
    /**
     * Removes active attribute from all tabs and sets active on the new tab.
     *
     * @internal
     *
     * @param tabId - The id of the new active tab.
     */
    setActiveTab(newTab) {
        var _a;
        (_a = this.tabs) === null || _a === void 0 ? void 0 : _a.forEach(tab => {
            if (tab === newTab) {
                tab.setAttribute('active', '');
            }
            else {
                tab.removeAttribute('active');
            }
        });
    }
    /**
     * Set focus on the new tab, then scroll it into view.
     *
     * @internal
     *
     * @param tab - Tab to set focus on.
     */
    async focusTab(tab) {
        if (!(tab instanceof Tab)) {
            return;
        }
        if (tab !== (document === null || document === void 0 ? void 0 : document.activeElement)) {
            this.resetTabIndexAndSetNewTabIndex(tab);
            tab.focus();
        }
        await this.handleArrowButtonVisibility();
    }
    /**
     * Handle the keydown event. The arrow keys, Home, End, Enter, and Space keys are supported.
     *
     * @internal
     *
     * @param event - HTML Keyboard Event.
     */
    async handleKeydown(event) {
        const tab = event.target;
        if (!(tab instanceof Tab)) {
            return;
        }
        if (!this.tabs) {
            return;
        }
        const previousTab = getPreviousTab(this.tabs, tab);
        const nextTab = getNextTab(this.tabs, tab);
        const firstTab = getFirstTab(this.tabs);
        const lastTab = getLastTab(this.tabs);
        switch (event.code) {
            case KEYCODES.LEFT:
                event.preventDefault();
                await this.focusTab(!this.isRtl() ? previousTab : nextTab);
                break;
            case KEYCODES.RIGHT:
                event.preventDefault();
                await this.focusTab(!this.isRtl() ? nextTab : previousTab);
                break;
            case KEYCODES.HOME:
                await this.focusTab(firstTab);
                break;
            case KEYCODES.END:
                await this.focusTab(lastTab);
                break;
            default:
        }
    }
    /**
     * Should the arrow button be shown.
     *
     * @internal
     *
     * @param direction - The direction of the arrow button.
     */
    shouldShowArrowButton(direction) {
        return direction === ARROW_BUTTON_DIRECTION.FORWARD ? this.showForwardArrowButton : this.showBackwardArrowButton;
    }
    /**
     * If an arrow button is in focus and it gets removed, switch focus to the opposite arrow button.
     * If both arrow buttons are removed, switch focus to the active tab.
     *
     * @internal
     */
    async switchFocus() {
        var _a, _b;
        await this.updateComplete;
        if (!this.showBackwardArrowButton && !this.showForwardArrowButton) {
            (_a = getActiveTab(this.tabs || [])) === null || _a === void 0 ? void 0 : _a.focus();
        }
        else if ((this.showBackwardArrowButton && !this.showForwardArrowButton) ||
            (this.showForwardArrowButton && !this.showBackwardArrowButton)) {
            (_b = this.notFocusedArrowButton) === null || _b === void 0 ? void 0 : _b.focus();
        }
    }
    /**
     * Show or hide the arrow buttons based on the position of the tabs
     * corresponding to the tab list.
     *
     * @internal
     */
    async handleArrowButtonVisibility() {
        var _a, _b, _c;
        if (!this.tabs) {
            return;
        }
        if (!this.tabsContainer) {
            return;
        }
        let isArrowButtonFocused = false;
        /**
         * If the active element is an arrow button,
         * then run the switchFocus function after setting the arrow button visibility.
         */
        if (((_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.activeElement) instanceof Button) {
            isArrowButtonFocused = true;
        }
        const firstTab = getFirstTab(this.tabs);
        const lastTab = getLastTab(this.tabs);
        const firstTabLeftEdgePosition = firstTab.getBoundingClientRect().left;
        const tabListLeftEdgePosition = (_b = this.tabsContainer) === null || _b === void 0 ? void 0 : _b.getBoundingClientRect().left;
        const lastTabRightEdgePosition = lastTab.getBoundingClientRect().right;
        const tabListRightEdgePosition = (_c = this.tabsContainer) === null || _c === void 0 ? void 0 : _c.getBoundingClientRect().right;
        // when RTL, the edges are reversed
        const firstTabRightEdgePosition = firstTab.getBoundingClientRect().right;
        const lastTabLeftEdgePosition = lastTab.getBoundingClientRect().left;
        if (!this.isRtl()) {
            if (firstTabLeftEdgePosition <= tabListLeftEdgePosition) {
                this.showBackwardArrowButton = true;
            }
            else {
                this.showBackwardArrowButton = false;
            }
            if (lastTabRightEdgePosition > tabListRightEdgePosition) {
                this.showForwardArrowButton = true;
            }
            else {
                this.showForwardArrowButton = false;
            }
            if (isArrowButtonFocused) {
                await this.switchFocus();
            }
            return;
        }
        if (firstTabRightEdgePosition > tabListRightEdgePosition) {
            this.showBackwardArrowButton = true;
        }
        else {
            this.showBackwardArrowButton = false;
        }
        if (lastTabLeftEdgePosition < tabListLeftEdgePosition) {
            this.showForwardArrowButton = true;
        }
        else {
            this.showForwardArrowButton = false;
        }
        if (isArrowButtonFocused) {
            await this.switchFocus();
        }
    }
    /**
     * Scroll tabs forward or backward.
     *
     * @internal
     *
     * @param direction - The direction to scroll the tabs in.
     */
    async scrollTabs(direction) {
        var _a;
        const forwardMultiplier = !this.isRtl() ? 1 : -1;
        const backwardMultiplier = !this.isRtl() ? -1 : 1;
        (_a = this.tabsContainer) === null || _a === void 0 ? void 0 : _a.scrollBy({
            left: this.tabsContainer.clientWidth *
                (direction === ARROW_BUTTON_DIRECTION.FORWARD ? forwardMultiplier : backwardMultiplier),
            // @ts-ignore : https://github.com/Microsoft/TypeScript/issues/28755
            behavior: 'instant',
        });
        await this.handleArrowButtonVisibility();
    }
    render() {
        const forwardArrowDirection = !this.isRtl() ? 'right' : 'left';
        const backwardArrowDirection = !this.isRtl() ? 'left' : 'right';
        const arrowButton = (direction) => html ` ${this.shouldShowArrowButton(direction)
            ? html `<mdc-button
            variant="tertiary"
            prefix-icon="arrow-${direction === ARROW_BUTTON_DIRECTION.FORWARD
                ? forwardArrowDirection
                : backwardArrowDirection}-regular"
            aria-label="Scroll tabs ${direction === ARROW_BUTTON_DIRECTION.FORWARD
                ? forwardArrowDirection
                : backwardArrowDirection}"
            @click="${() => this.scrollTabs(direction)}"
          ></mdc-button>`
            : nothing}`;
        return html ` ${arrowButton('backward')}
      <div class="container" role="${ROLE.TABLIST}" tabindex="-1" aria-label="${ifDefined(this.dataAriaLabel)}">
        <slot></slot>
      </div>
      ${arrowButton(ARROW_BUTTON_DIRECTION.FORWARD)}`;
    }
}
TabList.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, attribute: 'active-tab-id', reflect: true }),
    __metadata("design:type", String)
], TabList.prototype, "activeTabId", void 0);
__decorate([
    property({ type: String, attribute: 'data-aria-label' }),
    __metadata("design:type", String)
], TabList.prototype, "dataAriaLabel", void 0);
__decorate([
    query('.container'),
    __metadata("design:type", HTMLDivElement)
], TabList.prototype, "tabsContainer", void 0);
__decorate([
    query('mdc-button:not(:focus-visible)'),
    __metadata("design:type", Button)
], TabList.prototype, "notFocusedArrowButton", void 0);
__decorate([
    queryAssignedElements({ selector: 'mdc-tab' }),
    __metadata("design:type", Array)
], TabList.prototype, "tabs", void 0);
__decorate([
    state(),
    __metadata("design:type", Boolean)
], TabList.prototype, "showForwardArrowButton", void 0);
__decorate([
    state(),
    __metadata("design:type", Boolean)
], TabList.prototype, "showBackwardArrowButton", void 0);
export default TabList;
