// @momentum-design/components/dist/components/combobox/combobox.events.js
export class ComboboxEventManager {
    /**
     * Dispatches a custom event for the Combobox.
     *
     * @param instance - The combobox instance.
     * @param eventName - The name of the event.
     * @param option - The option that triggered the event.
     * @param bubbles - Whether the event should bubble up the DOM tree.
     */
    static dispatchComboboxEvent(instance, eventName, option, bubbles = true) {
        instance.dispatchEvent(new CustomEvent(eventName, {
            detail: { value: option === null || option === void 0 ? void 0 : option.value, label: option === null || option === void 0 ? void 0 : option.label },
            composed: true,
            bubbles,
        }));
    }
    /**
     * Dispatches a custom event for the Combobox when the input value changes.
     *
     * @param instance - The combobox instance.
     * @param option - The value and label of the selected option.
     */
    static onInputCombobox(instance, option) {
        this.dispatchComboboxEvent(instance, 'input', option);
    }
    /**
     * Dispatches a custom event for the Combobox when the selected option changes.
     *
     * @param instance - The combobox instance.
     * @param option - The value and label of the selected option.
     */
    static onChangeCombobox(instance, option) {
        this.dispatchComboboxEvent(instance, 'change', option);
    }
}
