// @momentum-design/components/dist/components/searchfield/searchfield.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { queryAssignedElements, state } from 'lit/decorators.js';
import { classMap } from 'lit-html/directives/class-map.js';
import Input from '../input/input.component';
import { KEYS } from '../../utils/keys';
import styles from './searchfield.styles';
import { DEFAULTS } from './searchfield.constants';
/**
 * `mdc-searchfield` component is used as an input field for search functionality.
 *
 * It supports `mdc-inputchip` as filters.
 *
 * This component is built by extending the `mdc-input` component.
 *
 * @tagname mdc-searchfield
 *
 * @event input - (React: onInput) This event is dispatched when the value of the input field changes (every press).
 * @event change - (React: onChange) This event is dispatched when the value of the input field changes (on blur).
 * @event focus - (React: onFocus) This event is dispatched when the input receives focus.
 * @event blur - (React: onBlur) This event is dispatched when the input loses focus.
 * @event clear - (React: onClear) This event is dispatched when the input text is cleared.
 *
 * @slot filters - Slot for input chips
 */
class Searchfield extends Input {
    constructor() {
        super(...arguments);
        /**
         * @internal
         */
        this.isInputFocused = false;
        /**
         * @internal
         */
        this.hasInputChips = false;
    }
    /**
     * Handles the keydown event of the search field.
     * If the key pressed is 'Enter', it submits the form.
     * If the key pressed is 'Escape', it clears the input text.
     * @param event - Keyboard event
     */
    handleKeyDown(event) {
        super.handleKeyDown(event);
        if (event.key === KEYS.ESCAPE) {
            this.clearInputText();
        }
    }
    connectedCallback() {
        super.connectedCallback();
        this.leadingIcon = DEFAULTS.ICON;
        this.trailingButton = DEFAULTS.CLOSE_BTN;
        this.helpText = undefined;
        this.helpTextType = undefined;
        this.required = undefined;
        this.validationMessage = undefined;
        this.prefixText = undefined;
    }
    /**
     * This method is used to render the input chips inside filters slot.
     * It will remove any elements that are not input chips.
     */
    renderInputChips() {
        var _a;
        this.hasInputChips = !!((_a = this.inputChips) === null || _a === void 0 ? void 0 : _a.length);
        if (this.inputChips) {
            this.inputChips.forEach(element => {
                if (!element.matches(DEFAULTS.INPUT_CHIP_TAG)) {
                    element.remove();
                }
            });
        }
    }
    /**
     * This sets the focus ring state based on the input focus state.
     * Eventually, this logic has to be omitted and achieved using CSS instead.
     * @override
     */
    firstUpdated() {
        this.inputElement.onfocus = () => {
            this.isInputFocused = true;
        };
        this.inputElement.onblur = () => {
            this.isInputFocused = false;
        };
    }
    clearInputText() {
        var _a;
        super.clearInputText();
        (_a = this.inputChips) === null || _a === void 0 ? void 0 : _a.forEach(element => {
            // Dispatch the custom 'remove' event from inputChip
            element.dispatchEvent(new CustomEvent('remove', { bubbles: true, composed: true }));
        });
    }
    render() {
        return html `
      ${this.renderLabelElement()}
      <div
        class="${classMap({
            'input-container': true,
            'mdc-focus-ring': this.isInputFocused,
        })}"
        part="input-container"
      >
        ${this.renderLeadingIcon()}
        <div part="scrollable-container" tabindex="-1">
          <div
            part="filters-container"
            @click=${() => this.inputElement.focus()}
            @keydown=${(e) => (e.key === KEYS.ENTER ? this.inputElement.focus() : null)}
            @keyup=${(e) => (e.key === KEYS.SPACE ? this.inputElement.focus() : null)}
          >
            <slot name="filters" @slotchange=${this.renderInputChips}></slot>
          </div>
          ${this.renderInputElement(DEFAULTS.TYPE, this.hasInputChips)}
        </div>
        ${this.renderTrailingButton(this.hasInputChips)}
      </div>
    `;
    }
}
Searchfield.styles = [...Input.styles, ...styles];
__decorate([
    queryAssignedElements({ slot: 'filters' }),
    __metadata("design:type", Array)
], Searchfield.prototype, "inputChips", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], Searchfield.prototype, "isInputFocused", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], Searchfield.prototype, "hasInputChips", void 0);
export default Searchfield;
