// @momentum-design/components/dist/components/progressbar/progressbar.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { v4 as uuidv4 } from 'uuid';
import FormfieldWrapper from '../formfieldwrapper';
import { DataAriaLabelMixin } from '../../utils/mixins/DataAriaLabelMixin';
import { VALIDATION } from '../formfieldwrapper/formfieldwrapper.constants';
import { ROLE } from '../../utils/roles';
import { DEFAULTS, VARIANT } from './progressbar.constants';
import styles from './progressbar.styles';
/**
 * mdc-progressbar component visually represents a progress indicator, typically used to show
 * the completion state of an ongoing process (e.g., loading, file upload, etc.).
 * It contains an optional label and an optional helper text.
 *
 * - It supports mainly two types: Default and Inline
 * - It supports three validation variants: Default, Success and Error.
 *
 * This component is created by extending FormfieldWrapper.
 *
 * @tagname mdc-progressbar
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @cssproperty --mdc-progressbar-default-background-color - Background color of the progressbar when inactive.
 * @cssproperty --mdc-progressbar-default-active-background-color - Background color of the progressbar when active.
 * @cssproperty --mdc-progressbar-success-background-color - Background color of the progressbar when in success state.
 * @cssproperty --mdc-progressbar-error-background-color - Background color of the progressbar when in error state.
 * @cssproperty --mdc-progressbar-height - The height of the progressbar.
 * @cssproperty --mdc-progressbar-border-radius - The border radius of the progressbar.
 * @cssproperty --mdc-progressbar-label-color - Color of the progressbar label text.
 * @cssproperty --mdc-progressbar-label-lineheight - Line height of the label text.
 * @cssproperty --mdc-progressbar-label-fontsize - Font size of the label text.
 * @cssproperty --mdc-progressbar-label-fontweight - Font weight of the label text.
 * @cssproperty --mdc-progressbar-help-text-color - Color of the help text.
 */
class Progressbar extends DataAriaLabelMixin(FormfieldWrapper) {
    constructor() {
        super(...arguments);
        /**
         * Types of the progressbar
         * - **Default**
         * - **Inline**
         *
         * @default default
         */
        this.variant = DEFAULTS.VARIANT;
        /**
         * The current progress as a percentage, 0 to 100.
         * The value will be clamped between 0 and 100.
         * @default 0
         */
        this.value = '0';
        /**
         * Define error state of the progressbar
         * - **true**
         * - **false**
         * @default false
         */
        this.error = false;
        /**
         * Renders the progress bar with dynamic width and variant styles.
         * @returns The rendered HTML for the progress bar.
         * @internal
         */
        this.renderProgressbar = () => {
            var _a;
            const variant = this.getValidationVariant();
            const isGap = this.clampedValue > 0 && this.clampedValue < 100;
            const progressWidth = this.error ? '100' : `${this.clampedValue}`;
            return html `
      <div
        part="progress-container ${isGap ? 'gap' : ''}"
        role="${ROLE.PROGRESSBAR}"
        aria-valuenow="${this.clampedValue}"
        aria-valuemin="0"
        aria-valuemax="100"
        aria-label="${(_a = this.dataAriaLabel) !== null && _a !== void 0 ? _a : ''}"
      >
        <div part="progress-bar ${variant}" style="width: ${progressWidth}%"></div>
        <div part="remaining"></div>
      </div>
    `;
        };
    }
    connectedCallback() {
        super.connectedCallback();
        this.id = `mdc-progressbar-${uuidv4()}`;
        this.disabled = undefined;
        this.toggletipText = undefined;
        this.toggletipPlacement = undefined;
    }
    /**
     * Ensures that the value is clamped between 0 and 100
     * @returns The clamped value
     * @internal
     */
    get clampedValue() {
        const value = Number(this.value);
        const clampedValue = Number.isNaN(value) ? 0 : Math.max(0, Math.min(100, value));
        return clampedValue;
    }
    /**
     * Determines the validation state (success, error, or default) based on progress value and error state.
     * @returns The appropriate validation state for the progressbar.
     */
    getValidationVariant() {
        if (this.error) {
            this.helpTextType = VALIDATION.ERROR;
        }
        else if (this.clampedValue === 100) {
            this.helpTextType = VALIDATION.SUCCESS;
        }
        else {
            this.helpTextType = VALIDATION.DEFAULT;
        }
        return this.helpTextType;
    }
    /**
     * Renders the component based on the variant (inline or default).
     * @returns The rendered HTML for the component.
     */
    render() {
        const isInline = this.variant === VARIANT.INLINE;
        return html `
      ${isInline
            ? html `<div part="inline-label-container">${this.renderLabelElement()} ${this.renderProgressbar()}</div>`
            : html `
            <div part="label-container">
              ${this.renderLabelElement()}
              ${this.variant === VARIANT.DEFAULT && this.label
                ? html `<span part="percentage">${this.clampedValue}%</span>`
                : ''}
            </div>
            ${this.renderProgressbar()} ${this.renderHelperText()}
          `}
    `;
    }
}
Progressbar.styles = [...FormfieldWrapper.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Progressbar.prototype, "variant", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Progressbar.prototype, "value", void 0);
__decorate([
    property({ type: Boolean, attribute: 'error' }),
    __metadata("design:type", Object)
], Progressbar.prototype, "error", void 0);
export default Progressbar;
