// @momentum-design/components/dist/components/inputchip/inputchip.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { Component } from '../../models';
import { DisabledMixin } from '../../utils/mixins/DisabledMixin';
import { IconNameMixin } from '../../utils/mixins/IconNameMixin';
import styles from './inputchip.styles';
import { DEFAULTS } from './inputchip.constants';
/**
 * mdc-inputchip component is an interactive chip that consumers can use to represent an input.
 *
 * - It supports a leading icon along with label.
 * - It supports a prefix slot for avatars (takes precedence over icon-name).
 * - It supports an error state for validation.
 * - It supports a close button to remove the chip.
 *
 * @tagname mdc-inputchip
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @event remove - This event is dispatched when the close button is activated. It bubbles and is composed.
 *
 * @slot prefix - A slot for prefix content such as avatars.
 *
 * @csspart label - The label part of the chip.
 * @csspart icon - The icon part of the chip.
 * @csspart close-icon - The close icon part of the chip.
 *
 * @cssproperty --mdc-chip-color - The color of the chip.
 * @cssproperty --mdc-chip-border-color - The border color of the chip.
 * @cssproperty --mdc-chip-background-color - The background color of the chip.
 *
 */
class InputChip extends IconNameMixin(DisabledMixin(Component)) {
    constructor() {
        super(...arguments);
        /**
         * The label of the inputchip.
         *
         * We recommend limiting the <b>maximum length of the label text to 20 characters</b>,
         * including empty spaces to split words.
         */
        this.label = '';
        /**
         * The error state of the inputchip.
         *
         * @default false
         */
        this.error = false;
        /**
         * The aria-label of the close button.
         */
        this.clearAriaLabel = '';
    }
    /**
     * Renders the prefix content, supporting both icons and slot content.
     * @returns The prefix content if available, otherwise nothing.
     */
    renderPrefix() {
        return html `
      <slot name="prefix">
        ${this.iconName
            ? html `<mdc-icon part="icon" name="${this.iconName}" length-unit="rem" size="1"></mdc-icon>`
            : nothing}
      </slot>
    `;
    }
    /**
     * Handles the behavior of the close button on click event.
     * @param event - The event object.
     */
    handleClose() {
        this.dispatchEvent(new CustomEvent('remove', { bubbles: true, composed: true }));
    }
    render() {
        return html `
      ${this.renderPrefix()}
      ${this.label
            ? html `<mdc-text part="label" type="${DEFAULTS.TEXT_TYPE}" tagname="${DEFAULTS.TAG_NAME}"
            >${this.label}</mdc-text
          >`
            : nothing}
      <mdc-button
        ?disabled="${this.disabled}"
        variant="tertiary"
        part="close-icon"
        aria-label="${this.clearAriaLabel}"
        prefix-icon="${DEFAULTS.CLOSE_ICON}"
        size="20"
        @click="${this.handleClose}"
      ></mdc-button>
    `;
    }
}
InputChip.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String }),
    __metadata("design:type", Object)
], InputChip.prototype, "label", void 0);
__decorate([
    property({ type: Boolean }),
    __metadata("design:type", Object)
], InputChip.prototype, "error", void 0);
__decorate([
    property({ type: String, attribute: 'clear-aria-label' }),
    __metadata("design:type", Object)
], InputChip.prototype, "clearAriaLabel", void 0);
export default InputChip;
