// @momentum-design/components/dist/react/slider/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/slider';
import { TAG_NAME } from '../../components/slider/slider.constants';
/**
 * Slider component is used to select a value or range of values from within a defined range.
 * It provides a visual representation of the current value(s) and allows users to adjust the value(s) by dragging the thumb(s) along the track.
 * It can be used as a single slider or a range slider. This is set by the boolean attribute `range`
 * If the step value is more than 1, tick marks are shown to represent the steps between the min and max values. The slider thumb will snap to the nearest tick mark.
 *
 * ## Spatial navigation
 *
 * In spatial navigation mode, the user has to step into edit mode to adjust the slider value using left and right navigation keys.
 * To enter edit mode, the user can press the Enter when one of the slider's handler is focused. User can leave edit mode by press
 * Enter or Escape.
 *
 * @tagname mdc-slider
 *
 * @dependency mdc-icon
 *
 * @csspart slider-tooltip - The tooltip of the slider
 * @csspart slider-track - The track of the slider
 * @csspart slider-wrapper - The wrapper around the slider input(s)
 * @csspart slider-ticks - The container for the tick marks
 * @csspart slider-tick - The individual tick marks
 * @csspart slider-input - The input element of the slider
 * @csspart slider-label - The label of the slider
 *
 * @event input - Fired when the slider value changes
 * @event change - Fired when the slider value is committed
 *
 * @cssproperty --mdc-slider-thumb-color - The color of the slider thumb
 * @cssproperty --mdc-slider-thumb-border-color - The color of the slider thumb border
 * @cssproperty --mdc-slider-thumb-size - The size of the slider thumb
 * @cssproperty --mdc-slider-input-size - The height of the slider input
 * @cssproperty --mdc-slider-tick-size - The size of the slider tick marks
 * @cssproperty --mdc-slider-track-height - The height of the slider track
 * @cssproperty --mdc-slider-tick-color - The color of the slider tick marks
 * @cssproperty --mdc-slider-progress-color - The color of the slider progress
 * @cssproperty --mdc-slider-track-color - The color of the slider track
 * @cssproperty --mdc-slider-tooltip-left - The left position of the slider tooltip
 * @cssproperty --mdc-slider-tick-left - The left position of the slider tick marks
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change',
        onInput: 'input',
    },
    displayName: 'Slider',
});
export default reactWrapper;
