// @momentum-design/components/dist/utils/date-utils.js
/**
 * Lightweight date utilities using native Date and Intl APIs.
 * Replaces Luxon for calendar/datepicker components.
 *
 * All dates are handled as local dates (no timezone conversion).
 * ISO strings are in yyyy-mm-dd format.
 */
function toISODate(d) {
    const y = d.getFullYear();
    const m = String(d.getMonth() + 1).padStart(2, '0');
    const day = String(d.getDate()).padStart(2, '0');
    return `${y}-${m}-${day}`;
}
function parseISO(iso) {
    if (!iso)
        return null;
    if (!/^\d{4}-\d{2}-\d{2}$/.test(iso))
        return null;
    const parts = iso.split('-');
    const y = parseInt(parts[0], 10);
    const m = parseInt(parts[1], 10);
    const d = parseInt(parts[2], 10);
    const date = new Date(y, m - 1, d);
    if (date.getFullYear() !== y || date.getMonth() !== m - 1 || date.getDate() !== d)
        return null;
    return date;
}
function createDate(year, month, day = 1) {
    return new Date(year, month - 1, day);
}
function today() {
    const now = new Date();
    return new Date(now.getFullYear(), now.getMonth(), now.getDate());
}
function addDays(d, days) {
    const result = new Date(d);
    result.setDate(result.getDate() + days);
    return result;
}
/**
 * Add months, clamping the day to the last day of the target month
 * to match Luxon behavior (e.g. March 31 - 1 month → Feb 28, not March 3).
 */
function addMonths(d, months) {
    const targetMonth = d.getMonth() + months;
    const result = new Date(d.getFullYear(), targetMonth, 1);
    const maxDay = new Date(result.getFullYear(), result.getMonth() + 1, 0).getDate();
    result.setDate(Math.min(d.getDate(), maxDay));
    return result;
}
/**
 * Add years, clamping the day to the last day of the target month
 * (handles Feb 29 in leap years → Feb 28 in non-leap years).
 */
function addYears(d, years) {
    const targetYear = d.getFullYear() + years;
    const result = new Date(targetYear, d.getMonth(), 1);
    const maxDay = new Date(result.getFullYear(), result.getMonth() + 1, 0).getDate();
    result.setDate(Math.min(d.getDate(), maxDay));
    return result;
}
function isSameDay(a, b) {
    return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate();
}
function isBefore(a, b) {
    return a.getTime() < b.getTime();
}
function isAfter(a, b) {
    return a.getTime() > b.getTime();
}
function isBeforeOrEqual(a, b) {
    return a.getTime() <= b.getTime();
}
function startOfMonth(d) {
    return new Date(d.getFullYear(), d.getMonth(), 1);
}
function endOfMonth(d) {
    return new Date(d.getFullYear(), d.getMonth() + 1, 0);
}
function daysInMonth(year, month) {
    return new Date(year, month, 0).getDate();
}
/**
 * Returns the ISO weekday: 1=Monday .. 7=Sunday
 */
function isoWeekday(d) {
    const day = d.getDay();
    return day === 0 ? 7 : day;
}
function startOfWeek(d, weekStartDay = 1) {
    const jsDay = d.getDay();
    const isoDay = jsDay === 0 ? 7 : jsDay;
    const startIso = weekStartDay === 7 ? 7 : weekStartDay;
    const diff = (isoDay - startIso + 7) % 7;
    return addDays(d, -diff);
}
function endOfWeek(d, weekStartDay = 1) {
    return addDays(startOfWeek(d, weekStartDay), 6);
}
/**
 * Locale-aware weekday names using Intl.DateTimeFormat.
 */
function getWeekdayNames(locale, style = 'narrow') {
    const formatter = new Intl.DateTimeFormat(locale, { weekday: style });
    // Jan 5 2026 is a Monday
    const monday = new Date(2026, 0, 5);
    const names = [];
    for (let i = 0; i < 7; i += 1) {
        names.push(formatter.format(addDays(monday, i)));
    }
    return names;
}
/**
 * Returns the starting day of the week for a locale.
 * 1 = Monday (ISO), 7 = Sunday.
 */
function getWeekStartDay(locale) {
    var _a, _b;
    try {
        const loc = new Intl.Locale(locale);
        const weekInfo = (_a = loc.weekInfo) !== null && _a !== void 0 ? _a : (_b = loc.getWeekInfo) === null || _b === void 0 ? void 0 : _b.call(loc);
        if (weekInfo === null || weekInfo === void 0 ? void 0 : weekInfo.firstDay) {
            return weekInfo.firstDay;
        }
    }
    catch {
        // fallback
    }
    const sundayLocales = ['en', 'ja', 'ko', 'zh', 'pt-BR'];
    const lang = locale.split('-')[0].toLowerCase();
    const full = locale.toLowerCase();
    if (sundayLocales.includes(lang) || sundayLocales.includes(full)) {
        return 7;
    }
    return 1;
}
/**
 * Locale-aware month + year label (e.g. "July 2025").
 */
function getMonthYearLabel(year, month, locale) {
    const d = createDate(year, month, 1);
    return new Intl.DateTimeFormat(locale, { month: 'long', year: 'numeric' }).format(d);
}
/**
 * Locale-aware full date label for aria (e.g. "July 15, 2025").
 */
function formatDateFull(iso, locale) {
    const d = parseISO(iso);
    if (!d)
        return '';
    return new Intl.DateTimeFormat(locale, { dateStyle: 'full' }).format(d);
}
/**
 * Locale-aware medium date format (e.g. "Jul 15, 2025").
 */
function formatDateMed(iso, locale) {
    const d = parseISO(iso);
    if (!d)
        return '';
    return new Intl.DateTimeFormat(locale, { dateStyle: 'medium' }).format(d);
}
/**
 * Format a date range for display using the native Intl.DateTimeFormat.formatRange API.
 * Produces locale-correct output for all cases: same month, cross-month, and cross-year ranges.
 * e.g. en-US: "July 13 – 19, 2025", de-DE: "13.–19. Juli 2025", ja-JP: "2025年7月13日～19日"
 */
function formatDateRange(startIso, endIso, locale) {
    const start = parseISO(startIso);
    const end = parseISO(endIso);
    if (!start || !end)
        return '';
    try {
        const formatter = new Intl.DateTimeFormat(locale, {
            year: 'numeric',
            month: 'long',
            day: 'numeric',
        });
        return formatter.formatRange(start, end);
    }
    catch {
        return `${formatDateMed(startIso, locale)} \u2013 ${formatDateMed(endIso, locale)}`;
    }
}
/**
 * Get the locale-specific date parts order and separator.
 */
function getLocaleDateParts(locale) {
    const formatter = new Intl.DateTimeFormat(locale, {
        year: 'numeric',
        month: '2-digit',
        day: '2-digit',
    });
    const parts = formatter.formatToParts(new Date(2000, 11, 31));
    const order = [];
    let separator = '/';
    for (const part of parts) {
        if (part.type === 'month' && !order.includes('month'))
            order.push('month');
        if (part.type === 'day' && !order.includes('day'))
            order.push('day');
        if (part.type === 'year' && !order.includes('year'))
            order.push('year');
        if (part.type === 'literal' && part.value.trim()) {
            separator = part.value.trim();
        }
    }
    if (order.length !== 3) {
        return { order: ['month', 'day', 'year'], separator };
    }
    return { order, separator };
}
export { addDays, addMonths, addYears, createDate, daysInMonth, endOfMonth, endOfWeek, formatDateFull, formatDateMed, formatDateRange, getLocaleDateParts, getMonthYearLabel, getWeekdayNames, getWeekStartDay, isAfter, isBefore, isBeforeOrEqual, isSameDay, isoWeekday, parseISO, startOfMonth, startOfWeek, today, toISODate, };
