// @momentum-design/components/dist/components/illustrationprovider/illustrationprovider.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import { Provider } from '../../models';
import IllustrationProviderContext from './illustrationprovider.context';
import { ALLOWED_FILE_EXTENSIONS, DEFAULTS } from './illustrationprovider.constants';
/**
 * IllustrationProvider component, which allows to be consumed from sub components
 * (see `providerUtils.consume` for how to consume)
 *
 * Attribute `illustrationSet` can be set to either `momentum-illustrations` or `custom-illustrations`.
 * If `momentum-illustrations` is selected, the illustrations will be fetched from the
 * Momentum Design System illustration set per a dynamic JS Import (no need to provide a URL).
 * This requires the consumer to have the `@momentum-design/illustrations` package installed and the
 * build tooling needs to support dynamic imports.
 *
 * If `custom-illustrations` is selected, the illustrations will be fetched from the provided URL.
 * This requires the consumer to provide a URL from which the illustrations will be fetched and
 * the consumer needs to make sure to bundle the illustrations in the application.
 *
 * If `cacheStrategy` is provided (only works with illustrationSet = `custom-illustrations`), the
 * IllustrationProvider will cache the illustrations in the selected cache (either web-api-cache or in-memory-cache),
 * to avoid fetching the same illustration multiple times over the network.
 * This is useful when the same illustration is used multiple times in the application.
 * To consider:
 * - The `in-memory-cache` is not persisted and will be lost when the
 * IllustrationProvider is removed from the DOM.
 * - The `web-api-cache` is persisted, but only works in https environments
 * (https://developer.mozilla.org/en-US/docs/Web/API/Cache).
 *
 * @tagname mdc-illustrationprovider
 *
 * @slot - children
 */
class IllustrationProvider extends Provider {
    constructor() {
        // initialise the context by running the Provider constructor:
        super({
            context: IllustrationProviderContext.context,
            initialValue: new IllustrationProviderContext(),
        });
        /**
         * Illustration set to be used
         *
         * If `momentum-illustrations` is selected, the illustrations will be fetched from the
         * Momentum Design System illustration set per a dynamic JS Import (no need to provide a URL).
         * This requires the consumer to have the `@momentum-designs` package installed and the
         * build tooling needs to support dynamic imports.
         *
         * If `custom-illustrations` is selected, the illustrations will be fetched from the provided URL.
         * This requires the consumer to provide a URL from which the illustrations will be fetched and
         * the consumer needs to make sure to bundle the illustrations in the application.
         *
         * @default momentum-illustrations
         */
        this.illustrationSet = DEFAULTS.ILLUSTRATION_SET;
        /**
         * File extension of illustrations
         * (if Illustration set is `custom-illustrations`, this will be the file extension for illustrations)
         * @default svg
         */
        this.fileExtension = DEFAULTS.FILE_EXTENSION;
    }
    /**
     * Context object of the IllustrationProviderContext, to be consumed by child components
     */
    static get Context() {
        return IllustrationProviderContext.context;
    }
    updateValuesInContext() {
        // only update fileExtension on context if its an allowed fileExtension
        if (this.fileExtension && ALLOWED_FILE_EXTENSIONS.includes(this.fileExtension)) {
            this.context.value.fileExtension = this.fileExtension;
        }
        else {
            // Ensure both fileExtension and context are updated to the default if its not an allowed fileExtension
            this.fileExtension = DEFAULTS.FILE_EXTENSION;
            this.context.value.fileExtension = DEFAULTS.FILE_EXTENSION;
        }
        this.context.value.illustrationSet = this.illustrationSet;
        this.context.value.url = this.url;
        this.context.value.cacheName = this.cacheName;
        this.context.value.cacheStrategy = this.cacheStrategy;
    }
    updateContext() {
        if (this.context.value.fileExtension !== this.fileExtension ||
            this.context.value.illustrationSet !== this.illustrationSet ||
            this.context.value.url !== this.url ||
            this.context.value.cacheName !== this.cacheName ||
            this.context.value.cacheStrategy !== this.cacheStrategy) {
            this.updateValuesInContext();
            this.context.updateObservers();
        }
    }
}
__decorate([
    property({ type: String, attribute: 'illustration-set', reflect: true }),
    __metadata("design:type", String)
], IllustrationProvider.prototype, "illustrationSet", void 0);
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], IllustrationProvider.prototype, "url", void 0);
__decorate([
    property({ type: String, attribute: 'file-extension', reflect: true }),
    __metadata("design:type", String)
], IllustrationProvider.prototype, "fileExtension", void 0);
__decorate([
    property({ type: String, attribute: 'cache-strategy' }),
    __metadata("design:type", String)
], IllustrationProvider.prototype, "cacheStrategy", void 0);
__decorate([
    property({ type: String, attribute: 'cache-name' }),
    __metadata("design:type", String)
], IllustrationProvider.prototype, "cacheName", void 0);
export default IllustrationProvider;
