// @momentum-design/components/dist/components/buttonsimple/buttonsimple.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { Component } from '../../models';
import { KEYS } from '../../utils/keys';
import { AutoFocusMixin } from '../../utils/mixins/AutoFocusMixin';
import { DisabledMixin } from '../../utils/mixins/DisabledMixin';
import { TabIndexMixin } from '../../utils/mixins/TabIndexMixin';
import { BUTTON_TYPE, DEFAULTS } from './buttonsimple.constants';
import styles from './buttonsimple.styles';
/**
 * `mdc-buttonsimple` is a component that can be configured in various ways to suit different use cases.
 * It is used as an internal component and is not intended to be used directly by consumers.
 * Consumers should use the `mdc-button` component instead.
 *
 * @event click - (React: onClick) This event is dispatched when the button is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the button.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the button.
 * @event focus - (React: onFocus) This event is dispatched when the button receives focus.
 *
 * @tagname mdc-buttonsimple
 */
class Buttonsimple extends AutoFocusMixin(TabIndexMixin(DisabledMixin(Component))) {
    /** @internal */
    get form() {
        return this.internals.form;
    }
    constructor() {
        super();
        /**
         * Simplebutton size is a super set of all the sizes supported by children components.
         * @default 32
         */
        this.size = DEFAULTS.SIZE;
        /**
         * This property defines the ARIA role for the element. By default, it is set to 'button'.
         * Consumers should override this role when:
         * - The element is being used in a context where a different role is more appropriate.
         * - Custom behaviors are implemented that require a specific ARIA role for accessibility purposes.
         * @default button
         */
        this.role = DEFAULTS.ROLE;
        /**
         * This property defines the type attribute for the button element.
         * The type attribute specifies the behavior of the button when it is clicked.
         * - **submit**: The button submits the form data to the server.
         * - **reset**: The button resets the form data to its initial state.
         * - **button**: The button does nothing when clicked.
         * @default button
         */
        this.type = DEFAULTS.TYPE;
        /**
         * @internal
         */
        this.prevTabindex = 0;
        /** @internal */
        this.internals = this.attachInternals();
        this.addEventListener('click', this.executeAction.bind(this));
        this.addEventListener('keydown', this.handleKeyDown.bind(this));
        this.addEventListener('keyup', this.handleKeyUp.bind(this));
        this.addEventListener('blur', this.handleBlur.bind(this));
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('disabled')) {
            this.setDisabled(this, this.disabled);
        }
        if (changedProperties.has('softDisabled')) {
            this.setSoftDisabled(this, this.softDisabled);
        }
        if (changedProperties.has('active')) {
            if (this.active !== undefined) {
                // if active is not undefined and no ariaStateKey is provided, set it to the default (= aria-pressed)
                // this will make sure that if active is set to true or false regardless
                // the ariaStateKey fallback will still work
                this.ariaStateKey = this.ariaStateKey || DEFAULTS.ARIA_STATE_KEY;
            }
            this.setActive(this, this.active);
        }
    }
    executeAction() {
        if (this.type === BUTTON_TYPE.SUBMIT && this.form) {
            this.form.requestSubmit();
        }
        if (this.type === BUTTON_TYPE.RESET && this.form) {
            this.form.reset();
        }
    }
    /**
     * Sets the ariaStateKey attributes based on the active state of the button.
     * @param element - The button element
     * @param active - The active state of the element
     */
    setActive(element, active) {
        if (this.ariaStateKey) {
            const ariaStateKeys = this.ariaStateKey.split(',');
            ariaStateKeys
                .filter(key => key.trim().startsWith('aria-'))
                .forEach(key => {
                if (active === true) {
                    element.setAttribute(key.trim(), 'true');
                }
                else if (active === false) {
                    element.setAttribute(key.trim(), 'false');
                }
                else {
                    // If the active state is not a boolean, remove the attribute
                    element.removeAttribute(key.trim());
                }
            });
        }
    }
    /**
     * Sets the soft-disabled attribute for the button.
     * When soft-disabled, the button looks to be disabled but remains focusable and clickable.
     * Also sets/removes aria-disabled attribute.
     *
     * @param element - The button element.
     * @param softDisabled - The soft-disabled state.
     */
    setSoftDisabled(element, softDisabled) {
        if (softDisabled) {
            element.setAttribute('aria-disabled', 'true');
        }
        else {
            element.removeAttribute('aria-disabled');
        }
    }
    /**
     * Sets the disabled attribute for the button.
     * When disabled, the button is not focusable or clickable, and tabindex is set to -1.
     * The previous tabindex is stored and restored when enabled.
     * Also sets/removes aria-disabled attribute.
     *
     * @param element - The button element.
     * @param disabled - The disabled state.
     */
    setDisabled(element, disabled) {
        if (disabled) {
            element.setAttribute('aria-disabled', 'true');
            this.prevTabindex = this.tabIndex;
            this.tabIndex = -1;
        }
        else {
            if (this.tabIndex === -1) {
                this.tabIndex = this.prevTabindex;
            }
            element.removeAttribute('aria-disabled');
        }
    }
    triggerClickEvent() {
        const clickEvent = new MouseEvent('click', {
            bubbles: true,
            cancelable: true,
            view: window,
        });
        this.dispatchEvent(clickEvent);
    }
    /**
     * In case the button is pressed and the focus is lost while pressing,
     * the pressed class is removed.
     */
    handleBlur() {
        if (this.classList.contains('pressed')) {
            this.classList.remove('pressed');
        }
    }
    /**
     * Handles the keydown event on the button.
     * If the key is 'Enter' or 'Space', the button is pressed.
     * If the key is 'Enter', the button is pressed. The native HTML button works in the same way.
     * If the key is 'Space', the button's default is prevent to avoid scrolling etc in the host application.
     *
     * @param event - The keyboard event.
     */
    handleKeyDown(event) {
        if ([KEYS.ENTER, KEYS.SPACE].includes(event.key)) {
            this.classList.add('pressed');
            if (event.key === KEYS.ENTER) {
                this.triggerClickEvent();
            }
            // preventing the default event behavior for space key
            // to avoid scrolling etc in the host application
            // preventing the default event behavior for enter key
            // to avoid the keyup event to be triggered
            event.preventDefault();
        }
    }
    /**
     * Handles the keyup event on the button.
     * If the key is 'Enter' or 'Space', the button is clicked.
     * If the key is 'Space', the button is pressed. The native HTML button works in the same way.
     *
     * @param event - The keyboard event.
     */
    handleKeyUp(event) {
        if ([KEYS.ENTER, KEYS.SPACE].includes(event.key)) {
            this.classList.remove('pressed');
            if (event.key === KEYS.SPACE) {
                this.triggerClickEvent();
            }
        }
    }
    render() {
        return html ` <slot></slot> `;
    }
}
/** @internal */
Buttonsimple.formAssociated = true;
Buttonsimple.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Boolean)
], Buttonsimple.prototype, "active", void 0);
__decorate([
    property({ type: Boolean, attribute: 'soft-disabled', reflect: true }),
    __metadata("design:type", Boolean)
], Buttonsimple.prototype, "softDisabled", void 0);
__decorate([
    property({ type: Number, reflect: true }),
    __metadata("design:type", Number)
], Buttonsimple.prototype, "size", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Buttonsimple.prototype, "role", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Buttonsimple.prototype, "ariaStateKey", void 0);
__decorate([
    property({ reflect: true }),
    __metadata("design:type", String)
], Buttonsimple.prototype, "type", void 0);
export default Buttonsimple;
