// @momentum-design/components/dist/components/popover/popover.stack.js
/**
 * Manages a stack of popovers to control their order and lifecycle.
 * This class allows adding, removing, and retrieving popovers
 * while maintaining their stacking behavior.
 *
 */
class PopoverStack {
    constructor() {
        /**
         * Stack to maintain the order of popovers
         * @internal
         */
        this.stack = [];
    }
    /**
     * Adds a popover to the stack
     *
     * @param popover - Popover instance
     */
    push(popover) {
        this.stack.push(popover);
    }
    /**
     * Removes the last popover from the stack
     *
     * @returns The last popover in the stack
     */
    pop() {
        return this.stack.pop();
    }
    /**
     * Returns the last popover in the stack
     * without removing it
     *
     * @returns The last popover in the stack
     */
    peek() {
        return this.stack[this.stack.length - 1];
    }
    /**
     * Removes a popover from the stack
     *
     * @param popover - Popover instance
     */
    remove(popover) {
        this.stack = this.stack.filter(item => item !== popover);
    }
    /**
     * Checks if the stack has a specific popover
     *
     * @param popover - Popover instance
     * @returns True if the stack has the popover, false otherwise
     */
    has(popover) {
        return this.stack.includes(popover);
    }
    /**
     * Clears the stack
     */
    clear() {
        this.stack = [];
    }
}
export const popoverStack = new PopoverStack();
