// @momentum-design/components/dist/components/popover/popover.utils.js
import { ROLE } from '../../utils/roles';
import { TAG_NAME as POPOVER_PORTAL_TAG_NAME } from './popover.portal.component';
export class PopoverUtils {
    constructor(popover) {
        /**
         * The portal element used when the popover is appended to another container.
         * @internal
         */
        this.portalElement = null;
        /**
         * Flag to indicate if the popover was diconnected because it was appended to another container, or
         * it was actually removed from the DOM.
         * @internal
         */
        this.disconnectAfterAppendTo = false;
        /** @internal */
        this.arrowPixelChange = false;
        this.popover = popover;
    }
    /**
     * Parses the delay string and sets the open and close delay values.
     * If the delay string is invalid, the delay is set to 0.
     * @returns An array containing the open and close delay values.
     * @throws An error if the delay value is invalid.
     */
    setupDelay() {
        try {
            const [openDelay, closeDelay] = this.popover.delay.split(',').map(delay => {
                const parsed = parseInt(delay, 10);
                if (Number.isNaN(parsed) || parsed < 0) {
                    throw new Error(`Invalid delay value: ${delay}`);
                }
                return parsed;
            });
            return [openDelay, closeDelay];
        }
        catch (error) {
            this.popover.delay = '0,0';
            return [0, 0];
        }
    }
    /**
     * Sets up the hover bridge for the popover.
     * The hover bridge is used to prevent the popover from closing when the mouse is over the popover.
     *
     * @param placement - The placement of the popover.
     */
    setupHoverBridge(placement) {
        const hoverBridge = this.popover.renderRoot.querySelector('div[part="popover-hover-bridge"]');
        Object.assign(hoverBridge.style, {
            top: '',
            left: '',
            right: '',
            bottom: '',
        });
        const triggerEl = this.popover.triggerElement;
        if (hoverBridge && triggerEl) {
            const popoverBBox = this.popover.getBoundingClientRect();
            const triggerBBox = triggerEl.getBoundingClientRect();
            let bridgeSize = 0;
            if (!triggerBBox)
                return;
            const popoverHeight = this.popover.offsetHeight || 0;
            const popoverWidth = this.popover.offsetWidth || 0;
            const side = placement.split('-')[0];
            // bridgeSize calculated here because, floating UI might flip side on the main axis, also it can be negative if the trigger and popover are overlapping
            if (side === 'top' || side === 'bottom') {
                bridgeSize = Math.max(0, Math.max(popoverBBox.top - triggerBBox.bottom, triggerBBox.top - popoverBBox.bottom));
            }
            else if (side === 'left' || side === 'right') {
                bridgeSize = Math.max(0, Math.max(triggerBBox.left - popoverBBox.right, popoverBBox.left - triggerBBox.right));
            }
            // ensure bridge overlap
            bridgeSize += 1;
            switch (side) {
                case 'top':
                    hoverBridge.style.height = `${bridgeSize}px`;
                    hoverBridge.style.bottom = `calc(-1 * (${bridgeSize}px))`;
                    hoverBridge.style.left = '50%';
                    hoverBridge.style.width = `${popoverWidth}px`;
                    break;
                case 'left':
                    hoverBridge.style.height = `${popoverHeight}px`;
                    hoverBridge.style.width = `${bridgeSize}px`;
                    hoverBridge.style.right = `calc(-1.5 * (${bridgeSize}px))`;
                    break;
                case 'right':
                    hoverBridge.style.height = `${popoverHeight}px`;
                    hoverBridge.style.width = `${bridgeSize}px`;
                    hoverBridge.style.left = `calc(-0.5 * (${bridgeSize}px))`;
                    break;
                case 'bottom':
                default:
                    hoverBridge.style.height = `${bridgeSize}px`;
                    hoverBridge.style.top = `calc(-1 * (${bridgeSize}px))`;
                    hoverBridge.style.left = '50%';
                    hoverBridge.style.width = `${popoverWidth}px`;
                    break;
            }
        }
    }
    /**
     * If the `appendTo` property is set, finds the corresponding
     * DOM element by its ID, and appends this popover as a child of that element.
     */
    setupAppendTo() {
        var _a, _b;
        if (this.popover.appendTo) {
            const appendToEl = document.getElementById(this.popover.appendTo);
            if (appendToEl && !Array.from(appendToEl.children).includes(this.popover)) {
                this.disconnectAfterAppendTo = true;
                this.portalElement = document.createElement(POPOVER_PORTAL_TAG_NAME);
                this.portalElement.onDisconnect = () => {
                    this.popover.remove();
                    this.portalElement = null;
                };
                (_b = (_a = this.popover.parentElement) === null || _a === void 0 ? void 0 : _a.appendChild) === null || _b === void 0 ? void 0 : _b.call(_a, this.portalElement);
                appendToEl.appendChild(this.popover);
            }
        }
    }
    /**
     * Remove portal component to when the popover appended to somewhere else and removed from the DOM
     */
    cleanupAppendTo() {
        if (!this.disconnectAfterAppendTo && this.portalElement) {
            this.portalElement.remove();
        }
        this.disconnectAfterAppendTo = false;
    }
    /**
     * Sets up the aria labels
     */
    updateAriaLabels() {
        var _a, _b, _c;
        if (this.popover.interactive && this.popover.role) {
            if (!this.popover.ariaLabel) {
                this.popover.ariaLabel =
                    ((_a = this.popover.triggerElement) === null || _a === void 0 ? void 0 : _a.ariaLabel) || ((_b = this.popover.triggerElement) === null || _b === void 0 ? void 0 : _b.textContent) || '';
            }
            if (!this.popover.ariaLabelledby && !this.popover.ariaLabel) {
                // If no aria-label is provided, link the popover to the trigger element via aria-labelledby
                this.popover.ariaLabelledby = ((_c = this.popover.triggerElement) === null || _c === void 0 ? void 0 : _c.id) || '';
            }
        }
    }
    /**
     * Updates the aria-modal attribute based on the popover's role.
     */
    updateAriaModal() {
        if (this.popover.role === ROLE.DIALOG || this.popover.role === ROLE.ALERTDIALOG) {
            this.popover.setAttribute('aria-modal', 'true');
        }
        else {
            this.popover.removeAttribute('aria-modal');
        }
    }
    /**
     * Updates the aria-expanded attribute on the trigger element.
     */
    updateAriaExpandedAttribute() {
        var _a, _b, _c;
        if (!this.popover.disableAriaExpanded) {
            (_a = this.popover.triggerElement) === null || _a === void 0 ? void 0 : _a.setAttribute('aria-expanded', `${this.popover.visible}`);
            (_b = this.popover.triggerElement) === null || _b === void 0 ? void 0 : _b.setAttribute('aria-haspopup', ((_c = this.popover.triggerElement) === null || _c === void 0 ? void 0 : _c.getAttribute('aria-haspopup')) || this.popover.role || 'dialog');
        }
    }
    /**
     * Updates the arrow style based on the arrow data and placement.
     *
     * @param arrowData - The arrow data x and y.
     * @param placement - The placement of the popover.
     */
    updateArrowStyle(arrowData, placement) {
        var _a;
        if (!this.popover.arrowElement)
            return;
        const side = placement.split('-')[0];
        const staticSide = {
            top: 'bottom',
            right: 'left',
            bottom: 'top',
            left: 'right',
        }[side];
        const { x: arrowX, y: arrowY } = arrowData;
        const rect = this.popover.arrowElement.getBoundingClientRect();
        const parent = (_a = this.popover.arrowElement.offsetParent) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect();
        if (!this.arrowPixelChange) {
            const pixelDiff = (parent === null || parent === void 0 ? void 0 : parent[staticSide]) ? 12 - Math.abs(rect[staticSide] - parent[staticSide]) : 0;
            if (Math.round(pixelDiff) === 1) {
                this.arrowPixelChange = true;
            }
            else {
                this.arrowPixelChange = false;
            }
        }
        const arrowPixelDiff = this.arrowPixelChange ? 0.5 : 0;
        this.popover.arrowElement.setAttribute('data-side', side);
        Object.assign(this.popover.arrowElement.style, {
            left: arrowX != null ? `${arrowX}px` : '',
            top: arrowY != null ? `${arrowY}px` : '',
            [staticSide]: `${-this.popover.arrowElement.offsetHeight / 2 - arrowPixelDiff}px`,
        });
    }
    /**
     * Updates the popover style based on the x and y position.
     *
     * @param x - The x position.
     * @param y - The y position.
     */
    updatePopoverStyle(x, y) {
        Object.assign(this.popover.style, {
            left: `${x}px`,
            top: `${y}px`,
        });
    }
}
