// @momentum-design/components/dist/components/popover/popover.events.js
export class PopoverEventManager {
    /**
     * Dispatches a custom event for the popover.
     *
     * @param eventName - The name of the event.
     * @param instance - The popover instance.
     */
    static dispatchPopoverEvent(eventName, instance) {
        instance.dispatchEvent(new CustomEvent(eventName, {
            detail: { show: instance.visible },
            composed: true,
            bubbles: true,
        }));
    }
    /**
     * Custom event that is fired when the popover is shown.
     *
     * @param instance - The popover instance.
     */
    static onShowPopover(instance) {
        this.dispatchPopoverEvent('shown', instance);
    }
    /**
     * Custom event that is fired when the popover is hidden.
     *
     * @param instance - The popover instance.
     */
    static onHidePopover(instance) {
        this.dispatchPopoverEvent('hidden', instance);
    }
    /**
     * Custom event that is fired when the popover is created.
     *
     * @param instance - The popover instance.
     */
    static onCreatedPopover(instance) {
        this.dispatchPopoverEvent('created', instance);
    }
    /**
     * Custom event that is fired when the popover is destroyed.
     *
     * @param instance - The popover instance.
     */
    static onDestroyedPopover(instance) {
        this.dispatchPopoverEvent('destroyed', instance);
    }
    /**
     * Custom event that is fired when the popover is closed by pressing the Escape key.
     *
     * @param instance - The popover instance.
     */
    static onEscapeKeyPressed(instance) {
        this.dispatchPopoverEvent('closebyescape', instance);
    }
    /**
     * Custom event that is fired when the popover is closed by clicking outside of it.
     *
     * @param instance - The popover instance.
     */
    static onClickOutside(instance) {
        this.dispatchPopoverEvent('closebyoutsideclick', instance);
    }
}
