// @momentum-design/components/dist/components/linkbutton/linkbutton.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { IconNameMixin } from '../../utils/mixins/IconNameMixin';
import Buttonsimple from '../buttonsimple/buttonsimple.component';
import Link from '../link/link.component';
import { DEFAULTS, LINKBUTTON_SIZES } from './linkbutton.constants';
import { getIconSize } from './linkbutton.utils';
import styles from './linkbutton.styles';
/**
 * `mdc-linkbutton` visually mimics a hyperlink while functioning as a button. It blends the appearance of `mdc-link` with the accessibility and interaction capabilities of `mdc-button`.
 *
 * ### Features:
 * - Looks like a link, behaves like a button.
 * - Supports slots for a text label and an optional trailing icon.
 * - Inherits accessibility and keyboard interaction behavior from `mdc-buttonsimple`.
 *
 * @dependency mdc-icon
 *
 * @tagname mdc-linkbutton
 *
 * @slot - Text label of the linkbutton.
 *
 * @event click - (React: onClick) This event is dispatched when the linkbutton is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the linkbutton.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the linkbutton.
 * @event focus - (React: onFocus) This event is dispatched when the linkbutton receives focus.
 *
 * @cssproperty --mdc-link-border-radius - Border radius of the linkbutton.
 * @cssproperty --mdc-link-color-active - Color of the linkbutton’s child content in the active state.
 * @cssproperty --mdc-link-color-disabled - Color of the linkbutton’s child content in the disabled state.
 * @cssproperty --mdc-link-color-hover - Color of the linkbutton’s child content in the hover state.
 * @cssproperty --mdc-link-color-normal - Color of the linkbutton’s child content in the normal state.
 * @cssproperty --mdc-link-inverted-color-active - Color of the inverted linkbutton’s child content in the active state.
 * @cssproperty --mdc-link-inverted-color-disabled - Color of the inverted linkbutton’s child content in the disabled state.
 * @cssproperty --mdc-link-inverted-color-hover - Color of the inverted linkbutton’s child content in the hover state.
 * @cssproperty --mdc-link-inverted-color-normal - Color of the inverted linkbutton’s child content in the normal state.
 */
class LinkButton extends IconNameMixin(Buttonsimple) {
    constructor() {
        super(...arguments);
        /**
         * Sets the size of the linkbutton.
         * Acceptable values:
         * - 12
         * - 14
         * - 16
         * @default 16
         */
        this.size = DEFAULTS.SIZE;
        /**
         * The linkbutton can be inline or standalone.
         * @default false
         */
        this.inline = DEFAULTS.INLINE;
        /**
         * The linkbutton color can be inverted by setting the inverted attribute to true.
         * @default false
         */
        this.inverted = DEFAULTS.INVERTED;
    }
    connectedCallback() {
        super.connectedCallback();
        this.active = undefined;
        this.role = DEFAULTS.ROLE;
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('size')) {
            this.setSize(this.size);
        }
    }
    /**
     * Sets the `size` attribute for the linkbutton, falling back to the default if the value is invalid.
     *
     * @param size - The desired link size.
     */
    setSize(size) {
        this.setAttribute('size', Object.values(LINKBUTTON_SIZES).includes(size) ? `${size}` : DEFAULTS.SIZE.toString());
    }
    render() {
        return html `
      <slot></slot>
      ${this.iconName
            ? html ` <mdc-icon name=${this.iconName} size=${getIconSize(this.size)} length-unit="rem"></mdc-icon> `
            : nothing}
    `;
    }
}
LinkButton.styles = [...Link.styles, ...styles];
__decorate([
    property({ type: Number, reflect: true }),
    __metadata("design:type", Number)
], LinkButton.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Boolean)
], LinkButton.prototype, "inline", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Boolean)
], LinkButton.prototype, "inverted", void 0);
export default LinkButton;
