// @momentum-design/components/dist/react/animation/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/animation';
import { TAG_NAME } from '../../components/animation/animation.constants';
/**
 * The `mdc-animation` component is a wrapper around the Lottie animation library.
 * It fetches the animation data dynamically based on the provided name and renders it.
 * This is a display only component that does not have any interactive functionality.
 * From accessibility perspective, (by default) it is a decorative image component.
 *
 * @tagname mdc-animation
 *
 * @event load - (React: onLoad) This event is dispatched when the animation is loaded
 * @event complete - (React: onComplete) This event is dispatched when all animation loops completed
 * @event error - (React: onError) This event is dispatched when animation loading failed
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onLoad: 'load',
        onComplete: 'complete',
        onError: 'error',
    },
    displayName: 'Animation',
});
export default reactWrapper;
