// @momentum-design/components/dist/react/accordion/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/accordion';
import { TAG_NAME } from '../../components/accordion/accordion.constants';
/**
 * An accordion contains a header and body section with a focusable heading that can be expanded or collapsed.
 *
 * The header section contains:
 * - Prefix Icon
 * - Header Text
 * - Leading Slot - Contains the leading controls of the accordion on the header section. This will be placed on the leading side, after the header text.
 * - Trailing Slot - Contains the trailing controls of the accordion on the header section. This will be placed on the trailing side, before the expand/collapse button.
 *
 * The body section contains:
 * - Default slot - User can place any content inside the body section.
 *
 * The accordion can be expanded or collapsed. The visibility of the body section can be controlled by `expanded` attribute. <br/>
 * There are two types of variants based on that the border styling of the accordion gets reflected. <br/>
 * There are two sizes of accordion, one is small and the other is large.
 * Small size has a padding of 1rem (16px) and large size has a padding of 1.5rem (24px) for the body section of accordion.
 *
 * By default, the header text in the accordion heading is of H3 with an aria-level of '3'.
 * If this accordion is placed on any other level in the entire webpage, then do adjust the aria-level number based on that.
 *
 * An accordion can be disabled, and when it's disabled, the header section will not be clickable.
 *
 * If you don't need any controls on your accordion heading, then it's advised to use `accordionbutton` component.
 *
 * If an accordion is expanded by default, then the screen reader might loose focus when toggling the visibilty of the accordion.
 *
 * @tagname mdc-accordion
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @slot leading-controls - The leading controls slot of the accordion on the header section.
 * @slot trailing-controls - The trailing controls slot of the accordion on the header section.
 * @slot default - The default slot contains the body section of the accordion. User can place anything inside this body slot.
 *
 * @event shown - (React: onShown) This event is triggered when the accordion is expanded.
 *
 * @cssproperty --mdc-accordionbutton-border-color - The border color of the accordion.
 * @cssproperty --mdc-accordionbutton-hover-color - The hover color of the accordion.
 * @cssproperty --mdc-accordionbutton-active-color - The active color of the accordion.
 * @cssproperty --mdc-accordionbutton-disabled-color - The disabled color of the accordion.
 *
 * @csspart header-section - The header section of the accordion.
 * @csspart leading-header - The leading header of the accordion.
 * @csspart trailing-header - The trailing header of the accordion.
 * @csspart trailing-header__button - The trailing header button of the accordion.
 * @csspart body-section - The body section of the accordion.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onShown: 'shown',
    },
    displayName: 'Accordion',
});
export default reactWrapper;
