// @momentum-design/components/dist/components/option/option.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { FormInternalsMixin } from '../../utils/mixins/FormInternalsMixin';
import ListItem from '../listitem/listitem.component';
import { LISTITEM_VARIANTS } from '../listitem/listitem.constants';
import { TYPE } from '../text/text.constants';
import { SELECTED_ICON_NAME } from './option.constants';
import styles from './option.styles';
/**
 * Option component, which is used within Selectlistbox within Select component.
 *
 * The label and secondary label of the option can be set using the `label` and `secondaryLabel` properties respectively.
 * The `label` is the primary text displayed in the option, while the `secondaryLabel` is the secondary text displayed below the primary label.
 *
 * The `selected` property is used to indicate whether the option is selected or not.
 * When the `selected` property is set to true, a checkmark icon will be displayed
 * on the right side of the option to indicate that it is selected.
 *
 * The `prefixIcon` property can be used to display an icon on the left side of the option label.
 * We can show a tooltip by setting `tooltip-text` attribute. This will be displayed on hover or focus of the option.
 * The tooltip will be helpful for a long label text which is truncated with ellipsis.
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-tooltip
 *
 * @tagname mdc-option
 *
 * @event click - (React: onClick) This event is dispatched when the option is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the option.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the option.
 * @event focus - (React: onFocus) This event is dispatched when the option receives focus.
 */
class Option extends FormInternalsMixin(ListItem) {
    constructor() {
        super(...arguments);
        /**
         * The selected attribute is used to indicate that the option is selected.
         */
        this.selected = false;
        /**
         * Any additional description can be provided here for screen readers.
         */
        this.ariaLabel = null;
    }
    connectedCallback() {
        super.connectedCallback();
        this.role = 'option';
        this.variant = LISTITEM_VARIANTS.INSET_RECTANGLE;
        this.setAttribute('aria-selected', `${this.selected}`);
        this.setAttribute('aria-disabled', `${!!this.disabled}`);
        // Option will not contain below fields
        this.name = undefined;
        this.sideHeaderText = undefined;
        this.sublineText = undefined;
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('selected')) {
            this.setAttribute('aria-selected', `${this.selected}`);
        }
    }
    render() {
        const prefixIconContent = this.prefixIcon
            ? html `
          <div part="leading-icon">
            <mdc-icon length-unit="rem" slot="leading-controls" name="${ifDefined(this.prefixIcon)}"></mdc-icon>
          </div>
        `
            : nothing;
        const selectedIcon = this.selected
            ? html ` <mdc-icon length-unit="rem" slot="trailing-controls" name="${SELECTED_ICON_NAME}"></mdc-icon> `
            : nothing;
        return html `
      ${prefixIconContent}
      <div part="leading-text">
        ${this.getText('leading-text-primary-label', TYPE.BODY_MIDSIZE_REGULAR, this.label)}
        ${this.getText('leading-text-secondary-label', TYPE.BODY_SMALL_REGULAR, this.secondaryLabel)}
      </div>
      <div part="trailing">${selectedIcon}</div>
    `;
    }
}
Option.styles = [...ListItem.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], Option.prototype, "selected", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'prefix-icon' }),
    __metadata("design:type", String)
], Option.prototype, "prefixIcon", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'aria-label' }),
    __metadata("design:type", Object)
], Option.prototype, "ariaLabel", void 0);
export default Option;
