// @momentum-design/components/dist/components/alertchip/alertchip.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { Component } from '../../models';
import Buttonsimple from '../buttonsimple/buttonsimple.component';
import { IconNameMixin } from '../../utils/mixins/IconNameMixin';
import styles from './alertchip.styles';
import { DEFAULTS } from './alertchip.constants';
import { getAlertIcon } from './alertchip.utils';
/**
 * mdc-alertchip component is an interactive chip that consumers can use to represent an alert.
 *
 * - It supports a leading icon along with label.
 * - It supports 5 variants of alerts - error, informational, neutral, success and warning.
 *
 * This component is built by extending Buttonsimple.
 *
 * @tagname mdc-alertchip
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @cssproperty --mdc-chip-color - The color of the label text
 * @cssproperty --mdc-chip-icon-color - The color of the icon
 * @cssproperty --mdc-chip-border-color - The border color of the alertchip
 * @cssproperty --mdc-chip-background-color - The background color of the alertchip
 *
 * @csspart icon - The alert icon
 * @csspart label - The text label of the alertchip
 *
 * @event click - (React: onClick) This event is dispatched when the alertchip is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the alertchip.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the alertchip.
 * @event focus - (React: onFocus) This event is dispatched when the alertchip receives focus.
 */
class AlertChip extends IconNameMixin(Buttonsimple) {
    constructor() {
        super(...arguments);
        /**
         * The variant of the alertchip. It supports 5 variants
         * - neutral
         * - error
         * - success
         * - warning
         * - informational
         *
         * @default neutral
         */
        this.variant = DEFAULTS.VARIANT;
    }
    connectedCallback() {
        super.connectedCallback();
        this.disabled = undefined;
        this.softDisabled = undefined;
        this.active = undefined;
        this.size = DEFAULTS.SIZE;
        this.role = DEFAULTS.ROLE;
    }
    render() {
        return html `
      <mdc-icon part="icon" name="${this.iconName || getAlertIcon(this.variant)}" length-unit="rem" size="1"></mdc-icon>
      ${this.label
            ? html `<mdc-text part="label" type="${DEFAULTS.TEXT_TYPE}" tagname="${DEFAULTS.TAG_NAME}"
            >${this.label}</mdc-text
          >`
            : nothing}
    `;
    }
}
AlertChip.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], AlertChip.prototype, "variant", void 0);
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], AlertChip.prototype, "label", void 0);
export default AlertChip;
