// @momentum-design/components/dist/components/textarea/textarea.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { property, query } from 'lit/decorators.js';
import FormfieldWrapper from '../formfieldwrapper';
import { DEFAULTS as FORMFIELD_DEFAULTS, VALIDATION } from '../formfieldwrapper/formfieldwrapper.constants';
import { AUTO_CAPITALIZE } from '../input/input.constants';
import { DataAriaLabelMixin } from '../../utils/mixins/DataAriaLabelMixin';
import { FormInternalsMixin } from '../../utils/mixins/FormInternalsMixin';
import { AUTO_COMPLETE, WRAP, DEFAULTS } from './textarea.constants';
import styles from './textarea.styles';
/**
 * mdc-textarea component, which is used to get the multi-line text input from the user.
 * It contains:
 * - label: It is the title of the textarea field.
 * - required: A boolean attribute depicting that the textarea field is required.
 * - Textarea: It is the multi-line text input field.
 * - helper-text: It is the text that provides additional information about the textarea field.
 * - max-character-limit: It is the text that shows the character count of the textarea field.
 * - Error, Warning, Success, Priority Help Text type: It is the text that provides additional information
 *   about the textarea field based on the validation state.
 * - limitexceeded: It is the event that is dispatched when the character limit exceeds or restored.
 *   This event exposes 3 properties:
 *   - currentCharacterCount - the current number of characters in the textarea field,
 *   - maxCharacterLimit - the maximum number of characters allowed in the textarea field,
 *   - value - the current value of the textarea field,
 *
 * **Note**: Consumers must set the help-text-type with 'error' and
 * help-text attribute with the error message using limitexceeded event.
 * The same help-text value will be used for the validation message to be displayed.
 *
 * @tagname mdc-textarea
 *
 * @event input - (React: onInput) This event is dispatched when the value of the textarea field changes (every press).
 * @event change - (React: onChange) This event is dispatched when the value of the textarea field changes (on blur).
 * @event focus - (React: onFocus) This event is dispatched when the textarea receives focus.
 * @event blur - (React: onBlur) This event is dispatched when the textarea loses focus.
 * @event limitexceeded - (React: onLimitExceeded) This event is dispatched once when the character limit
 * exceeds or restored.
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @cssproperty --mdc-textarea-disabled-border-color - Border color for the textarea container when disabled
 * @cssproperty --mdc-textarea-disabled-text-color - Text color for the textarea field when disabled
 * @cssproperty --mdc-textarea-disabled-background-color - Background color for the textarea field when disabled
 * @cssproperty --mdc-textarea-text-color - Text color for the textarea field
 * @cssproperty --mdc-textarea-background-color - Background color for the textarea field
 * @cssproperty --mdc-textarea-border-color - Border color for the textarea field
 * @cssproperty --mdc-textarea-text-secondary-normal - Text color for the character counter
 * @cssproperty --mdc-textarea-error-border-color - Border color for the error related help text
 * @cssproperty --mdc-textarea-warning-border-color - Border color for the warning related help text
 * @cssproperty --mdc-textarea-success-border-color - Border color for the success related help text
 * @cssproperty --mdc-textarea-primary-border-color - Border color for the priority related help text
 * @cssproperty --mdc-textarea-hover-background-color - Background color for the textarea container when hover
 * @cssproperty --mdc-textarea-focused-background-color - Background color for the textarea container when focused
 * @cssproperty --mdc-textarea-focused-border-color - Border color for the textarea container when focused
 */
class Textarea extends FormInternalsMixin(DataAriaLabelMixin(FormfieldWrapper)) {
    constructor() {
        super(...arguments);
        /**
         * readonly attribute of the textarea field. If true, the textarea field is read-only.
         * @default false
         */
        this.readonly = false;
        /**
         * The rows attribute specifies the visible number of lines in a text area.
         * @default 5
         */
        this.rows = DEFAULTS.ROWS;
        /**
         * The cols attribute specifies the visible number of lines in a text area.
         * @default 40
         */
        this.cols = DEFAULTS.COLS;
        /**
         * The wrap attribute specifies how the text in a text area is to be wrapped when submitted in a form.
         * @default 'soft'
         */
        this.wrap = WRAP.SOFT;
        /**
         * The autocapitalize attribute of the textarea field.
         * @default 'off'
         */
        this.autocapitalize = AUTO_CAPITALIZE.OFF;
        /**
         * The autocomplete attribute of the textarea field.
         * @default 'off'
         */
        this.autocomplete = AUTO_COMPLETE.OFF;
        /**
         * If true, the textarea field is focused when the component is rendered.
         * @default false
         */
        this.autofocus = false;
        this.characterLimitExceedingFired = false;
    }
    get textarea() {
        return this.inputElement;
    }
    connectedCallback() {
        var _a;
        super.connectedCallback();
        // Set the default value to the textarea field if the value is set through the text content directly
        this.value = ((_a = this.textContent) === null || _a === void 0 ? void 0 : _a.trim()) || this.value;
        this.updateComplete
            .then(() => {
            if (this.textarea) {
                this.textarea.checkValidity();
                this.setTextareaValidity();
                this.internals.setFormValue(this.textarea.value);
            }
        })
            .catch(error => {
            if (this.onerror) {
                this.onerror(error);
            }
        });
    }
    setTextareaValidity() {
        if (this.required && this.validationMessage && this.value === '') {
            this.textarea.setCustomValidity(this.validationMessage);
        }
        else if (this.maxCharacterLimit &&
            this.value.length > this.maxCharacterLimit &&
            this.helpTextType === VALIDATION.ERROR &&
            this.helpText) {
            // Set custom validity if the character limit is exceeded to stop form submission
            // helptext and helptexttype will be set by the consumers.
            this.textarea.setCustomValidity(this.helpText);
        }
        else {
            this.textarea.setCustomValidity('');
        }
        this.setValidity();
    }
    /** @internal */
    formResetCallback() {
        this.value = '';
        this.requestUpdate();
    }
    /** @internal */
    formStateRestoreCallback(state) {
        this.value = state;
    }
    /**
     * Handles the value change of the textarea field.
     * Sets the form value and updates the validity of the textarea field.
     * @returns void
     */
    handleValueChange() {
        this.updateComplete
            .then(() => {
            this.setTextareaValidity();
        })
            .catch(error => {
            if (this.onerror) {
                this.onerror(error);
            }
        });
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('value')) {
            this.handleValueChange();
            this.handleCharacterOverflowStateChange();
        }
    }
    /**
     * This function is called when the attribute changes.
     * It updates the validity of the textarea field based on the textarea field's validity.
     *
     * @param name - attribute name
     * @param old - old value
     * @param value - new value
     */
    attributeChangedCallback(name, old, value) {
        super.attributeChangedCallback(name, old, value);
        const validationRelatedAttributes = ['maxlength', 'minlength', 'required'];
        if (validationRelatedAttributes.includes(name)) {
            this.updateComplete
                .then(() => {
                this.setTextareaValidity();
            })
                .catch(error => {
                if (this.onerror) {
                    this.onerror(error);
                }
            });
        }
    }
    /**
     * Dispatches the character overflow state change event.
     * @returns void
     */
    dispatchCharacterOverflowStateChangeEvent() {
        this.dispatchEvent(new CustomEvent('limitexceeded', {
            detail: {
                currentCharacterCount: this.value.length,
                maxCharacterLimit: this.maxCharacterLimit,
                value: this.value,
            },
            bubbles: true,
            composed: true,
        }));
    }
    /**
     * Handles the character overflow state change.
     * Dispatches the character overflow state change event if the character limit is exceeded or restored.
     * @returns void
     */
    handleCharacterOverflowStateChange() {
        if (this.maxCharacterLimit) {
            if (this.value.length > this.maxCharacterLimit && !this.characterLimitExceedingFired) {
                this.dispatchCharacterOverflowStateChangeEvent();
                this.characterLimitExceedingFired = true;
            }
            else if (this.value.length <= this.maxCharacterLimit && this.characterLimitExceedingFired) {
                this.dispatchCharacterOverflowStateChangeEvent();
                this.characterLimitExceedingFired = false;
            }
        }
    }
    /**
     * Updates the value of the textarea field.
     * Sets the form value.
     * @returns void
     */
    updateValue() {
        this.value = this.textarea.value;
        this.internals.setFormValue(this.textarea.value);
    }
    /**
     * Handles the change event of the textarea field.
     * Updates the value and sets the validity of the textarea field.
     *
     * The 'change' event does not bubble up through the shadow DOM as it was not composed.
     * Therefore, we need to re-dispatch the same event to ensure it is propagated correctly.
     * Read more: https://developer.mozilla.org/en-US/docs/Web/API/Event/composed
     *
     * @param event - Event which contains information about the value change.
     */
    onChange(event) {
        this.updateValue();
        const EventConstructor = event.constructor;
        this.dispatchEvent(new EventConstructor(event.type, event));
    }
    renderCharacterCounter() {
        if (!this.maxCharacterLimit) {
            return nothing;
        }
        return html `
      <mdc-text part="character-counter" tagname="span" type=${DEFAULTS.CHARACTER_COUNTER_TYPE}>
        ${this.value.length < 10 ? `0${this.value.length}` : this.value.length}/${this.maxCharacterLimit}
      </mdc-text>
    `;
    }
    renderTextareaFooter() {
        if (!this.helpText && !this.maxCharacterLimit) {
            return nothing;
        }
        return html ` <div part="textarea-footer">${this.renderHelperText()} ${this.renderCharacterCounter()}</div> `;
    }
    render() {
        var _a;
        return html `
      ${this.renderLabel()}
      <div class="textarea-container mdc-focus-ring" part="textarea-container">
        <textarea
          aria-label="${(_a = this.dataAriaLabel) !== null && _a !== void 0 ? _a : ''}"
          part="textarea"
          id="${this.inputId}"
          name="${this.name}"
          .value="${this.value}"
          ?disabled="${this.disabled}"
          ?readonly="${this.readonly}"
          ?required="${this.required}"
          placeholder=${ifDefined(this.placeholder)}
          rows=${ifDefined(this.rows)}
          cols=${ifDefined(this.cols)}
          wrap=${ifDefined(this.wrap)}
          ?autofocus="${this.autofocus}"
          autocapitalize=${this.autocapitalize}
          autocomplete=${this.autocomplete}
          minlength=${ifDefined(this.minlength)}
          maxlength=${ifDefined(this.maxlength)}
          dirname=${ifDefined(this.dirname)}
          @input=${this.updateValue}
          @change=${this.onChange}
          aria-describedby="${ifDefined(this.helpText ? FORMFIELD_DEFAULTS.HELPER_TEXT_ID : '')}"
          aria-invalid="${this.helpTextType === 'error' ? 'true' : 'false'}"
        ></textarea>
      </div>
      ${this.renderTextareaFooter()}
    `;
    }
}
Textarea.styles = [...FormfieldWrapper.styles, ...styles];
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], Textarea.prototype, "placeholder", void 0);
__decorate([
    property({ type: Boolean }),
    __metadata("design:type", Object)
], Textarea.prototype, "readonly", void 0);
__decorate([
    property({ type: Number }),
    __metadata("design:type", Number)
], Textarea.prototype, "rows", void 0);
__decorate([
    property({ type: Number }),
    __metadata("design:type", Number)
], Textarea.prototype, "cols", void 0);
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], Textarea.prototype, "wrap", void 0);
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], Textarea.prototype, "autocapitalize", void 0);
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], Textarea.prototype, "autocomplete", void 0);
__decorate([
    property({ type: Boolean }),
    __metadata("design:type", Boolean)
], Textarea.prototype, "autofocus", void 0);
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], Textarea.prototype, "dirname", void 0);
__decorate([
    property({ type: Number }),
    __metadata("design:type", Number)
], Textarea.prototype, "maxlength", void 0);
__decorate([
    property({ type: Number }),
    __metadata("design:type", Number)
], Textarea.prototype, "minlength", void 0);
__decorate([
    property({ type: Number, attribute: 'max-character-limit' }),
    __metadata("design:type", Number)
], Textarea.prototype, "maxCharacterLimit", void 0);
__decorate([
    query('textarea'),
    __metadata("design:type", HTMLTextAreaElement)
], Textarea.prototype, "inputElement", void 0);
export default Textarea;
