// @momentum-design/components/dist/react/alertchip/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/alertchip';
import { TAG_NAME } from '../../components/alertchip/alertchip.constants';
/**
 * mdc-alertchip component is an interactive chip that consumers can use to represent an alert.
 *
 * - It supports a leading icon along with label.
 * - It supports 5 variants of alerts - error, informational, neutral, success and warning.
 *
 * This component is built by extending Buttonsimple.
 *
 * @tagname mdc-alertchip
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @cssproperty --mdc-chip-color - The color of the label text
 * @cssproperty --mdc-chip-icon-color - The color of the icon
 * @cssproperty --mdc-chip-border-color - The border color of the alertchip
 * @cssproperty --mdc-chip-background-color - The background color of the alertchip
 *
 * @csspart icon - The alert icon
 * @csspart label - The text label of the alertchip
 *
 * @event click - (React: onClick) This event is dispatched when the alertchip is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the alertchip.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the alertchip.
 * @event focus - (React: onFocus) This event is dispatched when the alertchip receives focus.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onKeyUp: 'keyup',
        onFocus: 'focus',
    },
    displayName: 'AlertChip',
});
export default reactWrapper;
