// @momentum-design/components/dist/components/datepicker/datepicker.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property, query, state } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { DataAriaLabelMixin } from '../../utils/mixins/DataAriaLabelMixin';
import { FormInternalsMixin } from '../../utils/mixins/FormInternalsMixin';
import { parseISO, isBefore, isAfter } from '../../utils/date-utils';
import { KEYS } from '../../utils/keys';
import FormfieldWrapper from '../formfieldwrapper/formfieldwrapper.component';
import { POPOVER_PLACEMENT, TRIGGER, DEFAULTS as POPOVER_DEFAULTS } from '../popover/popover.constants';
import { SELECTION_MODE } from '../calendar/calendar.constants';
import { formatDateForDisplay, formatDateRangeForDisplay, getWeekRange } from '../calendar/calendar.utils';
import { CALENDAR_ICON, DEFAULTS, TRIGGER_ID, VARIANT } from './datepicker.constants';
import styles from './datepicker.styles';
import { buildISODate, getDateSeparator, getFieldMax, getFieldMin, getFieldOrder, getPlaceholder, parseISODate, } from './datepicker.utils';
/**
 * mdc-datepicker is a component that allows users to select a date
 * or enter a date manually. It supports two variants:
 *
 * - `input`: Manual date entry with spinbutton inputs for month, day, and year
 * - `default`: A select-like trigger that opens a calendar for date range selection
 *
 * The component uses `mdc-calendar` inside a popover for date selection
 * and supports localization via a BCP 47 locale string.
 *
 * @tagname mdc-datepicker
 *
 * @dependency mdc-button
 * @dependency mdc-calendar
 * @dependency mdc-icon
 * @dependency mdc-popover
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @event input - (React: onInput) This event is dispatched when the date value changes.
 * @event change - (React: onChange) This event is dispatched when the date value is committed.
 * @event focus - (React: onFocus) This event is dispatched when the datepicker receives focus.
 * @event blur - (React: onBlur) This event is dispatched when the datepicker loses focus.
 *
 * @slot label - Slot for the label element.
 * @slot toggletip - Slot for the toggletip info icon button.
 * @slot help-icon - Slot for the helper/validation icon.
 * @slot help-text - Slot for the helper/validation text.
 *
 * @cssproperty --mdc-datepicker-background-color - Background color of the datepicker input.
 * @cssproperty --mdc-datepicker-border-color - Border color of the datepicker input.
 * @cssproperty --mdc-datepicker-text-color - Text color of the datepicker input.
 * @cssproperty --mdc-datepicker-height - The height of the datepicker trigger control.
 * @cssproperty --mdc-datepicker-width - Width of the datepicker component.
 *
 * @csspart label - The label element.
 * @csspart label-text - The container for the label and required indicator elements.
 * @csspart required-indicator - The required indicator element.
 * @csspart info-icon-btn - The info icon button element.
 * @csspart label-toggletip - The toggletip element.
 * @csspart help-text - The helper/validation text element.
 * @csspart helper-icon - The helper/validation icon element.
 * @csspart help-text-container - The container for helper/validation elements.
 * @csspart container - The outer container for the input and popover.
 * @csspart base-container - The input container with border and background.
 * @csspart spinbutton-group - The container for spinbutton elements.
 * @csspart spinbutton - A spinbutton input element.
 * @csspart separator - The separator between spinbuttons.
 * @csspart icon-container - The calendar/dropdown button container.
 * @csspart native-input - The hidden native input for form participation.
 * @csspart select-text - The selected value display text (default variant).
 */
class DatePicker extends FormInternalsMixin(DataAriaLabelMixin(FormfieldWrapper)) {
    constructor() {
        super(...arguments);
        /**
         * The variant of the date picker.
         * - `'input'`: Manual date entry with spinbutton inputs
         * - `'default'`: Select-like trigger for week selection
         * @default 'input'
         */
        this.variant = DEFAULTS.VARIANT;
        /**
         * The selected date value as an ISO string (yyyy-mm-dd).
         */
        this.value = '';
        /**
         * The selection mode for the calendar.
         * @default 'single'
         */
        this.selectionMode = SELECTION_MODE.SINGLE;
        /**
         * BCP 47 locale string for date formatting.
         * @default 'en-US'
         */
        this.locale = DEFAULTS.LOCALE;
        /**
         * The placement of the popover dropdown.
         * @default 'bottom-start'
         */
        this.placement = POPOVER_PLACEMENT.BOTTOM_START;
        /**
         * The strategy for positioning the popover.
         * @default 'absolute'
         */
        this.strategy = POPOVER_DEFAULTS.STRATEGY;
        /**
         * Determines whether the dropdown should flip its position at the boundary.
         * @default false
         */
        this.disableFlip = DEFAULTS.DISABLE_FLIP;
        /**
         * Accessible label for the month spinbutton.
         */
        this.localeMonthLabel = '';
        /**
         * Accessible label for the day spinbutton.
         */
        this.localeDayLabel = '';
        /**
         * Accessible label for the year spinbutton.
         */
        this.localeYearLabel = '';
        /**
         * Accessible label for the calendar toggle button.
         */
        this.localeCalendarLabel = '';
        /**
         * Accessible description for spinbutton inputs.
         */
        this.localeSpinbuttonDescription = '';
        /**
         * Localized label for the today button on the calendar.
         */
        this.localeTodayLabel = '';
        /**
         * Localized aria-label for the previous month button.
         */
        this.localePrevMonthLabel = '';
        /**
         * Localized aria-label for the next month button.
         */
        this.localeNextMonthLabel = '';
        /** @internal */
        this.displayPopover = false;
        /** @internal */
        this.internalMonth = '';
        /** @internal */
        this.internalDay = '';
        /** @internal */
        this.internalYear = '';
        /**
         * The end date value for range/week selection (ISO string).
         */
        this.endValue = '';
        /** @internal */
        this.pendingDigits = '';
    }
    connectedCallback() {
        super.connectedCallback();
        this.updateComplete
            .then(() => {
            this.parseValueToInternal();
            this.syncFormValue();
        })
            .catch(error => {
            if (this.onerror) {
                this.onerror(error);
            }
        });
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        if (this.pendingDigitTimeout) {
            clearTimeout(this.pendingDigitTimeout);
        }
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('value') && !this.displayPopover) {
            this.parseValueToInternal();
            this.syncFormValue();
        }
        else if (changedProperties.has('endValue') && !this.displayPopover) {
            this.syncFormValue();
        }
    }
    parseValueToInternal() {
        const parts = parseISODate(this.value);
        if (parts) {
            this.internalMonth = parts.month;
            this.internalDay = parts.day;
            this.internalYear = parts.year;
        }
        else {
            this.internalMonth = '';
            this.internalDay = '';
            this.internalYear = '';
        }
    }
    internalToValue() {
        return buildISODate(this.internalMonth, this.internalDay, this.internalYear);
    }
    syncFormValue() {
        const val = this.value || this.internalToValue();
        if (this.internals) {
            if (this.name && this.endValue) {
                const formData = new FormData();
                formData.append(this.name, val || '');
                formData.append(`${this.name}-end`, this.endValue);
                this.internals.setFormValue(formData);
            }
            else {
                this.internals.setFormValue(val || null);
            }
        }
    }
    commitValue() {
        const newVal = this.internalToValue();
        if (newVal && newVal !== this.value) {
            const dt = parseISO(newVal);
            if (dt) {
                const minDt = this.min ? parseISO(this.min) : undefined;
                const maxDt = this.max ? parseISO(this.max) : undefined;
                if (minDt && isBefore(dt, minDt))
                    return;
                if (maxDt && isAfter(dt, maxDt))
                    return;
            }
            this.value = newVal;
            this.syncFormValue();
            this.notifyValueChange();
        }
    }
    notifyValueChange() {
        const detail = { value: this.value, endValue: this.endValue };
        this.dispatchEvent(new CustomEvent('input', { detail, bubbles: true, composed: true }));
        this.dispatchEvent(new CustomEvent('change', { detail, bubbles: true, composed: true }));
    }
    flushPendingInput() {
        if (this.pendingDigitTimeout) {
            clearTimeout(this.pendingDigitTimeout);
            this.pendingDigitTimeout = undefined;
        }
        this.pendingDigits = '';
        this.commitValue();
    }
    formResetCallback() {
        this.value = '';
        this.endValue = '';
        this.parseValueToInternal();
        this.syncFormValue();
    }
    formStateRestoreCallback(state) {
        if (state instanceof FormData) {
            const name = this.name || '';
            this.value = state.get(name) || '';
            this.endValue = state.get(`${name}-end`) || '';
        }
        else if (typeof state === 'string') {
            this.value = state;
            this.endValue = '';
        }
        this.parseValueToInternal();
        this.syncFormValue();
    }
    // -- Popover handling --
    handleCalendarButtonClick() {
        if (this.disabled || this.readonly)
            return;
        if (!this.displayPopover) {
            this.flushPendingInput();
        }
        this.displayPopover = !this.displayPopover;
    }
    handleSelectTriggerClick() {
        if (this.disabled || this.readonly)
            return;
        this.displayPopover = !this.displayPopover;
    }
    handleSelectKeydown(event) {
        if (this.disabled || this.readonly)
            return;
        if (event.key === KEYS.ENTER || event.key === KEYS.SPACE || event.key === KEYS.ARROW_DOWN) {
            event.preventDefault();
            this.displayPopover = true;
        }
    }
    focusCalendarGrid() {
        // The popover's focus trap uses setTimeout(0) to call setInitialFocus after the 'shown' event.
        // We use setTimeout to run after that, so our focus override takes effect.
        setTimeout(() => {
            var _a;
            const calendar = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('mdc-calendar');
            if (calendar === null || calendar === void 0 ? void 0 : calendar.shadowRoot) {
                const focusedCell = calendar.shadowRoot.querySelector('.calendar-day[tabindex="0"]');
                focusedCell === null || focusedCell === void 0 ? void 0 : focusedCell.focus();
            }
        }, 0);
    }
    handleDateSelected(event) {
        const { detail } = event;
        const mode = this.effectiveSelectionMode;
        if (mode === SELECTION_MODE.WEEK || mode === SELECTION_MODE.RANGE) {
            if (detail.startDate && detail.endDate) {
                this.value = detail.startDate;
                this.endValue = detail.endDate;
            }
            else {
                this.value = detail.date;
                this.endValue = '';
            }
        }
        else {
            this.value = detail.date;
            this.endValue = '';
        }
        this.parseValueToInternal();
        this.syncFormValue();
        this.notifyValueChange();
        const rangeComplete = mode === SELECTION_MODE.RANGE && this.value && this.endValue;
        if (mode !== SELECTION_MODE.RANGE || rangeComplete) {
            this.displayPopover = false;
            this.focusBackToTrigger();
        }
    }
    // -- Spinbutton handling (input variant) --
    getSpinbuttonInput(field) {
        switch (field) {
            case 'month':
                return this.monthInput;
            case 'day':
                return this.dayInput;
            case 'year':
                return this.yearInput;
            default:
                return null;
        }
    }
    getFieldValue(field) {
        switch (field) {
            case 'month':
                return this.internalMonth;
            case 'day':
                return this.internalDay;
            case 'year':
                return this.internalYear;
            default:
                return '';
        }
    }
    setFieldValue(field, val) {
        const maxDigits = field === 'year' ? 4 : 2;
        const padded = String(val).padStart(maxDigits, '0');
        switch (field) {
            case 'month':
                this.internalMonth = padded;
                break;
            case 'day':
                this.internalDay = padded;
                break;
            case 'year':
                this.internalYear = padded;
                break;
            default:
                break;
        }
        this.requestUpdate();
    }
    handleSpinbuttonKeydown(event, field) {
        if (this.readonly)
            return;
        const currentMonth = parseInt(this.internalMonth, 10) || undefined;
        const currentYear = parseInt(this.internalYear, 10) || undefined;
        const minVal = getFieldMin(field);
        const maxVal = getFieldMax(field, currentMonth, currentYear);
        const currentVal = parseInt(this.getFieldValue(field), 10) || 0;
        switch (event.key) {
            case KEYS.ARROW_UP: {
                event.preventDefault();
                const newVal = currentVal >= maxVal ? minVal : currentVal + 1;
                this.setFieldValue(field, newVal);
                this.commitValue();
                break;
            }
            case KEYS.ARROW_DOWN: {
                event.preventDefault();
                const newVal = currentVal <= minVal ? maxVal : currentVal - 1;
                this.setFieldValue(field, newVal);
                this.commitValue();
                break;
            }
            case KEYS.ARROW_LEFT: {
                event.preventDefault();
                this.advanceToPrevField(field);
                break;
            }
            case KEYS.ARROW_RIGHT: {
                event.preventDefault();
                this.advanceToNextField(field);
                break;
            }
            case KEYS.TAB:
                break;
            default: {
                if (/^\d$/.test(event.key)) {
                    event.preventDefault();
                    this.handleDigitInput(event.key, field, minVal, maxVal);
                }
                else {
                    event.preventDefault();
                }
                break;
            }
        }
    }
    handleDigitInput(digit, field, minVal, maxVal) {
        if (this.pendingDigitTimeout) {
            clearTimeout(this.pendingDigitTimeout);
        }
        this.pendingDigits += digit;
        const maxDigits = field === 'year' ? 4 : 2;
        if (this.pendingDigits.length >= maxDigits) {
            let val = parseInt(this.pendingDigits, 10);
            if (val > maxVal)
                val = maxVal;
            if (val < minVal)
                val = minVal;
            this.setFieldValue(field, val);
            this.pendingDigits = '';
            this.commitValue();
            this.advanceToNextField(field);
        }
        else if (field !== 'year') {
            const firstDigit = parseInt(this.pendingDigits, 10);
            const maxFirstDigit = Math.floor(maxVal / 10);
            if (firstDigit > maxFirstDigit) {
                let val = firstDigit;
                if (val > maxVal)
                    val = maxVal;
                if (val < minVal)
                    val = minVal;
                this.setFieldValue(field, val);
                this.pendingDigits = '';
                this.commitValue();
                this.advanceToNextField(field);
            }
            else {
                this.setFieldValue(field, firstDigit);
                this.pendingDigitTimeout = setTimeout(() => {
                    let val = firstDigit;
                    if (val < minVal)
                        val = minVal;
                    this.setFieldValue(field, val);
                    this.pendingDigits = '';
                    this.commitValue();
                    this.advanceToNextField(field);
                }, 1000);
            }
        }
        else {
            const partialVal = parseInt(this.pendingDigits, 10);
            this.setFieldValue(field, partialVal);
            this.pendingDigitTimeout = setTimeout(() => {
                this.pendingDigits = '';
                this.commitValue();
            }, 1000);
        }
    }
    advanceToNextField(currentField) {
        const order = getFieldOrder(this.locale);
        const idx = order.indexOf(currentField);
        if (idx < order.length - 1) {
            const nextField = order[idx + 1];
            this.updateComplete
                .then(() => {
                const input = this.getSpinbuttonInput(nextField);
                input === null || input === void 0 ? void 0 : input.focus();
                input === null || input === void 0 ? void 0 : input.select();
            })
                .catch(() => { });
        }
    }
    advanceToPrevField(currentField) {
        const order = getFieldOrder(this.locale);
        const idx = order.indexOf(currentField);
        if (idx > 0) {
            const prevField = order[idx - 1];
            this.updateComplete
                .then(() => {
                const input = this.getSpinbuttonInput(prevField);
                input === null || input === void 0 ? void 0 : input.focus();
                input === null || input === void 0 ? void 0 : input.select();
            })
                .catch(() => { });
        }
    }
    handleSpinbuttonFocus(event) {
        const target = event.target;
        target.select();
        this.pendingDigits = '';
        if (this.pendingDigitTimeout) {
            clearTimeout(this.pendingDigitTimeout);
        }
    }
    handleSpinbuttonAreaClick(event) {
        const target = event.target;
        if (target.getAttribute('role') !== 'spinbutton') {
            const order = getFieldOrder(this.locale);
            const firstInput = this.getSpinbuttonInput(order[0]);
            firstInput === null || firstInput === void 0 ? void 0 : firstInput.focus();
            firstInput === null || firstInput === void 0 ? void 0 : firstInput.select();
        }
    }
    handleBaseKeydown(event) {
        if (event.key === KEYS.ESCAPE && this.displayPopover) {
            this.displayPopover = false;
        }
    }
    focusBackToTrigger() {
        this.updateComplete
            .then(() => {
            var _a, _b;
            if (this.variant === VARIANT.INPUT) {
                const calendarBtn = (_a = this.shadowRoot) === null || _a === void 0 ? void 0 : _a.querySelector('[part~="icon-container"]');
                calendarBtn === null || calendarBtn === void 0 ? void 0 : calendarBtn.focus();
            }
            else {
                const trigger = (_b = this.shadowRoot) === null || _b === void 0 ? void 0 : _b.getElementById(TRIGGER_ID);
                trigger === null || trigger === void 0 ? void 0 : trigger.focus();
            }
        })
            .catch(() => { });
    }
    // -- Display text --
    /** @internal */
    get effectiveSelectionMode() {
        if (this.selectionMode === SELECTION_MODE.WEEK) {
            return SELECTION_MODE.WEEK;
        }
        return this.variant === VARIANT.DEFAULT ? SELECTION_MODE.RANGE : this.selectionMode;
    }
    getDisplayText() {
        if (!this.value)
            return '';
        if (this.effectiveSelectionMode === SELECTION_MODE.WEEK) {
            const { start, end } = getWeekRange(this.value, this.locale);
            return formatDateRangeForDisplay(start, end, this.locale);
        }
        if (this.effectiveSelectionMode === SELECTION_MODE.RANGE && this.endValue) {
            return formatDateRangeForDisplay(this.value, this.endValue, this.locale);
        }
        return formatDateForDisplay(this.value, this.locale);
    }
    // -- Render methods --
    renderSpinbutton(field) {
        const value = this.getFieldValue(field);
        const currentMonth = parseInt(this.internalMonth, 10) || undefined;
        const currentYear = parseInt(this.internalYear, 10) || undefined;
        const minVal = getFieldMin(field);
        const maxVal = getFieldMax(field, currentMonth, currentYear);
        const labelMap = {
            month: this.localeMonthLabel,
            day: this.localeDayLabel,
            year: this.localeYearLabel,
        };
        const placeholderMap = {
            month: 'mm',
            day: 'dd',
            year: 'yyyy',
        };
        return html `
      <input
        id="${field}-spinbutton"
        part="spinbutton spinbutton-${field}"
        role="spinbutton"
        aria-label="${labelMap[field]}"
        aria-valuemin="${minVal}"
        aria-valuemax="${maxVal}"
        aria-valuenow="${ifDefined(value ? parseInt(value, 10) : undefined)}"
        aria-description="${this.localeSpinbuttonDescription}"
        .value="${value}"
        placeholder="${placeholderMap[field]}"
        autocomplete="off"
        ?disabled="${this.disabled}"
        ?readonly="${this.readonly}"
        tabindex="${this.disabled ? '-1' : '0'}"
        @keydown="${(e) => this.handleSpinbuttonKeydown(e, field)}"
        @focus="${this.handleSpinbuttonFocus}"
      />
    `;
    }
    renderInputVariant() {
        const fieldOrder = getFieldOrder(this.locale);
        const separator = getDateSeparator(this.locale);
        return html `
      <div
        id="${TRIGGER_ID}"
        part="base-container"
        class="mdc-focus-ring"
        @click="${this.handleSpinbuttonAreaClick}"
        @keydown="${this.handleBaseKeydown}"
      >
        <div part="spinbutton-group">
          ${fieldOrder.map((field, index) => html `
              ${this.renderSpinbutton(field)}
              ${index < fieldOrder.length - 1 ? html `<span part="separator">${separator}</span>` : nothing}
            `)}
        </div>
        <mdc-button
          part="icon-container"
          class="own-focus-ring"
          variant="tertiary"
          prefix-icon="${CALENDAR_ICON}"
          aria-label="${this.localeCalendarLabel}"
          aria-expanded="${this.displayPopover ? 'true' : 'false'}"
          aria-haspopup="dialog"
          ?disabled="${this.disabled}"
          size="20"
          @click="${this.handleCalendarButtonClick}"
        ></mdc-button>
      </div>
    `;
    }
    renderDefaultVariant() {
        const displayText = this.getDisplayText();
        const placeholderText = getPlaceholder(this.locale);
        return html `
      <div
        id="${TRIGGER_ID}"
        part="base-container"
        class="mdc-focus-ring"
        role="combobox"
        aria-expanded="${this.displayPopover ? 'true' : 'false'}"
        aria-haspopup="dialog"
        aria-label="${this.localeCalendarLabel}"
        tabindex="${this.disabled ? '-1' : '0'}"
        @click="${this.handleSelectTriggerClick}"
        @keydown="${this.handleSelectKeydown}"
      >
        ${displayText
            ? html `<span part="select-text">${displayText}</span>`
            : html `<span part="select-text select-placeholder">${placeholderText}</span>`}
        <mdc-icon part="icon-container" name="${CALENDAR_ICON}" size="1" aria-hidden="true"></mdc-icon>
      </div>
    `;
    }
    renderPopover() {
        const calendarSelectionMode = this.effectiveSelectionMode;
        return html `
      <mdc-popover
        trigger="${TRIGGER.MANUAL}"
        triggerid="${TRIGGER_ID}"
        interactive
        ?visible="${this.displayPopover}"
        backdrop
        backdrop-append-to="${ifDefined(this.backdropAppendTo)}"
        append-to="${ifDefined(this.appendTo)}"
        hide-on-outside-click
        hide-on-escape
        focus-trap
        disable-aria-expanded
        ?disable-flip="${this.disableFlip}"
        placement="${this.placement}"
        strategy="${ifDefined(this.strategy)}"
        @shown="${() => this.focusCalendarGrid()}"
        @closebyescape="${(event) => {
            if (event.target === event.currentTarget) {
                this.displayPopover = false;
                this.focusBackToTrigger();
            }
        }}"
        @closebyoutsideclick="${() => {
            this.displayPopover = false;
            this.focusBackToTrigger();
        }}"
        exportparts="popover-content"
      >
        <mdc-calendar
          value="${ifDefined(this.value || undefined)}"
          end-value="${ifDefined(this.endValue || undefined)}"
          selection-mode="${calendarSelectionMode}"
          locale="${this.locale}"
          min="${ifDefined(this.min)}"
          max="${ifDefined(this.max)}"
          locale-today-label="${this.localeTodayLabel}"
          locale-prev-month-label="${this.localePrevMonthLabel}"
          locale-next-month-label="${this.localeNextMonthLabel}"
          @date-selected="${this.handleDateSelected}"
        ></mdc-calendar>
      </mdc-popover>
    `;
    }
    render() {
        return html `
      ${this.renderLabel()}
      <div part="container">
        ${this.variant === VARIANT.INPUT ? this.renderInputVariant() : this.renderDefaultVariant()}
        <input
          id="${this.inputId}"
          part="native-input"
          name="${this.name}"
          type="text"
          ?disabled="${this.disabled}"
          ?required="${this.required}"
          ?readonly="${this.readonly}"
          tabindex="-1"
          aria-hidden="true"
          aria-disabled="${ifDefined(this.disabled || this.softDisabled)}"
        />
        ${this.renderPopover()}
      </div>
      ${this.helpText ? this.renderHelperText() : nothing}
    `;
    }
}
DatePicker.styles = [...FormfieldWrapper.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], DatePicker.prototype, "variant", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", Object)
], DatePicker.prototype, "value", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'selection-mode' }),
    __metadata("design:type", String)
], DatePicker.prototype, "selectionMode", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], DatePicker.prototype, "locale", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], DatePicker.prototype, "min", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], DatePicker.prototype, "max", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], DatePicker.prototype, "placement", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], DatePicker.prototype, "strategy", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'disable-flip' }),
    __metadata("design:type", Boolean)
], DatePicker.prototype, "disableFlip", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'append-to' }),
    __metadata("design:type", String)
], DatePicker.prototype, "appendTo", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'backdrop-append-to' }),
    __metadata("design:type", String)
], DatePicker.prototype, "backdropAppendTo", void 0);
__decorate([
    property({ type: String, attribute: 'locale-month-label' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localeMonthLabel", void 0);
__decorate([
    property({ type: String, attribute: 'locale-day-label' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localeDayLabel", void 0);
__decorate([
    property({ type: String, attribute: 'locale-year-label' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localeYearLabel", void 0);
__decorate([
    property({ type: String, attribute: 'locale-calendar-label' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localeCalendarLabel", void 0);
__decorate([
    property({ type: String, attribute: 'locale-spinbutton-description' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localeSpinbuttonDescription", void 0);
__decorate([
    property({ type: String, attribute: 'locale-today-label' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localeTodayLabel", void 0);
__decorate([
    property({ type: String, attribute: 'locale-prev-month-label' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localePrevMonthLabel", void 0);
__decorate([
    property({ type: String, attribute: 'locale-next-month-label' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "localeNextMonthLabel", void 0);
__decorate([
    query('#month-spinbutton'),
    __metadata("design:type", HTMLInputElement)
], DatePicker.prototype, "monthInput", void 0);
__decorate([
    query('#day-spinbutton'),
    __metadata("design:type", HTMLInputElement)
], DatePicker.prototype, "dayInput", void 0);
__decorate([
    query('#year-spinbutton'),
    __metadata("design:type", HTMLInputElement)
], DatePicker.prototype, "yearInput", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], DatePicker.prototype, "displayPopover", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], DatePicker.prototype, "internalMonth", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], DatePicker.prototype, "internalDay", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], DatePicker.prototype, "internalYear", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'end-value' }),
    __metadata("design:type", Object)
], DatePicker.prototype, "endValue", void 0);
export default DatePicker;
