// @momentum-design/components/dist/react/cardradio/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/cardradio';
import { TAG_NAME } from '../../components/cardradio/cardradio.constants';
/**
 * cardradio component extends `mdc-card` and supports radio selection interaction.
 * Only one card can be selected at a time within the same group (defined by `name` attribute).
 *
 * ## Features
 * - Supports two orientations (vertical and horizontal) and three visual variants (border, ghost, and promotional).
 * - Selecting a card automatically unselects other cards in the same group and dispatches a `change` event.
 * - Supports keyboard navigation with arrow keys to move between cards in the same group.
 * - Card has `role="radio"` and manages `aria-checked` and `aria-disabled` attributes automatically.
 *
 * ## Usage
 * - Both `card-title` and `name` attributes are required.
 * - When using within a form or group, wrap cards in a container with `role="radiogroup"` and provide an `aria-label`.
 *
 * **Note**: Only pass non-interactable elements within the slots to avoid nested interactive elements.
 *
 * @tagname mdc-cardradio
 *
 * @dependency mdc-icon
 * @dependency mdc-staticradio
 * @dependency mdc-text
 *
 * @slot before-body - This slot is for passing the content before the body
 * @slot body - This slot is for passing the text content for the card
 * @slot after-body - This slot is for passing the content after the body
 * @slot title - This slot is for passing the title of the card in the header section
 * @slot subtitle - This slot is for passing the subtitle of the card in the header section
 * @slot footer-link - This slot is for passing `mdc-link` component within the footer section.
 * @slot footer-button-primary - This slot is for passing primary variant of `mdc-button` component within the footer section.
 *
 * @event click - (React: onClick) Event that gets dispatched when the card is clicked. It selects the card and unselects others in the same group.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the card.
 * It selects the card when enter key or arrow keys are used.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the card.
 * It selects the card when space key is used.
 * @event focus - (React: onFocus) Event that gets dispatched when the card receives focus.
 * @event change - (React: onChange) Event that gets dispatched when the card's checked state changes.
 *
 * @csspart header - The header part of the card
 * @csspart icon - The icon part of the card header
 * @csspart body - The body part of the card
 * @csspart image - The image part of the card
 * @csspart footer - The footer part of the card
 * @csspart footer-link - The link part of the card footer
 * @csspart footer-button-primary - The primary button part of the card footer
 * @csspart footer-button-secondary - The secondary button part of the card footer
 * @csspart icon-button - The icon button part of the card header
 * @csspart text - The text part of the card
 * @csspart check - The check part of the card
 * @csspart check-icon - The check icon part of the card
 * @csspart check-icon-button - The check icon button part of the card
 *
 * @cssproperty --mdc-card-width - The width of the card
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change',
        onClick: 'click',
        onKeyDown: 'keydown',
        onKeyUp: 'keyup',
        onFocus: 'focus',
    },
    displayName: 'CardRadio',
});
export default reactWrapper;
