// @momentum-design/components/dist/components/stepper/stepper.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { Component, Provider } from '../../models';
import { ROLE } from '../../utils/roles';
import { DEFAULTS } from './stepper.constants';
import StepperContext from './stepper.context';
import styles from './stepper.styles';
/**
 * Stepper component, which orchestrates stepperitem and stepperconnector components, is a wrapper for the stepper functionality.
 * It provides the context for the stepper items and connectors, allowing them to adapt to the stepper's orientation and variant.
 *
 * @tagname mdc-stepper
 *
 * @slot default - Pass the list of `mdc-stepperitem` and `mdc-stepperconnector` elements to be rendered inside the stepper.
 *
 */
class Stepper extends Provider {
    constructor() {
        super({
            context: StepperContext.context,
            initialValue: new StepperContext(DEFAULTS.ORIENTATION, DEFAULTS.VARIANT),
        });
        /**
         * The orientation of the stepperconnector (vertical or horizontal)
         * @default "horizontal"
         */
        this.orientation = DEFAULTS.ORIENTATION;
        /**
         * The variant of the stepper item, which can be `inline` or `stacked`.
         * @default 'inline'
         */
        this.variant = DEFAULTS.VARIANT;
    }
    connectedCallback() {
        super.connectedCallback();
        this.role = ROLE.LIST;
    }
    /**
     * Get the context for the stepper component.
     * This context provides the orientation and variant for the stepper items and connectors.
     */
    static get Context() {
        return StepperContext.context;
    }
    updated(changedProperties) {
        if (changedProperties.has('orientation') || changedProperties.has('variant')) {
            this.updateContext();
        }
    }
    /**
     * Update all observing components of this
     * provider to update the values
     *
     * Is called on every re-render, see Provider class
     */
    updateContext() {
        if (this.context.value.variant !== this.variant || this.context.value.orientation !== this.orientation) {
            this.context.value.variant = this.variant;
            this.context.value.orientation = this.orientation;
            this.context.updateObservers();
        }
    }
    render() {
        return html ` <slot></slot> `;
    }
}
Stepper.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Stepper.prototype, "orientation", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Stepper.prototype, "variant", void 0);
export default Stepper;
