// @momentum-design/components/dist/components/checkbox/checkbox.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { KEYS } from '../../utils/keys';
import { DataAriaLabelMixin } from '../../utils/mixins/DataAriaLabelMixin';
import { FormInternalsMixin } from '../../utils/mixins/FormInternalsMixin';
import FormfieldWrapper from '../formfieldwrapper/formfieldwrapper.component';
import { DEFAULTS as FORMFIELD_DEFAULTS } from '../formfieldwrapper/formfieldwrapper.constants';
import styles from './checkbox.styles';
/**
 * Checkboxes allow users to select multiple options from a list or turn an item/feature on or off.
 * These are often used in forms, settings, and selections in lists.
 *
 * A checkbox component contains an optional label and an optional helper text.
 *
 * To create a group of checkboxes, use the FormFieldGroup component.
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-staticcheckbox
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @tagname mdc-checkbox
 *
 * @event change - (React: onChange) Event that gets dispatched when the checkbox state changes.
 * @event focus - (React: onFocus) Event that gets dispatched when the checkbox receives focus.
 *
 * @cssproperty --mdc-checkbox-background-color-hover - Allows customization of the background color on hover.
 * @cssproperty --mdc-checkbox-checked-background-color-hover - Background color for a selected checkbox when hovered.
 * @cssproperty --mdc-checkbox-checked-pressed-icon-color - Background color for a selected checkbox when pressed.
 * @cssproperty --mdc-checkbox-pressed-icon-color - Background color for a selected checkbox when pressed.
 * @cssproperty --mdc-checkbox-disabled-checked-icon-color - Background color for a selected checkbox when disabled.
 */
class Checkbox extends FormInternalsMixin(DataAriaLabelMixin(FormfieldWrapper)) {
    constructor() {
        super(...arguments);
        /**
         * Determines whether the checkbox is selected or unselected.
         *
         * @default false
         */
        this.checked = false;
        /**
         * This property is used to determine the parent checkbox in a nested checkbox group.
         * If any one of the children is unselected, then the parent checkbox will be indeterminate.
         * If all children are either selected or unselected, then the parent checkbox will not be indeterminate.
         *
         * @default false
         */
        this.indeterminate = false;
        /**
         * Automatically focus on the element when the page loads.
         * [MDN Reference](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autofocus)
         * @default false
         */
        this.autofocus = false;
        this.renderLabelAndHelperText = () => {
            if (!this.label)
                return nothing;
            return html `<div class="text-container">${this.renderLabel()} ${this.renderHelperText()}</div>`;
        };
    }
    connectedCallback() {
        super.connectedCallback();
        // Checkbox does not contain helpTextType property.
        this.helpTextType = undefined;
    }
    /**
     * Updates the form value to reflect the current state of the checkbox.
     * If checked, the value is set to either the user-provided value or 'on' if no value is provided.
     * If unchecked, the value is set to null.
     */
    setFormValue() {
        let actualValue = null;
        if (this.checked) {
            actualValue = !this.value ? 'on' : this.value;
        }
        else {
            actualValue = null;
        }
        this.manageRequired();
        this.internals.setFormValue(actualValue);
    }
    /**
     * Manages the required state of the checkbox.
     * If the checkbox is not checked and the required property is set, then the checkbox is invalid.
     */
    manageRequired() {
        if (!this.checked && this.required) {
            if (this.validationMessage) {
                this.inputElement.setCustomValidity(this.validationMessage);
            }
            else {
                this.inputElement.setCustomValidity('');
            }
            this.setValidity();
        }
        else {
            this.internals.setValidity({});
        }
    }
    /** @internal
     * Resets the checkbox to its initial state.
     * The checked property is set to false.
     */
    formResetCallback() {
        this.checked = false;
        this.indeterminate = false;
    }
    /** @internal */
    formStateRestoreCallback(state) {
        if (state) {
            this.checked = true;
        }
    }
    /**
     * Toggles the state of the checkbox element.
     * If the element is not disabled, then
     * the checked property is toggled and the indeterminate property is set to false.
     */
    toggleState() {
        if (!this.disabled) {
            this.checked = !this.checked;
            this.indeterminate = false;
        }
    }
    /**
     * Handles the keydown event on the checkbox.
     * When the user presses Enter, the form is submitted.
     * @param event - The keyboard event.
     */
    handleKeyDown(event) {
        var _a;
        if (event.key === KEYS.ENTER) {
            (_a = this.form) === null || _a === void 0 ? void 0 : _a.requestSubmit();
        }
    }
    /**
     * Toggles the state of the checkbox element.
     * and dispatch the new change event.
     */
    handleChange(event) {
        this.toggleState();
        const EventConstructor = event.constructor;
        this.dispatchEvent(new EventConstructor(event.type, event));
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('checked')) {
            this.setFormValue();
        }
    }
    render() {
        var _a;
        return html `
      <mdc-staticcheckbox
        class="mdc-focus-ring"
        ?checked="${this.checked}"
        ?indeterminate="${this.indeterminate}"
        ?disabled="${this.disabled}"
      >
        <input
          id="${this.inputId}"
          type="checkbox"
          class="input"
          ?autofocus="${this.autofocus}"
          name="${ifDefined(this.name)}"
          value="${ifDefined(this.value)}"
          ?required="${this.required}"
          .checked="${this.checked}"
          aria-checked="${this.indeterminate ? 'mixed' : this.checked}"
          .indeterminate="${this.indeterminate}"
          .disabled="${this.disabled}"
          aria-label="${(_a = this.dataAriaLabel) !== null && _a !== void 0 ? _a : ''}"
          tabindex="${this.disabled ? -1 : 0}"
          aria-describedby="${ifDefined(this.helpText ? FORMFIELD_DEFAULTS.HELPER_TEXT_ID : '')}"
          @change=${this.handleChange}
          @keydown=${this.handleKeyDown}
        />
      </mdc-staticcheckbox>
      ${this.renderLabelAndHelperText()}
    `;
    }
}
Checkbox.styles = [...FormfieldWrapper.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], Checkbox.prototype, "checked", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], Checkbox.prototype, "indeterminate", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Object)
], Checkbox.prototype, "autofocus", void 0);
export default Checkbox;
