// @momentum-design/components/dist/components/chip/chip.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { IconNameMixin } from '../../utils/mixins/IconNameMixin';
import Buttonsimple from '../buttonsimple/buttonsimple.component';
import StaticChip from '../staticchip/staticchip.component';
import { DEFAULTS } from './chip.constants';
import styles from './chip.styles';
/**
 * mdc-chip is an interactive element that can be used to represent a chip. It supports a leading icon along with label.
 * Consumers can wrap this component around a tooltip to provide additional context.
 *
 * It is recommended to keep the label text for the chip component concise and compact.<br/>
 * For best results, we recommend limiting the <b>maximum length of the label text to 20 characters</b>,
 * including empty spaces to split words.
 *
 * This component is built by extending `Buttonsimple`.
 *
 * @tagname mdc-chip
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 *
 * @cssproperty --mdc-chip-color - The color of the chip.
 * @cssproperty --mdc-chip-border-color - The border color of the chip.
 * @cssproperty --mdc-chip-background-color - The background color of the chip.
 *
 * @csspart icon - The icon part of the chip.
 * @csspart label - The label part of the chip.
 *
 * @event click - (React: onClick) This event is dispatched when the chip is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the chip.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the chip.
 * @event focus - (React: onFocus) This event is dispatched when the chip receives focus.
 */
class Chip extends IconNameMixin(Buttonsimple) {
    constructor() {
        super(...arguments);
        /**
         * The color of the chip. It supports the following colors
         * - default
         * - cobalt
         * - gold
         * - lime
         * - mint
         * - orange
         * - pink
         * - purple
         * - slate
         * - violet
         *
         * @default default
         */
        this.color = DEFAULTS.COLOR;
    }
    connectedCallback() {
        super.connectedCallback();
        this.role = DEFAULTS.ROLE;
        this.size = DEFAULTS.SIZE;
        this.active = undefined;
        this.softDisabled = undefined;
    }
    /**
     * Renders the icon element if available.
     * @returns The icon element if available, otherwise nothing.
     */
    renderIcon() {
        if (!this.iconName)
            return nothing;
        return html ` <mdc-icon part="icon" name="${this.iconName}" length-unit="rem" size="1"></mdc-icon> `;
    }
    render() {
        return html `
      ${this.renderIcon()}
      ${this.label
            ? html `<mdc-text part="label" type="${DEFAULTS.TEXT_TYPE}" tagname="${DEFAULTS.TAG_NAME}"
            >${this.label}</mdc-text
          >`
            : nothing}
    `;
    }
}
Chip.styles = [...StaticChip.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Chip.prototype, "color", void 0);
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], Chip.prototype, "label", void 0);
export default Chip;
