// @momentum-design/components/dist/components/select/select.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property, query, queryAssignedElements, state } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { KEYS } from '../../utils/keys';
import { DataAriaLabelMixin } from '../../utils/mixins/DataAriaLabelMixin';
import { FormInternalsMixin } from '../../utils/mixins/FormInternalsMixin';
import { ROLE } from '../../utils/roles';
import FormfieldWrapper from '../formfieldwrapper/formfieldwrapper.component';
import { DEFAULTS as FORMFIELD_DEFAULTS, VALIDATION } from '../formfieldwrapper/formfieldwrapper.constants';
import { TAG_NAME as OPTION_TAG_NAME } from '../option/option.constants';
import { POPOVER_PLACEMENT, DEFAULTS as POPOVER_DEFAULTS } from '../popover/popover.constants';
import { TYPE, VALID_TEXT_TAGS } from '../text/text.constants';
import { ARROW_ICON, LISTBOX_ID, TRIGGER_ID } from './select.constants';
import styles from './select.styles';
/**
 * The mdc-select component is a dropdown selection control that allows users to pick an option from a predefined list.
 * It is designed to work with `mdc-option` for individual options and `mdc-optgroup` for grouping related options.
 * The component ensures accessibility and usability while handling various use cases,
 * including long text truncation with tooltip support.
 *
 * Every mdc-option should have a `value` attribute set to ensure proper form submission.
 *
 * To set a default option, use the `selected` attribute on the `mdc-option` element.
 *
 * **Note:** Make sure to add `mdc-selectlistbox` as a child of `mdc-select` and wrap options/optgroup in it to ensure proper accessibility functionality. Read more about it in SelectListBox documentation.
 *
 * @dependency mdc-button
 * @dependency mdc-icon
 * @dependency mdc-popover
 * @dependency mdc-text
 * @dependency mdc-toggletip
 *
 * @tagname mdc-select
 *
 * @slot default - This is a default/unnamed slot for Selectlistbox including options and/or option group.
 *
 * @event click - (React: onClick) This event is dispatched when the select is clicked.
 * @event change - (React: onChange) This event is dispatched when the select is changed.
 * @event input - (React: onInput) This event is dispatched when the select is changed.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the select.
 * @event focus - (React: onFocus) This event is dispatched when the select receives focus.
 *
 * @cssproperty --mdc-select-background-color - The background color of the combobox of select.
 * @cssproperty --mdc-select-background-color-hover - The background color of the combobox of select when hovered.
 * @cssproperty --mdc-select-background-color-active - The background color of the combobox of select when active.
 * @cssproperty --mdc-select-background-color-disabled - The background color of the combobox of select when disabled.
 * @cssproperty --mdc-select-text-color - The text color of the select.
 * @cssproperty --mdc-select-text-color-selected - The text color of the selected option in the select.
 * @cssproperty --mdc-select-text-color-disabled - The text color of the select when disabled.
 * @cssproperty --mdc-select-border-color - The border color of the select.
 * @cssproperty --mdc-select-border-color-disabled - The border color of the select when disabled.
 * @cssproperty --mdc-select-border-color-success - The border color of the select when in success state.
 * @cssproperty --mdc-select-border-color-warning - The border color of the select when in warning state.
 * @cssproperty --mdc-select-border-color-error - The border color of the select when in error state.
 * @cssproperty --mdc-select-width - The width of the select.
 * @cssproperty --mdc-select-listbox-height - The height of the listbox inside the select.
 * @cssproperty --mdc-select-listbox-width - The width of the listbox inside the select (default: `--mdc-select-width`).
 */
class Select extends FormInternalsMixin(DataAriaLabelMixin(FormfieldWrapper)) {
    constructor() {
        super(...arguments);
        /**
         * readonly attribute of the select field. If true, the select is read-only.
         * @default false
         */
        this.readonly = false;
        /**
         * The placeholder text which will be shown on the text if provided.
         */
        this.placement = POPOVER_PLACEMENT.BOTTOM_START;
        /**
         * This describes the clipping element(s) or area that overflow of the used popover will be checked relative to.
         * The default is 'clippingAncestors', which are the overflow ancestors which will cause the
         * element to be clipped.
         *
         * Possible values:
         *  - 'clippingAncestors'
         *  - any css selector
         *
         * @default 'clippingAncestors'
         *
         * @see [Floating UI - boundary](https://floating-ui.com/docs/detectOverflow#boundary)
         */
        this.boundary = POPOVER_DEFAULTS.BOUNDARY;
        /**
         * The strategy of the popover within Select.
         * This determines how the popover is positioned in the DOM.
         *
         * In case `boundary` is set to something other than 'clippingAncestors',
         * it might be necessary to set the `strategy` to 'fixed' to ensure that the popover
         * is not getting clipped by scrollable containers enclosing the select.
         *
         * @default absolute
         * @see [Floating UI - strategy](https://floating-ui.com/docs/computePosition#strategy)
         */
        this.strategy = POPOVER_DEFAULTS.STRATEGY;
        /**
         * The z-index of the popover within Select.
         *
         * Override this to make sure this stays on top of other components.
         * @default 1000
         */
        this.popoverZIndex = POPOVER_DEFAULTS.Z_INDEX;
        /** @internal */
        this.displayPopover = false;
        /** @internal */
        this.initialSelectedOption = null;
    }
    getAllValidOptions() {
        var _a;
        return Array.from(((_a = this.slottedListboxes[0]) === null || _a === void 0 ? void 0 : _a.querySelectorAll(`${OPTION_TAG_NAME}:not([disabled])`)) || []);
    }
    getFirstValidOption() {
        var _a;
        return (_a = this.slottedListboxes[0]) === null || _a === void 0 ? void 0 : _a.querySelector(`${OPTION_TAG_NAME}:not([disabled])`);
    }
    getLastValidOption() {
        const options = this.getAllValidOptions();
        return options.length > 0 ? options[options.length - 1] : null;
    }
    getFirstSelectedOption() {
        var _a;
        return (_a = this.slottedListboxes[0]) === null || _a === void 0 ? void 0 : _a.querySelector(`${OPTION_TAG_NAME}[selected]:not([disabled])`);
    }
    /**
     * Handles the first updated lifecycle event.
     * If an option is selected, use that as the value.
     * If not, use the placeholder if it exists, otherwise use the first option.
     */
    async firstUpdated() {
        await this.updateComplete;
        this.modifyListBoxWrapper();
        const firstSelectedOption = this.getFirstSelectedOption();
        if (firstSelectedOption) {
            this.initialSelectedOption = firstSelectedOption;
            // do not fire events when setting the selected value
            // which is already selected in the DOM on first update
            this.setSelectedOption(firstSelectedOption);
        }
        else if (!this.placeholder) {
            const firstValidOption = this.getFirstValidOption();
            // We will show the first option as selected & fire
            // and event since the selected option changed
            this.setSelectedOption(firstValidOption);
            this.fireEvents();
        }
        else if (this.placeholder) {
            // If there is no default selected option
            // then we call the native validity
            this.setInputValidity();
        }
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('disabled') ||
            changedProperties.has('softDisabled') ||
            changedProperties.has('readonly')) {
            if (this.disabled || this.softDisabled || this.readonly) {
                // If the select is disabled, soft-disabled or readonly,
                // we close the popover if it is open.
                this.displayPopover = false;
            }
        }
        if (changedProperties.has('dataAriaLabel')) {
            this.modifyListBoxWrapper();
        }
    }
    /**
     * Modifies the listbox wrapper to ensure it has the correct attributes
     * and IDs for accessibility.
     *
     * Once [ariaOwnsElements](https://developer.mozilla.org/en-US/docs/Web/API/ElementInternals/ariaOwnsElements) is supported in browsers,
     * this an be removed and mdc-option can be used directly in the select component with a listbox in a different
     * shadow root and aria-owns attribute to connect them.
     */
    modifyListBoxWrapper() {
        const slottedListBox = this.slottedListboxes[0];
        if (!slottedListBox) {
            return;
        }
        slottedListBox.setAttribute('id', LISTBOX_ID);
        slottedListBox.setAttribute('aria-label', this.dataAriaLabel || '');
        slottedListBox.setAttribute('aria-labelledby', TRIGGER_ID);
    }
    /**
     * A private method which is called when an option is clicked.
     * It sets the selected option, removes selected from other options, updates the tabindex for all options,
     * closes the popover, and fires the change and input events.
     * @param event - The event which triggered this function.
     */
    handleOptionsClick(event) {
        const option = event.target;
        if (option &&
            option.tagName === OPTION_TAG_NAME.toUpperCase() &&
            !option.hasAttribute('disabled') &&
            !option.hasAttribute('soft-disabled')) {
            this.setSelectedOption(option);
            this.displayPopover = false;
            this.fireEvents();
        }
    }
    /**
     * Sets the selected option in the component state and updates the input element's value.
     * This method ensures that only the selected option is marked as selected in the DOM,
     * and updates the tabindex for all options accordingly.
     * It also updates the validity of the input element based on the selected option.
     * This method is called when an option is selected.
     *
     * @param option - The option element in DOM which gets selected.
     */
    setSelectedOption(option) {
        var _a, _b, _c;
        // set the attribute 'selected' on the option in HTML and remove it from others
        this.updateSelectedInChildOptions(option);
        // update the tabindex for all options
        this.updateTabIndexForAllOptions(option);
        // set the selected option in the component state
        this.selectedOption = option;
        // update all form related values
        this.value = (_b = (_a = this.selectedOption) === null || _a === void 0 ? void 0 : _a.value) !== null && _b !== void 0 ? _b : '';
        this.internals.setFormValue(this.value);
        (_c = this.inputElement) === null || _c === void 0 ? void 0 : _c.setAttribute('value', this.value);
        this.setInputValidity();
    }
    /**
     * Updates the tabindex of all options.
     * Sets the tabindex of the selected option to '0' and others to '-1'.
     *
     * @param option - The option which tabIndex should be set to 0.
     */
    updateTabIndexForAllOptions(option) {
        const options = this.getAllValidOptions();
        const optionToGetTabIndex0 = option || options[0];
        options.forEach(option => {
            option.setAttribute('tabindex', option === optionToGetTabIndex0 ? '0' : '-1');
        });
    }
    /**
     * Sets selected attribute on the selected option and removes it from all options
     * @param selectedOption - The option which gets selected
     */
    updateSelectedInChildOptions(selectedOption) {
        selectedOption === null || selectedOption === void 0 ? void 0 : selectedOption.setAttribute('selected', 'true');
        const options = this.getAllValidOptions();
        options.forEach(option => {
            if (option !== selectedOption) {
                option.removeAttribute('selected');
            }
        });
    }
    /**
     * A private method which is called to fire the change and input events.
     * It dispatches the input and change events with the selected option's value and label.
     */
    fireEvents() {
        this.dispatchInput(this.selectedOption);
        this.dispatchChange(this.selectedOption);
    }
    /**
     * Sets the validity of the input element based on the selected option.
     * If the selected option is not set and the select is required,
     * it sets a custom validation message.
     * If the selected option is set or the select is not required,
     * it clears the custom validation message.
     * This method is called to ensure that the select component behaves correctly
     * in form validation scenarios, especially when the select is required.
     * @internal
     */
    setInputValidity() {
        var _a, _b;
        if (!this.selectedOption && this.required && this.validationMessage) {
            (_a = this.inputElement) === null || _a === void 0 ? void 0 : _a.setCustomValidity(this.validationMessage);
        }
        else {
            (_b = this.inputElement) === null || _b === void 0 ? void 0 : _b.setCustomValidity('');
        }
        this.setValidity();
    }
    /**
     * Resets the select to its initially selected option.
     * @internal
     */
    formResetCallback() {
        var _a;
        const optionToResetTo = this.initialSelectedOption || null;
        if (((_a = this.selectedOption) === null || _a === void 0 ? void 0 : _a.value) !== (optionToResetTo === null || optionToResetTo === void 0 ? void 0 : optionToResetTo.value)) {
            this.setSelectedOption(optionToResetTo);
            // fire events to notify the change in case of reset
            this.fireEvents();
        }
    }
    /** @internal */
    formStateRestoreCallback(state) {
        var _a;
        const optionToRestoreTo = this.getAllValidOptions().find(option => option.value === state || option.label === state);
        if (((_a = this.selectedOption) === null || _a === void 0 ? void 0 : _a.value) !== (optionToRestoreTo === null || optionToRestoreTo === void 0 ? void 0 : optionToRestoreTo.value)) {
            this.setSelectedOption(optionToRestoreTo || null);
            // fire events to notify the change in case of restore
            this.fireEvents();
        }
    }
    dispatchChange(option) {
        this.dispatchEvent(new CustomEvent('change', {
            detail: { value: option === null || option === void 0 ? void 0 : option.value, label: option === null || option === void 0 ? void 0 : option.label },
            composed: true,
            bubbles: true,
        }));
    }
    dispatchInput(option) {
        this.dispatchEvent(new CustomEvent('input', {
            detail: { value: option === null || option === void 0 ? void 0 : option.value, label: option === null || option === void 0 ? void 0 : option.label },
            composed: true,
            bubbles: true,
        }));
    }
    /**
     * Handles the click event on the visual combobox.
     * If the select is disabled, soft-disabled or readonly, it does nothing.
     * If the popover is already open, it closes it.
     * If it is closed, it opens it.
     * @param event - The mouse event which triggered this function.
     */
    handleClickCombobox(event) {
        if (this.disabled || this.softDisabled || this.readonly) {
            return;
        }
        // If the popover is already open, we close it.
        // If it is closed, we open it.
        this.displayPopover = !this.displayPopover;
        event.stopPropagation();
    }
    /**
     * Handles the keydown event on the select element when the popover is closed.
     * The options are as follows:
     * - ARROW_DOWN, ARROW_UP, SPACE: Opens the popover and prevents the default scrolling behavior.
     * - ENTER: Opens the popover, prevents default scrolling, and submits the form if the popover is closed.
     * - HOME: Opens the popover and sets focus and tabindex on the first option.
     * - END: Opens the popover and sets focus and tabindex on the last option.
     * @param event - The keyboard event.
     */
    handleKeydownCombobox(event) {
        if (this.disabled || this.softDisabled || this.readonly) {
            return;
        }
        switch (event.key) {
            case KEYS.ARROW_DOWN:
            case KEYS.ARROW_UP:
                this.displayPopover = true;
                // Prevent the default browser behavior of scrolling down
                event.preventDefault();
                event.stopPropagation();
                break;
            case KEYS.ENTER:
            case KEYS.SPACE:
                this.displayPopover = true;
                // Prevent the default browser behavior of scrolling down
                event.preventDefault();
                event.stopPropagation();
                break;
            case KEYS.HOME: {
                this.displayPopover = true;
                const firstOption = this.getFirstValidOption();
                if (firstOption) {
                    firstOption === null || firstOption === void 0 ? void 0 : firstOption.focus();
                    this.updateTabIndexForAllOptions(firstOption);
                }
                event.preventDefault();
                break;
            }
            case KEYS.END: {
                this.displayPopover = true;
                const lastOption = this.getLastValidOption();
                if (lastOption) {
                    lastOption.focus();
                    this.updateTabIndexForAllOptions(lastOption);
                }
                event.preventDefault();
                break;
            }
            default:
                break;
        }
    }
    /**
     * Handles the keydown event on the select element when the popover is open.
     * The options are as follows:
     * - HOME: Sets focus and tabindex on the first option.
     * - END: Sets focus and tabindex on the last option.
     * - ARROW_DOWN, ARROW_UP, PAGE_DOWN, PAGE_UP: Handles navigation between options.
     * @param event - The keyboard event.
     */
    handlePopoverKeydown(event) {
        let optionToFocus = null;
        switch (event.key) {
            case KEYS.HOME: {
                optionToFocus = this.getFirstValidOption();
                break;
            }
            case KEYS.END: {
                optionToFocus = this.getLastValidOption();
                break;
            }
            case KEYS.ARROW_DOWN: {
                const options = this.getAllValidOptions();
                const currentIndex = options.findIndex(option => option === event.target);
                const newIndex = Math.min(currentIndex + 1, options.length - 1);
                optionToFocus = options[newIndex];
                break;
            }
            case KEYS.ARROW_UP: {
                const options = this.getAllValidOptions();
                const currentIndex = options.findIndex(option => option === event.target);
                const newIndex = Math.max(currentIndex - 1, 0);
                optionToFocus = options[newIndex];
                break;
            }
            case KEYS.PAGE_DOWN: {
                const options = this.getAllValidOptions();
                const currentIndex = options.findIndex(option => option === event.target);
                const newIndex = Math.min(currentIndex + 10, options.length - 1);
                optionToFocus = options[newIndex];
                break;
            }
            case KEYS.PAGE_UP: {
                const options = this.getAllValidOptions();
                const currentIndex = options.findIndex(option => option === event.target);
                const newIndex = Math.max(currentIndex - 10, 0);
                optionToFocus = options[newIndex];
                break;
            }
            default:
                break;
        }
        if (optionToFocus) {
            this.focusAndUpdateTabIndexes(optionToFocus);
            event.preventDefault();
            event.stopPropagation();
        }
    }
    /**
     * Focuses the given option and updates the tabindex for all options.
     * @param option - The option to focus.
     * @internal
     */
    focusAndUpdateTabIndexes(option) {
        if (option) {
            option.focus();
            this.updateTabIndexForAllOptions(option);
        }
    }
    /**
     * If the native input is focused, it will focus the visual combobox.
     * This is to ensure that the visual combobox is focused when the native input is focused.
     * For example when a form is submitted and the native input is focused,
     * we want to focus the visual combobox so that the user can see the selected option
     * and can interact with it.
     * @internal
     */
    handleNativeInputFocus() {
        this.visualCombobox.focus();
    }
    /**
     * Updates the state of the select component.
     * This public method should be fired when the selected on the option components is updated from the outside.
     * It ensures that the selected attribute is set correctly on the options
     * and that the aria-selected attribute is updated accordingly.
     */
    updateState() {
        var _a;
        const newSelectedOption = this.getFirstSelectedOption();
        if (!this.inputElement) {
            return;
        }
        if (!newSelectedOption) {
            this.setSelectedOption(this.placeholder ? null : this.getFirstValidOption());
        }
        else if (((_a = this.selectedOption) === null || _a === void 0 ? void 0 : _a.value) !== newSelectedOption.value) {
            this.setSelectedOption(newSelectedOption);
        }
    }
    render() {
        var _a, _b, _c, _d, _e, _f;
        return html `
      ${this.renderLabel()}
      <div part="container">
        <div
          id="${TRIGGER_ID}"
          part="base-container"
          @click="${this.handleClickCombobox}"
          @keydown="${this.handleKeydownCombobox}"
          tabindex="${this.disabled ? '-1' : '0'}"
          class="${this.disabled ? '' : 'mdc-focus-ring'}"
          role="${ROLE.COMBOBOX}"
          aria-controls="${LISTBOX_ID}"
          aria-label="${(_a = this.dataAriaLabel) !== null && _a !== void 0 ? _a : ''}"
          aria-labelledby="${this.label ? FORMFIELD_DEFAULTS.HEADING_ID : ''}"
          aria-expanded="${this.displayPopover ? 'true' : 'false'}"
          aria-haspopup="${ROLE.LISTBOX}"
          aria-required="${this.required ? 'true' : 'false'}"
          aria-invalid="${this.helpTextType === VALIDATION.ERROR ? 'true' : 'false'}"
          aria-disabled="${ifDefined(this.disabled || this.softDisabled)}"
          aria-readonly="${ifDefined(this.readonly)}"
        >
          ${((_b = this.selectedOption) === null || _b === void 0 ? void 0 : _b.prefixIcon)
            ? html `<mdc-icon
                length-unit="rem"
                size="1"
                name="${(_c = this.selectedOption) === null || _c === void 0 ? void 0 : _c.prefixIcon}"
                part="selected-icon"
              ></mdc-icon>`
            : nothing}
          <mdc-text
            part="base-text ${((_d = this.selectedOption) === null || _d === void 0 ? void 0 : _d.label) ? 'selected' : ''}"
            type="${TYPE.BODY_MIDSIZE_REGULAR}"
            tagname="${VALID_TEXT_TAGS.SPAN}"
          >
            ${(_f = (_e = this.selectedOption) === null || _e === void 0 ? void 0 : _e.label) !== null && _f !== void 0 ? _f : this.placeholder}
          </mdc-text>
          <div part="icon-container">
            <mdc-icon
              size="1"
              length-unit="rem"
              name="${this.displayPopover ? ARROW_ICON.ARROW_UP : ARROW_ICON.ARROW_DOWN}"
            ></mdc-icon>
          </div>
        </div>
        <input
          id="${this.inputId}"
          part="native-input"
          name="${this.name}"
          type="text"
          ?autofocus="${this.autofocus}"
          ?disabled=${this.disabled}
          ?required=${this.required}
          ?readonly=${this.readonly}
          tabindex="-1"
          aria-hidden="true"
          @focus=${this.handleNativeInputFocus}
          @invalid=${this.setInputValidity}
          aria-disabled="${ifDefined(this.disabled || this.softDisabled)}"
        />
        <mdc-popover
          trigger="manual"
          triggerid="${TRIGGER_ID}"
          @keydown="${this.handlePopoverKeydown}"
          interactive
          ?visible="${this.displayPopover}"
          role=""
          backdrop
          backdrop-append-to="${ifDefined(this.backdropAppendTo)}"
          hide-on-outside-click
          hide-on-escape
          focus-back-to-trigger
          focus-trap
          size
          boundary="${ifDefined(this.boundary)}"
          strategy="${ifDefined(this.strategy)}"
          placement="${this.placement}"
          @closebyescape="${() => {
            this.displayPopover = false;
        }}"
          @closebyoutsideclick="${() => {
            this.displayPopover = false;
        }}"
          z-index="${ifDefined(this.popoverZIndex)}"
          exportparts="popover-content"
        >
          <slot @click="${this.handleOptionsClick}"></slot>
        </mdc-popover>
      </div>
      ${this.renderHelperText()}
    `;
    }
}
Select.styles = [...FormfieldWrapper.styles, ...styles];
__decorate([
    property({ type: String }),
    __metadata("design:type", String)
], Select.prototype, "placeholder", void 0);
__decorate([
    property({ type: Boolean }),
    __metadata("design:type", Object)
], Select.prototype, "readonly", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Select.prototype, "placement", void 0);
__decorate([
    property({ type: Boolean, attribute: 'soft-disabled', reflect: true }),
    __metadata("design:type", Boolean)
], Select.prototype, "softDisabled", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'boundary' }),
    __metadata("design:type", String)
], Select.prototype, "boundary", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'strategy' }),
    __metadata("design:type", String)
], Select.prototype, "strategy", void 0);
__decorate([
    property({ type: Number, reflect: true, attribute: 'popover-z-index' }),
    __metadata("design:type", Number)
], Select.prototype, "popoverZIndex", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'backdrop-append-to' }),
    __metadata("design:type", String)
], Select.prototype, "backdropAppendTo", void 0);
__decorate([
    queryAssignedElements({ selector: 'mdc-selectlistbox' }),
    __metadata("design:type", Array)
], Select.prototype, "slottedListboxes", void 0);
__decorate([
    query(`[id="${TRIGGER_ID}"]`),
    __metadata("design:type", HTMLDivElement)
], Select.prototype, "visualCombobox", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], Select.prototype, "selectedOption", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], Select.prototype, "displayPopover", void 0);
export default Select;
