// @momentum-design/components/dist/components/sidenavigation/sidenavigation.context.js
/* eslint-disable no-param-reassign */
import { createContext } from '@lit/context';
import { TAG_NAME as MENUPOPOVER_TAGNAME } from '../menupopover/menupopover.constants';
import { TAG_NAME as NAVMENUITEM_TAGNAME } from '../navmenuitem/navmenuitem.constants';
import { TAG_NAME } from './sidenavigation.constants';
class SideNavigationContext {
    constructor(defaultVariant, defaultExpanded) {
        this.variant = defaultVariant;
        this.expanded = defaultExpanded;
    }
    hasSiblingWithTriggerId(navMenuItem) {
        var _a, _b;
        const id = navMenuItem === null || navMenuItem === void 0 ? void 0 : navMenuItem.getAttribute('id');
        if (!id)
            return false;
        const siblings = Array.from((_b = (_a = navMenuItem === null || navMenuItem === void 0 ? void 0 : navMenuItem.parentElement) === null || _a === void 0 ? void 0 : _a.children) !== null && _b !== void 0 ? _b : []);
        return siblings.some(sibling => sibling !== navMenuItem &&
            sibling.tagName.toLowerCase() === MENUPOPOVER_TAGNAME &&
            sibling.getAttribute('triggerid') === id);
    }
    getParentNavMenuItems(navMenuItem) {
        if (!navMenuItem)
            return [];
        const parents = [];
        let current = navMenuItem;
        while (current) {
            // Walk up to find the menupopover
            const popover = current === null || current === void 0 ? void 0 : current.closest(MENUPOPOVER_TAGNAME);
            if (!popover)
                break;
            const triggerId = popover.getAttribute('triggerid');
            if (!triggerId)
                break;
            // Find the NavMenuItem that triggered this menupopover
            const triggeringNavMenuItem = document.getElementById(triggerId);
            if (triggeringNavMenuItem && triggeringNavMenuItem.tagName.toLowerCase() === NAVMENUITEM_TAGNAME) {
                parents.push(triggeringNavMenuItem);
                current = triggeringNavMenuItem;
            }
            else {
                break;
            }
        }
        return parents;
    }
    setCurrentActiveNavMenuItem(navMenuItem) {
        var _a;
        const isSameItem = ((_a = this.currentActiveNavMenuItem) === null || _a === void 0 ? void 0 : _a.navId) === (navMenuItem === null || navMenuItem === void 0 ? void 0 : navMenuItem.navId);
        const shouldSkip = (navMenuItem === null || navMenuItem === void 0 ? void 0 : navMenuItem.cannotActivate) || this.hasSiblingWithTriggerId(navMenuItem) || (navMenuItem === null || navMenuItem === void 0 ? void 0 : navMenuItem.softDisabled);
        if (isSameItem || shouldSkip)
            return;
        // Clean up previous active item
        if (this.currentActiveNavMenuItem) {
            this.currentActiveNavMenuItem.removeAttribute('aria-current');
            this.currentActiveNavMenuItem.removeAttribute('active');
            const previousParents = this.getParentNavMenuItems(this.currentActiveNavMenuItem);
            previousParents.forEach(parent => {
                parent.hasActiveChild = false;
                parent.removeAttribute('active');
            });
        }
        // Apply attributes to new active item
        if (!navMenuItem)
            return;
        this.currentActiveNavMenuItem = navMenuItem;
        if (!(navMenuItem === null || navMenuItem === void 0 ? void 0 : navMenuItem.disableAriaCurrent)) {
            navMenuItem.setAttribute('aria-current', 'page');
        }
        navMenuItem.setAttribute('active', '');
        const newParents = this.getParentNavMenuItems(navMenuItem);
        newParents.forEach(parent => {
            parent.hasActiveChild = true;
            parent.setAttribute('active', '');
        });
    }
}
SideNavigationContext.context = createContext(TAG_NAME);
export default SideNavigationContext;
