// @momentum-design/components/dist/utils/mixins/KeyDownHandledMixin.js
/**
 * Mixin to provide abstract key mapping for navigation and actions keys.
 */
export const KeyDownHandledMixin = (superClass) => {
    class InnerMixinClass extends superClass {
        constructor(...rest) {
            super(...rest);
            /**
             * Flag to track if the keydown event has been handled.
             * @internal
             */
            this.keyDownEventHandledFlag = false;
            /**
             * Reset keyDownEventHadnle to false
             * @internal
             */
            this.keyDownEventHandledHandler = () => {
                setTimeout(() => {
                    this.keyDownEventHandledFlag = false;
                });
            };
            /**
             * Handle navbeforeprocess event to prevent default behavior if the keydown event has been handled by the combobox.
             * @internal
             */
            this.handleNavBeforeProcessEvent = (event) => {
                if (this.keyDownEventHandledFlag) {
                    event.preventDefault();
                }
            };
            this.addEventListener('keydown', this.keyDownEventHandledHandler);
            this.addEventListener('navbeforeprocess', this.handleNavBeforeProcessEvent);
        }
        /**
         * @see KeyToActionInterface.keyDownEventHandled
         * @internal
         */
        keyDownEventHandled() {
            this.keyDownEventHandledFlag = true;
        }
        /**
         * @see KeyToActionInterface.isKeyDownEventHandled
         * @internal
         */
        get isKeyDownEventHandled() {
            return this.keyDownEventHandledFlag;
        }
    }
    // Cast return type to your mixin's interface intersected with the superClass type
    return InnerMixinClass;
};
