// @momentum-design/components/dist/react/toggletip/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/toggletip';
import { TAG_NAME } from '../../components/toggletip/toggletip.constants';
/**
 * A toggletip is triggered by clicking a trigger element.
 *
 * It can contain interactive content and can be closed by
 * clicking outside the toggletip or pressing the escape key.
 *
 * It can have optional close button to close the toggletip.
 *
 * Toggletip component uses `mdc-screenreaderannouncer` internally to
 * announce the toggletip text content with screen readers when the toggletip is shown.
 *
 * `screenreader-announcer-identity` attribute can be used to provide ID of an element
 * in DOM to which announcement elements are added. If not set, a visually hidden
 * div element is created in DOM to which announcement elements are added.
 *
 * Please refer to the `mdc-screenreaderannouncer` component for more details.
 *
 * @dependency mdc-screenreaderannouncer
 * @dependency mdc-button
 *
 * @tagname mdc-toggletip
 *
 * @event shown - (React: onShown) This event is dispatched when the toggletip is shown
 * @event hidden - (React: onHidden) This event is dispatched when the toggletip is hidden
 * @event created - (React: onCreated) This event is dispatched when the toggletip is created (added to the DOM)
 * @event destroyed - (React: onDestroyed) This event is dispatched when the toggletip
 * is destroyed (removed from the DOM)
 *
 * @cssproperty --mdc-toggletip-max-width - The maximum width of the toggletip.
 * @cssproperty --mdc-toggletip-text-color - The text color of the toggletip.
 * @cssproperty --mdc-toggletip-text-color-contrast - The text color of the toggletip
 *  when the color is contrast.
 *
 * @slot - Default slot for the toggletip content
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onShown: 'shown',
        onHidden: 'hidden',
        onCreated: 'created',
        onDestroyed: 'destroyed',
    },
    displayName: 'ToggleTip',
});
export default reactWrapper;
