// @momentum-design/components/dist/react/link/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/link';
import { TAG_NAME } from '../../components/link/link.constants';
/**
 * `mdc-link` component can be used to navigate to a different page
 * within the application or to an external site. It can be used to link to
 * emails or phone numbers.
 *
 * The `children` of the link component is expected to be the text content.
 *
 * For `icon`, the `mdc-icon` component is used to render the icon.
 *
 * @dependency mdc-icon
 *
 * @tagname mdc-link
 *
 * @event click - (React: onClick) Fired when the user activates the Link using a mouse or assistive technology.
 * @event keydown - (React: onKeyDown) Fired when the user presses a key while the Link has focus.
 * @event focus - (React: onFocus) Fired when the Link receives keyboard or mouse focus.
 * @event blur - (React: onBlur) Fired when the Link loses keyboard or mouse focus.
 *
 * @csspart anchor - The anchor element that wraps the link content.
 * @csspart text - The element that wraps the link text.
 * @csspart icon - The icon element.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onFocus: 'focus',
        onBlur: 'blur',
    },
    displayName: 'Link',
});
export default reactWrapper;
