// @momentum-design/components/dist/react/buttonsimple/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/buttonsimple';
import { TAG_NAME } from '../../components/buttonsimple/buttonsimple.constants';
/**
 * `mdc-buttonsimple` is a simple button component that can be configured in various ways to suit different use cases.
 * It can be used standalone or as a base for more complex button components like `mdc-button`.
 *
 * ### Button Types
 *
 * The button type defines the behavior when clicked:
 *
 * - **button**: Standard button that performs an action on click
 * - **submit**: Submits the form data to the server
 * - **reset**: Resets the form data to its initial state
 *
 * ### Active State
 *
 * The `active` attribute toggles the button's active/toggled state,
 * automatically setting the appropriate ARIA state (e.g., `aria-pressed`).
 * This is useful for toggle buttons or buttons indicating a selected state.
 *
 * ### Disabled States
 *
 * - **disabled**: Completely disables the button, removing it from the tab order
 * - **soft-disabled**: Button appears disabled but remains focusable for accessibility
 *
 * @tagname mdc-buttonsimple
 *
 * @slot - Text label of the button.
 *
 * @event click - (React: onClick) This event is dispatched when the button is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the button.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the button.
 * @event focus - (React: onFocus) This event is dispatched when the button receives focus.
 *
 * @cssproperty --mdc-button-height - Height of the button
 * @cssproperty --mdc-button-background - Background color of the button
 * @cssproperty --mdc-button-border-color - Border color of the button
 * @cssproperty --mdc-button-text-color - Text color of the button
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onKeyUp: 'keyup',
        onFocus: 'focus',
    },
    displayName: 'Buttonsimple',
});
export default reactWrapper;
