// @momentum-design/components/dist/react/avatarbutton/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/avatarbutton';
import { TAG_NAME } from '../../components/avatarbutton/avatarbutton.constants';
/**
 * The `mdc-avatarbutton` component is an interactable version of the `mdc-avatar` component.
 *
 * This component is made by extending `buttonsimple` class.
 * The button component acts as a wrapper for the avatar component.
 *
 * To override styles of the avatar inside, use the specified css parts.
 *
 * @dependency mdc-avatar
 *
 * @event click - (React: onClick) This event is dispatched when the avatarbutton is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the avatarbutton.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the avatarbutton.
 * @event focus - (React: onFocus) This event is dispatched when the avatarbutton receives focus.
 *
 * @tagname mdc-avatarbutton
 *
 * @cssproperty --mdc-avatarbutton-overlay-background-color-rest - Background color of the overlay in rest state
 * @cssproperty --mdc-avatarbutton-overlay-background-color-hover - Background color of the overlay in hover state
 * @cssproperty --mdc-avatarbutton-overlay-background-color-active - Background color of the overlay in active state
 *
 * @csspart overlay - The overlay part of the avatar button.
 * @csspart content - The main content of the avatar.
 * @csspart photo - The photo part of the avatar.
 * @csspart presence - The presence indicator part of the avatar.
 * @csspart loading-wrapper - The wrapper for the loading indicator of the avatar.
 * @csspart loader - The loading indicator part of the avatar.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onKeyUp: 'keyup',
        onFocus: 'focus',
    },
    displayName: 'AvatarButton',
});
export default reactWrapper;
