// @momentum-design/components/dist/components/button/button.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property, state } from 'lit/decorators.js';
import { ButtonComponentMixin } from '../../utils/mixins/ButtonComponentMixin';
import { ROLE } from '../../utils/roles';
import Buttonsimple from '../buttonsimple/buttonsimple.component';
import { DEFAULTS } from './button.constants';
import styles from './button.styles';
import { getIconNameWithoutStyle } from './button.utils';
/**
 * `mdc-button` is a component that can be configured in various ways to suit different use cases.
 *
 * Button Variants:
 * - **Primary**: Solid background color.
 * - **Secondary**: Transparent background with a solid border.
 * - **Tertiary**: No background or border, appears as plain text but retains all button functionalities.
 *
 * Button Colors:
 * - **Positive**: Green color.
 * - **Negative**: Red color.
 * - **Accent**: Blue color.
 * - **Promotional**: Purple color.
 * - **Default**: White color.
 *
 * Button Sizes (in REM units):
 * - **Pill button**: 40, 32, 28, 24.
 * - **Icon button**: 64, 52, 40, 32, 28, 24.
 * - **Tertiary icon button**: 20.
 *
 * Button Types:
 * - **Pill button**: A button that contains text value. Commonly used for call to action, tags, or filters.
 * - **Pill button with icons**: A button containing an icon either on the left or right side of the button.
 * - **Icon button**: A button represented by just an icon without any text.
 * The type of button is inferred based on the presence of slot and/or prefix and postfix icons/slots.
 *
 * @dependency mdc-icon
 *
 * @tagname mdc-button
 *
 * @slot - Text label of the button.
 * @slot prefix - Content to be displayed before the text label.
 * @slot postfix - Content to be displayed after the text label.
 *
 * @csspart button-text - Text label of the button, passed in default slot
 * @csspart prefix - Content before the text label, passed in `prefix` slot
 * @csspart postfix - Content after the text label, passed in `postfix` slot
 */
class Button extends ButtonComponentMixin(Buttonsimple) {
    constructor() {
        super(...arguments);
        /**
         * Button sizing is based on the button type.
         * - **Pill button**: 40, 32, 28, 24.
         * - **Icon button**: 64, 52, 40, 32, 28, 24.
         * - Tertiary icon button can also be 20.
         * @default 32
         */
        this.size = DEFAULTS.SIZE;
        /**
         * The button color can be inverted by setting the inverted attribute to true.
         *
         * Only applies when variant is set to `primary`, color is set to `default`
         * and button is not `active`.
         * @default false
         */
        this.inverted = DEFAULTS.INVERTED;
        /**
         * This property defines the ARIA role for the element. By default, it is set to 'button'.
         * Consumers should override this role when:
         * - The element is being used in a context where a different role is more appropriate.
         * - Custom behaviors are implemented that require a specific ARIA role for accessibility purposes.
         * @default button
         */
        this.role = ROLE.BUTTON;
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('active')) {
            this.inferFilledIconName(this.active);
        }
        if (changedProperties.has('size')) {
            this.setSize(this.size);
        }
        if (changedProperties.has('variant')) {
            this.setVariant(this.variant);
            this.setSize(this.size);
        }
        if (changedProperties.has('color')) {
            this.setColor(this.color);
        }
        if (changedProperties.has('typeInternal')) {
            this.setSize(this.size);
        }
        if (changedProperties.has('prefixIcon') || changedProperties.has('postfixIcon')) {
            this.inferFilledIconName(this.active);
            this.inferButtonType();
        }
    }
    /**
     * Modifies the icon name based on the active state.
     * If the button is active, the icon name is suffixed with '-filled'.
     * If the button is inactive, the icon name is restored to its original value.
     * If '-filled' icon is not available, the icon name remains unchanged.
     *
     * @param active - The active state.
     */
    inferFilledIconName(active) {
        if (active) {
            if (this.prefixIcon) {
                this.prefixFilledIconName = `${getIconNameWithoutStyle(this.prefixIcon)}-filled`;
            }
            if (this.postfixIcon) {
                this.postfixFilledIconName = `${getIconNameWithoutStyle(this.postfixIcon)}-filled`;
            }
        }
        else {
            this.prefixFilledIconName = this.prefixIcon;
            this.postfixFilledIconName = this.postfixIcon;
        }
    }
    render() {
        return html `
      <slot name="prefix" part="prefix">
        ${this.prefixFilledIconName
            ? html ` <mdc-icon
              name="${this.prefixFilledIconName}"
              part="prefix-icon"
              length-unit="rem"
            ></mdc-icon>`
            : ''}
      </slot>
      <slot @slotchange=${this.inferButtonType} part="button-text"></slot>
      <slot name="postfix" part="postfix">
        ${this.postfixFilledIconName
            ? html ` <mdc-icon
              name="${this.postfixFilledIconName}"
              part="postfix-icon"
              length-unit="rem"
            ></mdc-icon>`
            : ''}
      </slot>
    `;
    }
}
Button.styles = [...Buttonsimple.styles, ...styles];
__decorate([
    property({ type: Number }),
    __metadata("design:type", Number)
], Button.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Boolean)
], Button.prototype, "inverted", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], Button.prototype, "role", void 0);
__decorate([
    state(),
    __metadata("design:type", String)
], Button.prototype, "prefixFilledIconName", void 0);
__decorate([
    state(),
    __metadata("design:type", String)
], Button.prototype, "postfixFilledIconName", void 0);
export default Button;
