// @momentum-design/components/dist/components/controltypeprovider/controltypeprovider.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import { Provider } from '../../models';
import ControlTypeProviderContext from './controltypeprovider.context';
import { DEFAULTS, VALID_VALUES } from './controltypeprovider.constants';
/**
 * ContolTypeProvider component, which allows to be consumed from sub components
 * (see `providerUtils.consume` for how to consume)
 *
 * attribute 'control-type' sets the default control type for all consumers
    - 'controlled' the consumer component will not handle any interaction itself, e.g. toggling a checkbox.
    - 'uncontrolled' the consumer component will handle interactions
 *
 * ControlTypeMixin allows components to consume this context
 * note that this is intended to be set once, ControlTypeMixin will not update on changes to control-type
 *
 * @tagname mdc-controltypeprovider
 *
 */
class ControlTypeProvider extends Provider {
    constructor() {
        // initialise the context by running the Provider constructor:
        super({
            context: ControlTypeProviderContext,
        });
        this.innerControlType = DEFAULTS.CONTROL_TYPE;
    }
    /**
     * Context object of the ControlTypeProvider, to be consumed by child components
     */
    static get Context() {
        return ControlTypeProviderContext;
    }
    /**
     * Whether the control type is 'controlled' or 'uncontrolled'
     * @default 'uncontrolled'
     */
    set controlType(value) {
        if (VALID_VALUES.includes(value)) {
            this.innerControlType = value;
        }
    }
    get controlType() {
        return this.innerControlType;
    }
    updateContext() {
        if (this.context.value !== this.controlType) {
            this.context.value = this.controlType;
            this.context.updateObservers();
        }
    }
}
__decorate([
    property({ type: String, attribute: 'control-type', reflect: true }),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [String])
], ControlTypeProvider.prototype, "controlType", null);
export default ControlTypeProvider;
