// @momentum-design/components/dist/react/menuitem/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/menuitem';
import { TAG_NAME } from '../../components/menuitem/menuitem.constants';
/**
 * menuitem component is inherited by listitem component with the role set `menuitem`.<br/>
 * A menu item can contain an icon on the leading or trailing side.
 *
 * The leading and trailing slots can be used to display controls and text.<br/>
 * Based on the leading/trailing slot, the position of the controls and text can be adjusted.
 *
 * Please use element with role=menu as a parent element even when there is only menuitem for a11y purpose.
 * For example mdc-menupopover or mdc-menubar.
 *
 * Menu item has `name` and `value` attribute that can be used to identify the menu item when it is selected.
 *
 * @dependency mdc-text
 * @dependency mdc-icon
 * @dependency mdc-tooltip
 *
 * @tagname mdc-menuitem
 *
 * @slot leading-controls - slot for menu item controls to appear of leading end.
 * @slot leading-text-primary-label - slot for menu item primary label.
 * @slot leading-text-secondary-label - slot for menu item secondary label.
 * @slot leading-text-tertiary-label - slot for menu item tertiary label.
 * @slot trailing-text-side-header - slot for menu item side header text.
 * @slot trailing-text-subline - slot for menu item subline text.
 * @slot trailing-controls - slot for menu item controls to appear of trailing end.
 *
 * @event click - (React: onClick) This event is dispatched when the menuitem is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the menuitem.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the menuitem.
 * @event focus - (React: onFocus) This event is dispatched when the menuitem receives focus.
 * @event enabled - (React: onEnabled) This event is dispatched after the menuitem is enabled
 * @event disabled - (React: onDisabled) This event is dispatched after the menuitem is disabled
 * @event created - (React: onCreated) This event is dispatched after the menuitem is created (added to the DOM)
 * @event destroyed - (React: onDestroyed) This event is dispatched after the menuitem is destroyed (removed from the DOM)
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onKeyUp: 'keyup',
        onFocus: 'focus',
        onEnabled: 'enabled',
        onDisabled: 'disabled',
        onCreated: 'created',
        onDestroyed: 'destroyed',
    },
    displayName: 'MenuItem',
});
export default reactWrapper;
