// @momentum-design/components/dist/components/navmenuitem/navmenuitem.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import providerUtils from '../../utils/provider';
import { TYPE, VALID_TEXT_TAGS } from '../text/text.constants';
import { TAG_NAME as MENUPOPOVER_TAGNAME } from '../menupopover/menupopover.constants';
import { IconNameMixin } from '../../utils/mixins/IconNameMixin';
import MenuItem from '../menuitem/menuitem.component';
import { getIconNameWithoutStyle } from '../button/button.utils';
import SideNavigation from '../sidenavigation/sidenavigation.component';
import { DEFAULTS, ALLOWED_BADGE_TYPES, ICON_NAME } from './navmenuitem.constants';
import styles from './navmenuitem.styles';
/**
 * `mdc-navmenuitem` is a menuitem styled to work as a navigation tab.
 * It supports a leading icon, optional badge and dynamic text rendering.
 *
 * Note: `mdc-navmenuitem` is intended to be used inside `mdc-menubar` as part of the sideNavigation
 * component. Its structure, spacing, and interactions are designed to align with
 * the visual and functional requirements of side navigation layouts.
 *
 * @tagname mdc-navmenuitem
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-badge
 * @dependency mdc-tooltip
 *
 * @event click - (React: onClick) This event is dispatched when the navmenuitem is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the navmenuitem.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the navmenuitem.
 * @event focus - (React: onFocus) This event is dispatched when the navmenuitem receives focus.
 * @event activechange - (React: onActiveChange) Dispatched when the active state of the navmenuitem changes.
 *
 * @cssproperty --mdc-navmenuitem-color - Text color of the navmenuitem in its normal state.
 * @cssproperty --mdc-navmenuitem-border-color - Border color of the navmenuitem in its normal state.
 * @cssproperty --mdc-navmenuitem-disabled-color - Text color of the navmenuitem when disabled.
 * @cssproperty --mdc-navmenuitem-expanded-width - Width of the navmenuitem when expanded.
 * @cssproperty --mdc-navmenuitem-hover-background-color - Background color of the navmenuitem when hovered.
 * @cssproperty --mdc-navmenuitem-hover-active-background-color - Background color of the active navmenuitem when hovered.
 * @cssproperty --mdc-navmenuitem-pressed-background-color - Background color of the navmenuitem when pressed.
 * @cssproperty --mdc-navmenuitem-pressed-active-background-color - Background color of the active navmenuitem when pressed.
 * @cssproperty --mdc-navmenuitem-disabled-background-color - Background color of the navmenuitem when disabled.
 * @cssproperty --mdc-navmenuitem-disabled-active-background-color - Background color of the active navmenuitem when disabled.
 * @cssproperty --mdc-navmenuitem-rest-active-background-color - Background color of the active nav item in its rest state.
 */
class NavMenuItem extends IconNameMixin(MenuItem) {
    constructor() {
        super();
        /**
         * The maximum number can be set up to 999, anything above that will be rendered as _999+_.
         * The max counter can be `9`, `99` or `999`.
         * @default 99
         */
        this.maxCounter = DEFAULTS.MAX_COUNTER;
        /**
         * Aria-label attribute to be set for accessibility
         */
        this.ariaLabel = null;
        /**
         * @internal
         */
        this.sideNavigationContext = providerUtils.consume({ host: this, context: SideNavigation.Context });
        /**
         * Dispatch the activechange event.
         * @internal
         * @param active - The active state of the navMenuItem.
         */
        this.emitNavMenuItemActiveChange = (active) => {
            const event = new CustomEvent('activechange', {
                detail: { navId: this.navId, active },
                bubbles: true,
            });
            this.dispatchEvent(event);
        };
        this.addEventListener('click', this.handleClickEvent.bind(this));
    }
    connectedCallback() {
        super.connectedCallback();
        this.variant = undefined;
        this.tooltipPlacement = undefined;
        if (!this.navId && this.onerror) {
            this.onerror('[mdc-navmenuitem] navId is required and was not provided.');
        }
        // Set in-menupopover attribute if nested
        this.toggleAttribute('in-menupopover', this.isNested());
    }
    updated() {
        var _a, _b;
        const context = (_a = this.sideNavigationContext) === null || _a === void 0 ? void 0 : _a.value;
        if (!context)
            return;
        // Determine expansion state
        this.showLabel = this.isNested() ? true : context.expanded;
        // Manage aria-label for accessibility
        if (this.showLabel) {
            this.removeAttribute('aria-label');
        }
        else {
            const label = (_b = this.label) !== null && _b !== void 0 ? _b : '';
            this.ariaLabel = this.ariaLabel || label;
            this.setAttribute('aria-label', label);
        }
    }
    /**
     * Check whether the navmenuitem is inside a nested nav structure.
     * Returns `true` if there is a parent `mdc-menupopover`
     * This method assumes nesting implies deeper levels of nav hierarchy.
     */
    isNested() {
        let parent = this.parentElement;
        while (parent) {
            if (parent.tagName.toLowerCase() === MENUPOPOVER_TAGNAME) {
                return true;
            }
            parent = parent.parentElement;
        }
        return false;
    }
    /**
     * Modifies the icon name based on the active state.
     * If the navMenuItem is active, the icon name is suffixed with '-filled'.
     * If the navMenuItem is inactive, the icon name is restored to its original value.
     * If '-filled' icon is not available, the icon name remains unchanged.
     * @internal
     * @param active - The active state.
     */
    modifyIconName(active) {
        if (!this.iconName)
            return;
        const isFilled = this.iconName.endsWith('-filled');
        const baseIcon = getIconNameWithoutStyle(this.iconName);
        if (active) {
            if (!isFilled) {
                this.prevIconName = this.iconName;
                this.iconName = `${baseIcon}-filled`;
            }
        }
        else if (this.prevIconName) {
            this.iconName = this.prevIconName;
        }
    }
    handleClickEvent() {
        if (this.disabled)
            return;
        this.emitNavMenuItemActiveChange(this.active);
    }
    update(changedProperties) {
        super.update(changedProperties);
        if (changedProperties.has('active')) {
            this.modifyIconName(this.active);
        }
    }
    renderTextLabel(label) {
        return html `
      <mdc-text
        type=${this.active ? TYPE.BODY_MIDSIZE_BOLD : TYPE.BODY_MIDSIZE_MEDIUM}
        tagname=${VALID_TEXT_TAGS.SPAN}
        part="text-container"
      >
        ${label}
      </mdc-text>
    `;
    }
    renderArrowIcon(showLabel) {
        const arrowClass = showLabel ? '' : 'arrow';
        return html `
      <mdc-icon name=${ICON_NAME.RIGHT_ARROW} length-unit="rem" part="trailing-arrow" class="${arrowClass}"> </mdc-icon>
    `;
    }
    renderBadge(showLabel) {
        const badgeClass = showLabel ? '' : 'badge';
        const isValidBadgeType = Object.values(ALLOWED_BADGE_TYPES).includes(this.badgeType);
        if (!isValidBadgeType) {
            return nothing;
        }
        return html `
      <mdc-badge
        class="${badgeClass}"
        type="${ifDefined(this.badgeType)}"
        counter="${ifDefined(this.counter)}"
        max-counter="${this.maxCounter}"
      >
      </mdc-badge>
    `;
    }
    render() {
        var _a;
        const context = (_a = this.sideNavigationContext) === null || _a === void 0 ? void 0 : _a.value;
        return html `
      <div part="icon-container">
        <mdc-icon name="${this.iconName}" size="1.5" length-unit="rem" part="icon"></mdc-icon>
        ${!this.showLabel ? this.renderBadge(this.showLabel) : nothing}
      </div>
      ${this.showLabel ? html `${this.renderTextLabel(this.label)}${this.renderBadge(this.showLabel)}` : nothing}
      ${(context === null || context === void 0 ? void 0 : context.hasSiblingWithTriggerId(this)) ? this.renderArrowIcon(this.showLabel) : nothing}
    `;
    }
}
NavMenuItem.styles = [...MenuItem.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "active", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'badge-type' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "badgeType", void 0);
__decorate([
    property({ type: Number, reflect: true }),
    __metadata("design:type", Number)
], NavMenuItem.prototype, "counter", void 0);
__decorate([
    property({ type: Number, attribute: 'max-counter', reflect: true }),
    __metadata("design:type", Number)
], NavMenuItem.prototype, "maxCounter", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'nav-id' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "navId", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'show-label' }),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "showLabel", void 0);
__decorate([
    property({ type: String, attribute: 'aria-label' }),
    __metadata("design:type", Object)
], NavMenuItem.prototype, "ariaLabel", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'disable-aria-current' }),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "disableAriaCurrent", void 0);
export default NavMenuItem;
