// @momentum-design/components/dist/components/navmenuitem/navmenuitem.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property, state } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { v4 } from 'uuid';
import { TYPE, VALID_TEXT_TAGS } from '../text/text.constants';
import { TAG_NAME as MENUPOPOVER_TAGNAME } from '../menupopover/menupopover.constants';
import MenuItem from '../menuitem/menuitem.component';
import providerUtils from '../../utils/provider';
import SideNavigation from '../sidenavigation/sidenavigation.component';
import { TAG_NAME as TOOLTIP_TAG_NAME } from '../tooltip/tooltip.constants';
import { getIconNameWithoutStyle } from '../button/button.utils';
import { ALLOWED_BADGE_TYPES, DEFAULTS, ICON_NAME } from './navmenuitem.constants';
import styles from './navmenuitem.styles';
/**
 * `mdc-navmenuitem` is a menuitem styled to work as a navigation tab.
 * It supports a leading icon, optional badge and dynamic text rendering.
 *
 * The navmenuitem can be in an active or inactive state, indicating whether it is currently selected.
 *
 * Note: `mdc-navmenuitem` is intended to be used inside `mdc-menubar` as part of the sideNavigation
 * component. Its structure, spacing, and interactions are designed to align with
 * the visual and functional requirements of side navigation layouts.
 *
 * By providing the tooltip-text attribute, a tooltip will be displayed on hover of the navmenuitem. This tooltip is useful when an active navmenuitem is present within a submenu of the sidenavigation component.
 *
 * @tagname mdc-navmenuitem
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-badge
 * @dependency mdc-tooltip
 *
 * @event click - (React: onClick) This event is dispatched when the navmenuitem is clicked.
 * @event keydown - (React: onKeyDown) This event is dispatched when a key is pressed down on the navmenuitem.
 * @event keyup - (React: onKeyUp) This event is dispatched when a key is released on the navmenuitem.
 * @event focus - (React: onFocus) This event is dispatched when the navmenuitem receives focus.
 * @event activechange - (React: onActiveChange) Dispatched when the active state of the navmenuitem changes.
 *
 * @cssproperty --mdc-navmenuitem-in-sidenav-expanded-width - Width of the navmenuitem when expanded.
 * @cssproperty --mdc-navmenuitem-in-sidenav-expanded-margin-left - Left margin of the navmenuitem, when expanded.
 * @cssproperty --mdc-navmenuitem-in-sidenav-expanded-margin-right - Right margin of the navmenuitem, when expanded.
 * @cssproperty --mdc-navmenuitem-in-sidenav-collapsed-width - Width of the navmenuitem when collapsed.
 * @cssproperty --mdc-navmenuitem-in-sidenav-collapsed-margin-left - Left margin of the navmenuitem, when collapsed.
 * @cssproperty --mdc-navmenuitem-in-sidenav-collapsed-margin-right - Right margin of the navmenuitem, when collapsed.
 * @cssproperty --mdc-navmenuitem-color - Text color of the navmenuitem in its normal state.
 * @cssproperty --mdc-navmenuitem-disabled-color - Text color of the navmenuitem when disabled.
 * @cssproperty --mdc-navmenuitem-rest-active-background-color - Background color of the active nav item in its rest state.
 * @cssproperty --mdc-navmenuitem-hover-background-color - Background color of the navmenuitem when hovered.
 * @cssproperty --mdc-navmenuitem-hover-active-background-color - Background color of the active navmenuitem when hovered.
 * @cssproperty --mdc-navmenuitem-pressed-background-color - Background color of the navmenuitem when pressed.
 * @cssproperty --mdc-navmenuitem-pressed-active-background-color - Background color of the active navmenuitem when pressed.
 * @cssproperty --mdc-navmenuitem-disabled-background-color - Background color of the navmenuitem when disabled.
 * @cssproperty --mdc-navmenuitem-disabled-active-background-color - Background color of the active navmenuitem when disabled.
 *
 * @csspart arrow - The arrow of the navmenuitem.
 * @csspart badge - The badge of the navmenuitem.
 * @csspart icon-container - The container of the icon.
 * @csspart text-container - The container of the text.
 * @csspart trailing-arrow - The trailing arrow of the navmenuitem.
 */
class NavMenuItem extends MenuItem {
    constructor() {
        super();
        /**
         * When set to true, prevents the navmenuitem from being activated.
         * This is useful for cases where the navmenuitem is used as a button
         * or link, where not the content on the current page is changing.
         * By default this is set to false, making the navmenuitem activatable (= aria-current gets set when active).
         *
         * @default false
         */
        this.cannotActivate = DEFAULTS.CANNOT_ACTIVATE;
        /**
         * The maximum number can be set up to 999, anything above that will be rendered as _999+_.
         * The max counter can be `9`, `99` or `999`.
         * @default 99
         */
        this.maxCounter = DEFAULTS.MAX_COUNTER;
        /**
         * Aria-label attribute to be set for accessibility
         */
        this.ariaLabel = null;
        /**
         * The placement of the tooltip relative to the nav item.
         * @default 'right'
         */
        this.tooltipPlacement = 'right';
        /**
         * The appearance behavior of the tooltip.
         * Options are 'when-collapsed' (default) or 'always'.
         *
         * @default 'when-collapsed'
         */
        this.tooltipAppearance = 'when-collapsed';
        /**
         * Indicates whether this navmenuitem has an active child navmenuitem.
         * This is used to manage tooltip display for parent navmenuitems in nested navigation structures.
         * @internal
         */
        this.hasActiveChild = false;
        /**
         * @internal
         */
        this.sideNavigationContext = providerUtils.consume({ host: this, context: SideNavigation.Context });
        /**
         * Dispatch the activechange event.
         * @internal
         * @param active - The active state of the navMenuItem.
         */
        this.emitNavMenuItemActiveChange = (active) => {
            const event = new CustomEvent('activechange', {
                detail: { navId: this.navId, active },
                bubbles: true,
            });
            this.dispatchEvent(event);
        };
        this.addEventListener('click', this.handleClickEvent.bind(this));
    }
    connectedCallback() {
        super.connectedCallback();
        this.variant = undefined;
        if (!this.navId && this.onerror) {
            this.onerror('[mdc-navmenuitem] navId is required and was not provided.');
        }
        // Set in-menupopover attribute if nested
        this.toggleAttribute('in-menupopover', this.isNested());
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeTooltip();
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        // if active on initial render, set aria-current
        if (this.active && !this.disableAriaCurrent && !this.cannotActivate) {
            this.setAttribute('aria-current', 'page');
        }
    }
    updated(changedProperties) {
        var _a;
        super.updated(changedProperties);
        if (changedProperties.has('tooltipText') ||
            changedProperties.has('showLabel') ||
            changedProperties.has('hasActiveChild')) {
            this.renderDynamicTooltip();
        }
        if (changedProperties.has('showLabel')) {
            // If collapsed and aria-label is not set, use visible label
            if (!this.showLabel && !this.getAttribute('aria-label') && this.label) {
                this.setAttribute('aria-label', this.label);
            }
        }
        const context = (_a = this.sideNavigationContext) === null || _a === void 0 ? void 0 : _a.value;
        if (!context)
            return;
        // Determine expansion state
        this.showLabel = this.isNested() ? true : context.expanded;
    }
    renderDynamicTooltip() {
        var _a;
        this.removeTooltip();
        if (this.hasActiveChild && !this.isActiveParentTooltipText) {
            return;
        }
        if (!this.hasActiveChild &&
            (!this.tooltipText || (this.tooltipAppearance === 'when-collapsed' && this.showLabel))) {
            return;
        }
        if (!this.id) {
            this.id = `mdc-navmenuitem-${v4()}`;
        }
        const tooltip = document.createElement(TOOLTIP_TAG_NAME);
        tooltip.id = `mdc-navmenuitem-tooltip-${v4()}`;
        tooltip.textContent = this.hasActiveChild ? this.isActiveParentTooltipText : this.tooltipText;
        tooltip.setAttribute('triggerid', this.id);
        tooltip.setAttribute('show-arrow', '');
        tooltip.setAttribute('placement', this.tooltipPlacement);
        if (this.tooltipType) {
            tooltip.setAttribute('tooltip-type', this.tooltipType);
        }
        tooltip.setAttribute('boundary-padding', ((_a = this.tooltipBoundaryPadding) === null || _a === void 0 ? void 0 : _a.toString()) || '0');
        // Set the slot attribute if the parent element has a slot.
        if (this.hasAttribute('slot')) {
            tooltip.setAttribute('slot', this.getAttribute('slot') || '');
        }
        // Attach the tooltip programmatically after the nearest parent element.
        this.after(tooltip);
    }
    removeTooltip() {
        // Remove any existing tooltip.
        const existingTooltip = document.querySelector(`${TOOLTIP_TAG_NAME}[triggerid="${this.id}"]`);
        if (existingTooltip) {
            existingTooltip.remove();
        }
    }
    /**
     * Check whether the navmenuitem is inside a nested nav structure.
     * Returns `true` if there is a parent `mdc-menupopover`
     * This method assumes nesting implies deeper levels of nav hierarchy.
     */
    isNested() {
        let parent = this.parentElement;
        while (parent) {
            if (parent.tagName.toLowerCase() === MENUPOPOVER_TAGNAME) {
                return true;
            }
            parent = parent.parentElement;
        }
        return false;
    }
    handleClickEvent() {
        if (this.disabled || this.cannotActivate)
            return;
        this.emitNavMenuItemActiveChange(this.active);
    }
    getFilledIconName() {
        if (!this.iconName) {
            return undefined;
        }
        const isFilled = this.iconName.endsWith('-filled');
        if (isFilled) {
            return undefined;
        }
        const baseIcon = getIconNameWithoutStyle(this.iconName);
        return `${baseIcon}-filled`;
    }
    renderBadge(showLabel) {
        const isValidBadgeType = Object.values(ALLOWED_BADGE_TYPES).includes(this.badgeType);
        if (!isValidBadgeType) {
            return nothing;
        }
        return html `
      <mdc-badge
        part="${showLabel ? '' : 'badge'}"
        type="${ifDefined(this.badgeType)}"
        counter="${ifDefined(this.counter)}"
        max-counter="${this.maxCounter}"
        aria-hidden="true"
      >
      </mdc-badge>
    `;
    }
    render() {
        var _a;
        const context = (_a = this.sideNavigationContext) === null || _a === void 0 ? void 0 : _a.value;
        return html `
      <div part="icon-container">
        <mdc-icon name="${this.iconName}" size="1.5" length-unit="rem" part="regular-icon"></mdc-icon>
        ${!this.cannotActivate
            ? html `<mdc-icon
              name="${this.getFilledIconName()}"
              size="1.5"
              length-unit="rem"
              part="filled-icon"
            ></mdc-icon>`
            : nothing}
        ${!this.showLabel ? this.renderBadge(this.showLabel) : nothing}
      </div>
      ${this.showLabel
            ? html `
            <mdc-text
              type=${this.active ? TYPE.BODY_MIDSIZE_BOLD : TYPE.BODY_MIDSIZE_MEDIUM}
              tagname=${VALID_TEXT_TAGS.SPAN}
              part="text-container"
            >
              ${this.label}
            </mdc-text>
            ${this.renderBadge(this.showLabel)}
          `
            : nothing}
      ${(context === null || context === void 0 ? void 0 : context.hasSiblingWithTriggerId(this))
            ? html ` <mdc-icon name=${ICON_NAME.RIGHT_ARROW} length-unit="rem" part="trailing-arrow"> </mdc-icon>`
            : nothing}
    `;
    }
}
NavMenuItem.styles = [...MenuItem.styles, ...styles];
__decorate([
    property({ type: Boolean, reflect: true }),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "active", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'cannot-activate' }),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "cannotActivate", void 0);
__decorate([
    property({ type: String, attribute: 'icon-name', reflect: true }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "iconName", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'badge-type' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "badgeType", void 0);
__decorate([
    property({ type: Number, reflect: true }),
    __metadata("design:type", Number)
], NavMenuItem.prototype, "counter", void 0);
__decorate([
    property({ type: Number, attribute: 'max-counter', reflect: true }),
    __metadata("design:type", Number)
], NavMenuItem.prototype, "maxCounter", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'nav-id' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "navId", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'show-label' }),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "showLabel", void 0);
__decorate([
    property({ type: String, attribute: 'aria-label' }),
    __metadata("design:type", Object)
], NavMenuItem.prototype, "ariaLabel", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'disable-aria-current' }),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "disableAriaCurrent", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'tooltip-text' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "tooltipText", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'tooltip-placement' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "tooltipPlacement", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'tooltip-type' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "tooltipType", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'tooltip-appearance' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "tooltipAppearance", void 0);
__decorate([
    property({ type: Number, reflect: true, attribute: 'tooltip-boundary-padding' }),
    __metadata("design:type", Number)
], NavMenuItem.prototype, "tooltipBoundaryPadding", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'is-active-parent-tooltip-text' }),
    __metadata("design:type", String)
], NavMenuItem.prototype, "isActiveParentTooltipText", void 0);
__decorate([
    state(),
    __metadata("design:type", Boolean)
], NavMenuItem.prototype, "hasActiveChild", void 0);
export default NavMenuItem;
