// @momentum-design/components/dist/components/password/password.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html, nothing } from 'lit';
import { property, state } from 'lit/decorators.js';
import Input from '../input/input.component';
import { INPUT_TYPE } from '../input/input.constants';
import { BUTTON_VARIANTS } from '../button/button.constants';
import { DEFAULTS } from '../formfieldwrapper/formfieldwrapper.constants';
import { PASSWORD_VISIBILITY_ICONS } from './password.constants';
/**
 * `mdc-password` is a component that allows users to input their password.
 *  It extends the `mdc-input` component and provides additional features specific to password fields.
 *  It contains:
 * - `label` field - describe the password field.
 * - `password` field - contains the value
 * - `help-text` or `validation-message` - displayed below the password field.
 * - `help-text-type` - type of the help text, can be 'default', 'error', 'warning', 'success', 'priority'.
 * - `show-hide-button-aria-label` - aria label for the trailing show/hide button.
 * - all the attributes of the native password field.
 *
 * @tagname mdc-password
 *
 * @event input - (React: onInput) This event is dispatched when the value of the password field changes (every press).
 * @event change - (React: onChange) This event is dispatched when the value of the password field changes (on blur).
 * @event focus - (React: onFocus) This event is dispatched when the password receives focus.
 * @event blur - (React: onBlur) This event is dispatched when the password loses focus.
 *
 * @dependency mdc-icon
 * @dependency mdc-text
 * @dependency mdc-button
 * @dependency mdc-toggletip
 *
 * @cssproperty --mdc-input-disabled-border-color - Border color for the password container when disabled
 * @cssproperty --mdc-input-disabled-text-color - Text color for the password field when disabled
 * @cssproperty --mdc-input-disabled-background-color - Background color for the password field when disabled
 * @cssproperty --mdc-input-border-color - Border color for the password container
 * @cssproperty --mdc-input-text-color - Text color for the password field
 * @cssproperty --mdc-input-background-color - Background color for the password field
 * @cssproperty --mdc-input-selection-background-color - Background color for the selected text
 * @cssproperty --mdc-input-selection-text-color - Text color for the selected text
 * @cssproperty --mdc-input-support-text-color - Text color for the help text
 * @cssproperty --mdc-input-hover-background-color - Background color for the password field when hovered
 * @cssproperty --mdc-input-focused-background-color - Background color for the password field when focused
 * @cssproperty --mdc-input-focused-border-color - Border color for the password container when focused
 * @cssproperty --mdc-input-error-border-color - Border color for the password container when error
 * @cssproperty --mdc-input-warning-border-color - Border color for the password container when warning
 * @cssproperty --mdc-input-success-border-color - Border color for the password container when success
 * @cssproperty --mdc-input-primary-border-color - Border color for the password container when primary
 *
 */
class Password extends Input {
    constructor() {
        super(...arguments);
        /**
         * Aria label for the show or hide password icon button.
         */
        this.showHideButtonAriaLabel = '';
        /**
         * The type of help text. It can be 'default', 'error', 'warning', 'success', 'priority'.
         * @override
         */
        this.helpTextType = DEFAULTS.VALIDATION;
        /**
         * Internal state to track whether the password is visible (shown as text) or hidden (shown as password).
         */
        this.showPassword = false;
    }
    connectedCallback() {
        super.connectedCallback();
        // Ensure the trailing button is always true for password inputs
        this.trailingButton = true;
    }
    /**
     * Toggles the visibility of the password.
     */
    toggleShowPassword() {
        this.showPassword = !this.showPassword;
    }
    /**
     * Renders the trailing button for showing or hiding the password.
     */
    renderTrailingButton(show = false) {
        const showBtn = show || this.value;
        if (!showBtn) {
            return nothing;
        }
        return html `
      <mdc-button
        part="trailing-button"
        class="own-focus-ring ${!showBtn ? 'hidden' : ''}"
        variant=${BUTTON_VARIANTS.TERTIARY}
        ?disabled=${this.disabled || this.readonly || !showBtn}
        size="${DEFAULTS.ICON_SIZE}"
        @click=${this.toggleShowPassword}
        aria-label=${this.showHideButtonAriaLabel}
        prefix-icon=${this.showPassword ? PASSWORD_VISIBILITY_ICONS.HIDE_BOLD : PASSWORD_VISIBILITY_ICONS.SHOW_BOLD}
      ></mdc-button>
    `;
    }
    /**
     * Renders the input element for the password field.
     */
    renderInputElement(_, hidePlaceholder) {
        const inputType = this.showPassword ? INPUT_TYPE.TEXT : INPUT_TYPE.PASSWORD;
        return super.renderInputElement(inputType, hidePlaceholder);
    }
}
Password.styles = [...Input.styles];
__decorate([
    property({ type: String, attribute: 'show-hide-button-aria-label' }),
    __metadata("design:type", Object)
], Password.prototype, "showHideButtonAriaLabel", void 0);
__decorate([
    property({ type: String, attribute: 'help-text-type' }),
    __metadata("design:type", String)
], Password.prototype, "helpTextType", void 0);
__decorate([
    state(),
    __metadata("design:type", Object)
], Password.prototype, "showPassword", void 0);
export default Password;
