// @momentum-design/components/dist/components/stepperconnector/stepperconnector.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { Component } from '../../models';
import providerUtils from '../../utils/provider';
import Stepper from '../stepper/stepper.component';
import { DEFAULTS } from './stepperconnector.constants';
import styles from './stepperconnector.styles';
/**
 * StepperConnector component visually represents the connection between two stepper items.
 * Indicates whether the connection is complete or incomplete, and supports vertical or horizontal orientation.
 * They are used between 2 `mdc-stepperitem` components to visually connect them and wrapped in a `mdc-stepper` component.
 *
 * @tagname mdc-stepperconnector
 *
 * @csspart connector - The main connector line between steps
 *
 * @cssproperty --mdc-stepperconnector-complete-background - Background color for the complete connector
 * @cssproperty --mdc-stepperconnector-incomplete-background - Background color for the incomplete connector
 */
class StepperConnector extends Component {
    constructor() {
        super(...arguments);
        /**
         * The status of the connector (complete or incomplete)
         * @default "incomplete"
         */
        this.status = DEFAULTS.STATUS;
        /**
         * The orientation of the connector (vertical or horizontal)
         * @default "horizontal"
         */
        this.orientation = DEFAULTS.ORIENTATION;
        /**
         * @internal
         */
        this.stepperContext = providerUtils.consume({ host: this, context: Stepper.Context });
    }
    updated(changedProperties) {
        var _a;
        super.updated(changedProperties);
        const context = (_a = this.stepperContext) === null || _a === void 0 ? void 0 : _a.value;
        if (!context || !context.orientation)
            return;
        this.orientation = context.orientation;
    }
    render() {
        return html ` <div part="connector"></div> `;
    }
}
StepperConnector.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], StepperConnector.prototype, "status", void 0);
__decorate([
    property({ type: String, reflect: true }),
    __metadata("design:type", String)
], StepperConnector.prototype, "orientation", void 0);
export default StepperConnector;
