// @momentum-design/components/dist/utils/mixins/ControlTypeMixin.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import ControlTypeProvider from '../../components/controltypeprovider';
import providerUtils from '../provider';
import { DEFAULTS, VALID_VALUES } from '../../components/controltypeprovider/controltypeprovider.constants';
export const ControlTypeMixin = (superClass) => {
    // The tests for this mixin are in the menuitemcheckbox component
    class InnerMixinClass extends superClass {
        constructor() {
            super(...arguments);
            this.controlTypeProviderContext = providerUtils.consume({
                host: this,
                context: ControlTypeProvider.Context,
            });
        }
        willUpdate() {
            if (!(this.controlType && VALID_VALUES.includes(this.controlType))) {
                if (this.controlTypeProviderContext.value) {
                    this.controlType = this.controlTypeProviderContext.value;
                }
                else {
                    this.controlType = DEFAULTS.CONTROL_TYPE;
                }
            }
        }
    }
    __decorate([
        property({ type: String, attribute: 'control-type', reflect: true }),
        __metadata("design:type", String)
    ], InnerMixinClass.prototype, "controlType", void 0);
    // Cast return type to your mixin's interface intersected with the superClass type
    return InnerMixinClass;
};
