// @momentum-design/components/dist/utils/mixins/FormInternalsMixin.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property, query } from 'lit/decorators.js';
export const FormInternalsMixin = (superClass) => {
    class InnerMixinClass extends superClass {
        constructor() {
            super(...arguments);
            /**
             * Indicates the name of the component group.
             * They are used to group elements in a form together.
             * @default ''
             */
            this.name = '';
            /**
             * Indicates the value of the component group (ex: input, checkbox, radio, select etc...)
             * @default ''
             */
            this.value = '';
            /** @internal */
            this.internals = this.attachInternals();
        }
        /** @internal */
        get form() {
            return this.internals.form;
        }
        get validity() {
            return this.internals.validity;
        }
        get willValidate() {
            return this.internals.willValidate;
        }
        /**
         * Sets the validity of the input field based on the input field's validity.
         * @returns void
         */
        setValidity() {
            if (this.inputElement) {
                this.internals.setValidity(this.inputElement.validity, this.inputElement.validationMessage, this.inputElement);
            }
        }
        checkValidity() {
            this.setValidity();
            return this.internals.checkValidity();
        }
        reportValidity() {
            this.setValidity();
            return this.internals.reportValidity();
        }
    }
    /** @internal */
    InnerMixinClass.formAssociated = true;
    __decorate([
        property({ reflect: true, type: String }),
        __metadata("design:type", Object)
    ], InnerMixinClass.prototype, "name", void 0);
    __decorate([
        property({ reflect: true, type: String }),
        __metadata("design:type", Object)
    ], InnerMixinClass.prototype, "value", void 0);
    __decorate([
        property({ reflect: true, type: String, attribute: 'validation-message' }),
        __metadata("design:type", String)
    ], InnerMixinClass.prototype, "validationMessage", void 0);
    __decorate([
        query('input'),
        __metadata("design:type", Object)
    ], InnerMixinClass.prototype, "inputElement", void 0);
    // Cast return type to your mixin's interface intersected with the superClass type
    return InnerMixinClass;
};
