// @momentum-design/components/dist/utils/mixins/KeyToActionMixin.js
import { KEYS } from '../keys';
import providerUtils from '../provider';
import SpatialNavigationProvider from '../../components/spatialnavigationprovider';
export const ACTIONS = {
    /** Action key, e.g., Enter */
    ENTER: 'enter',
    /** Back key, e.g., Escape/Back/Cancel */
    ESCAPE: 'escape',
    /** Navigation key up */
    UP: 'up',
    /** Navigation key down */
    DOWN: 'down',
    /** Navigation key left */
    LEFT: 'left',
    /** Navigation key right */
    RIGHT: 'right',
    /** Space key, some actions and scrolling */
    SPACE: 'space',
    /** Tab key */
    TAB: 'tab',
    /** Home key */
    HOME: 'home',
    /** End key */
    END: 'end',
    /** Backspace key */
    BACKSPACE: 'backspace',
};
/**
 * Default key to action mapping if no spatial navigation context is available
 */
const DEFAULT_KEY_TO_ACTION = {
    [KEYS.ARROW_UP]: ACTIONS.UP,
    [KEYS.ARROW_DOWN]: ACTIONS.DOWN,
    [KEYS.ARROW_LEFT]: ACTIONS.LEFT,
    [KEYS.ARROW_RIGHT]: ACTIONS.RIGHT,
    [KEYS.ENTER]: ACTIONS.ENTER,
    [KEYS.SPACE]: ACTIONS.SPACE,
    [KEYS.ESCAPE]: ACTIONS.ESCAPE,
    [KEYS.TAB]: ACTIONS.TAB,
    [KEYS.HOME]: ACTIONS.HOME,
    [KEYS.END]: ACTIONS.END,
    [KEYS.BACKSPACE]: ACTIONS.BACKSPACE,
};
export const NAV_MODES = {
    /** Normal navigation mode, focus moved with Tab and Shift+Tab keys */
    DEFAULT: 'default',
    /** Spatial navigation mode, focus moved with arrow keys */
    SPATIAL: 'spatial',
};
/**
 * Mixin to provide abstract key mapping for navigation and actions keys.
 *
 * Instead of using hardcoded key names this mixin provides a way to map keys to abstract actions
 * and use different key mappings based on context.
 *
 * All components should implement this mixin if it handles keyboard events for navigation or actions,
 * e.g. buttons, lists, popups, etc.
 *
 * Navigation keys mapped directly:
 * - 'up'
 * - 'down'
 * - 'left'
 * - 'right'
 * - 'tab'
 * - 'home'
 * - 'end'
 *
 * Action keys:
 * - 'action' (Enter key)
 *
 * Special keys:
 * - 'space' (Space key)
 *
 * Space is separated from action keys as it is
 *  - not always trigger the same action as the enter key
 *  - not every platform has a space key equivalent for example on a TV remote or gamepad
 *  - often used for scrolling as well.
 *
 *  From the above lists only 'up', 'down', 'left', 'right', 'action' and 'abort' are mandatory to implement,
 *  because those are essential for spatial navigation and basic actions and all platforms have equivalents for those.
 */
export const KeyToActionMixin = (superClass) => {
    class InnerMixinClass extends superClass {
        constructor() {
            super(...arguments);
            /** @internal */
            this.spatialNavigationContext = providerUtils.consume({
                host: this,
                context: SpatialNavigationProvider.Context,
            });
        }
        /**
         * @see KeyToActionInterface.getKeyboardNavMode
         * @internal
         */
        getKeyboardNavMode() {
            var _a;
            const provider = (_a = this.spatialNavigationContext) === null || _a === void 0 ? void 0 : _a.value;
            return provider ? NAV_MODES.SPATIAL : NAV_MODES.DEFAULT;
        }
        /**
         * @see KeyToActionInterface.getMappedKeyFromEvent
         * @internal
         */
        getActionForKeyEvent(evt, applyWritingDirection = false) {
            var _a, _b, _c;
            const mapping = (_c = (_b = (_a = this.spatialNavigationContext) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.keyToActionMap) !== null && _c !== void 0 ? _c : DEFAULT_KEY_TO_ACTION;
            const key = mapping[evt.key];
            if (applyWritingDirection) {
                const isRtl = window.getComputedStyle(this).direction === 'rtl';
                if (!isRtl)
                    return key;
                if (key === ACTIONS.LEFT)
                    return ACTIONS.RIGHT;
                if (key === ACTIONS.RIGHT)
                    return ACTIONS.LEFT;
            }
            return key;
        }
        /**
         * @see KeyToActionInterface.isDirectionAction
         * @internal
         */
        isDirectionAction(action) {
            return action === ACTIONS.UP || action === ACTIONS.DOWN || action === ACTIONS.LEFT || action === ACTIONS.RIGHT;
        }
    }
    // Cast return type to your mixin's interface intersected with the superClass type
    return InnerMixinClass;
};
