// @momentum-design/components/dist/utils/mixins/PreventScrollMixin.js
export const PreventScrollMixin = (superClass) => {
    class PreventScroll extends superClass {
        constructor() {
            super(...arguments);
            /**
             * @internal
             */
            this.isPreventScrollActive = false;
            /**
             * @internal
             */
            this.previousDocumentBodyStyleOverflow = '';
        }
        activatePreventScroll() {
            if (this.preventScroll && !this.isPreventScrollActive) {
                this.isPreventScrollActive = true;
                // Store the previous body overflow style
                this.previousDocumentBodyStyleOverflow = document.body.style.overflow;
                // Set body overflow to hidden to prevent scrolling
                document.body.style.overflow = 'hidden';
            }
        }
        deactivatePreventScroll() {
            if (this.isPreventScrollActive) {
                this.isPreventScrollActive = false;
                // Restore the previous body overflow style
                document.body.style.overflow = this.previousDocumentBodyStyleOverflow;
            }
        }
    }
    return PreventScroll;
};
