// @momentum-design/components/dist/utils/mixins/TabListComponentMixin.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { property } from 'lit/decorators.js';
import Tab from '../../components/tab';
import { findTab } from '../../components/tablist/tablist.utils';
export const TabListComponentMixin = (superClass) => {
    class InnerMixinClass extends superClass {
        constructor(...rest) {
            super(...rest);
            // Reason for assertion below: https://github.com/microsoft/TypeScript/issues/28357
            this.addEventListener('activechange', this.handleNestedTabActiveChange);
        }
        /**
         * Observe the tablist element for changes in the activeTabId attribute.
         * Find the new tab with the new activeTabId.
         * If the new tab is not found, then do nothing.
         *
         * If the new tab exists:
         * Set the new tab as active.
         *
         * Fire the tab change event.
         *
         * @param changedProperties - Map of changed properties with old values.
         */
        async update(changedProperties) {
            super.update(changedProperties);
            if (changedProperties.has('activeTabId')) {
                if (!this.navItems || !this.activeTabId) {
                    return;
                }
                const newTab = findTab(this.navItems, this.activeTabId);
                if (!(newTab instanceof Tab)) {
                    return;
                }
                this.setActiveTab(newTab);
                // Only fire the change event if activeTabId was previously set (not the initial assignment).
                // On first render, old value is undefined by default, which means it's being set for the first time.
                if (changedProperties.get('activeTabId')) {
                    this.fireTabChangeEvent(newTab);
                }
            }
        }
        /**
         * Removes active attribute from all tabs and sets active on the new tab.
         *
         * @param newTab - The new active tab.
         * @internal
         */
        setActiveTab(newTab) {
            var _a;
            (_a = this.navItems) === null || _a === void 0 ? void 0 : _a.forEach(tab => {
                if (tab === newTab) {
                    tab.setAttribute('active', '');
                }
                else {
                    tab.removeAttribute('active');
                }
            });
        }
        /**
         * Dispatch the change event.
         *
         * @param newTab - the new tab that is active.
         * @internal
         */
        fireTabChangeEvent(newTab) {
            const event = new CustomEvent('change', {
                detail: { tabId: newTab.tabId },
            });
            this.dispatchEvent(event);
        }
        /**
         * Handle the tab active change event fired from the nested tab.
         *
         * @param event - Custom Event fired from the nested tab.
         * @internal
         */
        async handleNestedTabActiveChange(event) {
            event.stopPropagation();
            const tab = event.target;
            if (!(tab instanceof Tab)) {
                return;
            }
            this.activeTabId = tab.tabId;
        }
    }
    __decorate([
        property({ type: String, attribute: 'active-tab-id', reflect: true }),
        __metadata("design:type", String)
    ], InnerMixinClass.prototype, "activeTabId", void 0);
    // Cast return type to your mixin's interface intersected with the superClass type
    return InnerMixinClass;
};
