// @momentum-design/components/dist/react/avatarbutton/index.js
import * as React from 'react';
import { createComponent } from '@lit/react';
import Component from '../../components/avatarbutton';
import { TAG_NAME } from '../../components/avatarbutton/avatarbutton.constants';
/**
 * The `mdc-avatarbutton` component is an interactive, clickable version of `mdc-avatar`.
 * It wraps the avatar component in a button, making it focusable and actionable.
 *
 * This component extends `buttonsimple` and includes all avatar display features (photo, initials, icon, counter, presence).
 * Use CSS parts to customize the avatar's appearance inside the button.
 *
 * ## Accessibility
 * - Always provide an `aria-label` attribute to describe the button's purpose
 *
 * @dependency mdc-avatar
 *
 * @event click - (React: onClick) Dispatched when the avatar button is clicked.
 * @event keydown - (React: onKeyDown) Dispatched when a key is pressed on the avatar button.
 * @event keyup - (React: onKeyUp) Dispatched when a key is released on the avatar button.
 * @event focus - (React: onFocus) Dispatched when the avatar button receives focus.
 *
 * @tagname mdc-avatarbutton
 *
 * @cssproperty --mdc-avatarbutton-overlay-background-color-rest - Background color of the overlay in rest state.
 * @cssproperty --mdc-avatarbutton-overlay-background-color-hover - Background color of the overlay in hover state.
 * @cssproperty --mdc-avatarbutton-overlay-background-color-active - Background color of the overlay in active state.
 * @cssproperty --mdc-avatarbutton-default-background-color - Default background color of the avatar.
 * @cssproperty --mdc-avatarbutton-default-foreground-color - Default foreground color of the avatar.
 * @cssproperty --mdc-avatarbutton-loading-indicator-background-color - Background color of the loading indicator.
 * @cssproperty --mdc-avatarbutton-loading-indicator-foreground-color - Foreground color of the loading indicator.
 * @cssproperty --mdc-avatarbutton-loading-overlay-background-color - Background color of the loading overlay.
 *
 * @csspart overlay - The overlay part of the avatar button.
 * @csspart content - The main content of the avatar.
 * @csspart photo - The photo part of the avatar.
 * @csspart presence - The presence indicator part of the avatar.
 * @csspart loading-wrapper - The wrapper for the loading indicator of the avatar.
 * @csspart loader - The loading indicator part of the avatar.
 */
const reactWrapper = createComponent({
    tagName: TAG_NAME,
    elementClass: Component,
    react: React,
    events: {
        onClick: 'click',
        onKeyDown: 'keydown',
        onKeyUp: 'keyup',
        onFocus: 'focus',
    },
    displayName: 'AvatarButton',
});
export default reactWrapper;
