// @momentum-design/components/dist/components/menusection/menusection.component.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { Component } from '../../models';
import providerUtils from '../../utils/provider';
import { ROLE } from '../../utils/roles';
import SideNavigation from '../sidenavigation/sidenavigation.component';
import styles from './menusection.styles';
import { DEFAULTS } from './menusection.constants';
/**
 * `mdc-menusection` is a container element used to group a set of menu items.
 *
 * This component supports a mix of `menuitem`, `menuitemcheckbox`, and `menuitemradio` components.
 *
 * - If multiple `menuitemradio` components are slotted into the section, the group enforces a single-selection rule:
 *   only one radio item can be selected at a time.
 * - If `menuitemcheckbox` components are included, their checked state can be toggled independently.
 *
 * @tagname mdc-menusection
 *
 * @slot - Default slot for inserting `menuitem`, `menuitemcheckbox`, or `menuitemradio`
 *
 * @event change - (React: onChange) This event is dispatched when a `menuitemcheckbox`, or `menuitemradio` changes.
 * @event action - (React: onAction) This event is dispatched when a `menuitem` selected.
 */
class MenuSection extends Component {
    constructor() {
        super(...arguments);
        /**
         * The aria-label for the section.
         * This is used for accessibility purposes to describe the section.
         * If not provided, it defaults to the `headerText`.
         */
        this.ariaLabel = null;
        /**
         * The primary headerText of the list item.
         * This appears on the leading side of the list item.
         */
        this.headerText = null;
        /**
         * Whether to show a divider below the section header.
         * This is useful for visually separating sections in the menu.
         */
        this.showDivider = false;
        /**
         * The variant of the divider.
         * Can be set to 'solid' or 'gradient'.
         *
         * Keep 'solid' if used in MenuPopovers, as it is the default style.
         *
         * @default 'solid'
         */
        this.dividerVariant = DEFAULTS.DIVIDER_VARIANT;
        /**
         * Shows or hides the section headers based on the expanded state of the side navigation.
         *
         * @internal
         */
        this.hideHeaderText = false;
        /**
         * @internal
         */
        this.sideNavigationContext = providerUtils.consume({ host: this, context: SideNavigation.Context });
    }
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('role', ROLE.GROUP);
    }
    update(changedProperties) {
        var _a;
        super.update(changedProperties);
        if ((changedProperties.has('ariaLabel') || changedProperties.has('headerText')) &&
            (!this.ariaLabel || this.ariaLabel === changedProperties.get('headerText'))) {
            // Because IDREF attribute reflection does not work across light and shadow DOM, we either set the
            // `aria-label` directly or use the `ariaLabelledByElements`.
            // Since the later one just released in the major browsers, we do the first one for now.
            // more details: https://nolanlawson.com/2022/11/28/shadow-dom-and-accessibility-the-trouble-with-aria/
            this.ariaLabel = this.headerText || '';
        }
        const context = (_a = this.sideNavigationContext) === null || _a === void 0 ? void 0 : _a.value;
        if (!context)
            return;
        const { expanded } = context;
        this.hideHeaderText = !expanded;
    }
    renderHeader() {
        var _a, _b;
        if (this.headerText) {
            return html ` <mdc-listheader
        part="header ${((_b = (_a = this.sideNavigationContext) === null || _a === void 0 ? void 0 : _a.value) === null || _b === void 0 ? void 0 : _b.expanded) ? 'align-header' : ''}"
        header-text="${this.headerText}"
        prefix-icon="${ifDefined(this.prefixIcon)}"
      >
      </mdc-listheader>`;
        }
        return null;
    }
    render() {
        return html `
      ${!this.hideHeaderText ? this.renderHeader() : null}
      <slot part="container"></slot>
      ${this.showDivider ? html `<mdc-divider variant="${this.dividerVariant}" part="divider"></mdc-divider>` : null}
    `;
    }
}
MenuSection.styles = [...Component.styles, ...styles];
__decorate([
    property({ type: String, reflect: true, attribute: 'aria-label' }),
    __metadata("design:type", Object)
], MenuSection.prototype, "ariaLabel", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'header-text' }),
    __metadata("design:type", Object)
], MenuSection.prototype, "headerText", void 0);
__decorate([
    property({ type: String, attribute: 'prefix-icon' }),
    __metadata("design:type", String)
], MenuSection.prototype, "prefixIcon", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'show-divider' }),
    __metadata("design:type", Object)
], MenuSection.prototype, "showDivider", void 0);
__decorate([
    property({ type: String, reflect: true, attribute: 'divider-variant' }),
    __metadata("design:type", Object)
], MenuSection.prototype, "dividerVariant", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'hide-header-text' }),
    __metadata("design:type", Object)
], MenuSection.prototype, "hideHeaderText", void 0);
export default MenuSection;
